
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Primitives, Listed by Category &mdash; GPI Data Pipeline 1.3.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc_gpi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.3.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="GPI Data Pipeline 1.3.0 documentation" href="../index.html" />
    <link rel="up" title="GPI Data Pipeline Userâ€™s Guide" href="index.html" />
    <link rel="next" title="Logging and History records" href="logging.html" />
    <link rel="prev" title="The Calibration Database" href="calibdb.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="logging.html" title="Logging and History records"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="calibdb.html" title="The Calibration Database"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">GPI Data Pipeline 1.3.0 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">GPI Data Pipeline User&#8217;s Guide</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/GPI_Logo_v2012.png" alt="Logo"/>
            </a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Primitives, Listed by Category</a><ul>
<li><a class="reference internal" href="#spectralscience">SpectralScience</a></li>
<li><a class="reference internal" href="#polarimetricscience">PolarimetricScience</a></li>
<li><a class="reference internal" href="#calibration">Calibration</a></li>
</ul>
</li>
<li><a class="reference internal" href="#primitive-detailed-documentation">Primitive Detailed Documentation</a><ul>
<li><a class="reference internal" href="#display-raw-image-with-gpitv">Display raw image with GPItv</a></li>
<li><a class="reference internal" href="#correct-for-interpixel-capacitance">Correct for Interpixel Capacitance</a></li>
<li><a class="reference internal" href="#flag-quicklook">Flag Quicklook</a></li>
<li><a class="reference internal" href="#load-wavelength-calibration">Load Wavelength Calibration</a></li>
<li><a class="reference internal" href="#load-high-res-psfs">Load High-Res PSFs</a></li>
<li><a class="reference internal" href="#load-polarimetry-spot-calibration">Load Polarimetry Spot Calibration</a></li>
<li><a class="reference internal" href="#load-instrumental-polarization-calibration">Load Instrumental Polarization Calibration</a></li>
<li><a class="reference internal" href="#check-data-quality">Check Data Quality</a></li>
<li><a class="reference internal" href="#subtract-dark-background">Subtract Dark Background</a></li>
<li><a class="reference internal" href="#remove-persistence-from-previous-images">Remove Persistence from Previous Images</a></li>
<li><a class="reference internal" href="#subtract-thermal-sky-background-if-k-band">Subtract Thermal/Sky Background if K band</a></li>
<li><a class="reference internal" href="#clean-cosmic-rays">Clean Cosmic Rays</a></li>
<li><a class="reference internal" href="#apply-reference-pixel-correction">Apply Reference Pixel Correction</a></li>
<li><a class="reference internal" href="#update-spot-shifts-for-flexure">Update Spot Shifts for Flexure</a></li>
<li><a class="reference internal" href="#flexure-2d-x-correlation-with-polcal">Flexure 2D x correlation with polcal</a></li>
<li><a class="reference internal" href="#destripe-for-darks-only">Destripe for Darks Only</a></li>
<li><a class="reference internal" href="#destripe-science-image">Destripe science image</a></li>
<li><a class="reference internal" href="#interpolate-bad-pixels-in-2d-frame">Interpolate bad pixels in 2D frame</a></li>
<li><a class="reference internal" href="#combine-2d-images">Combine 2D images</a></li>
<li><a class="reference internal" href="#combine-2d-thermal-sky-backgrounds">Combine 2D Thermal/Sky Backgrounds</a></li>
<li><a class="reference internal" href="#d-wavelength-solution-developer">2D Wavelength Solution Developer</a></li>
<li><a class="reference internal" href="#quick-wavelength-solution-update">Quick Wavelength Solution Update</a></li>
<li><a class="reference internal" href="#d-wavelength-solution">2D Wavelength Solution</a></li>
<li><a class="reference internal" href="#measure-wavelength-calibration">Measure Wavelength Calibration</a></li>
<li><a class="reference internal" href="#measure-polarization-spot-calibration">Measure Polarization Spot Calibration</a></li>
<li><a class="reference internal" href="#assemble-polarization-cube">Assemble Polarization Cube</a></li>
<li><a class="reference internal" href="#assemble-undispersed-image">Assemble Undispersed Image</a></li>
<li><a class="reference internal" href="#assemble-spectral-datacube-using-mlens-psf">Assemble Spectral Datacube using mlens PSF</a></li>
<li><a class="reference internal" href="#assemble-spectral-datacube-using-epsf">Assemble Spectral Datacube using ePSF</a></li>
<li><a class="reference internal" href="#assemble-spectral-datacube">Assemble Spectral Datacube</a></li>
<li><a class="reference internal" href="#filter-datacube-spatially">Filter datacube spatially</a></li>
<li><a class="reference internal" href="#noise-and-flux-analysis">Noise and Flux Analysis</a></li>
<li><a class="reference internal" href="#divide-by-lenslet-flat-field">Divide by Lenslet Flat Field</a></li>
<li><a class="reference internal" href="#divide-by-spectral-flat-field">Divide by Spectral Flat Field</a></li>
<li><a class="reference internal" href="#remove-flat-lamp-spectrum">Remove Flat Lamp spectrum</a></li>
<li><a class="reference internal" href="#interpolate-wavelength-axis">Interpolate Wavelength Axis</a></li>
<li><a class="reference internal" href="#subtract-thermal-sky-background-cube-if-k-band">Subtract Thermal/Sky Background Cube if K band</a></li>
<li><a class="reference internal" href="#check-for-closed-loop-coronagraphic-image">Check for closed-loop coronagraphic image</a></li>
<li><a class="reference internal" href="#correct-distortion">Correct Distortion</a></li>
<li><a class="reference internal" href="#measure-satellite-spot-locations">Measure satellite spot locations</a></li>
<li><a class="reference internal" href="#measure-gpi-distortion-from-grid-pattern">Measure GPI distortion from grid pattern</a></li>
<li><a class="reference internal" href="#measure-star-position-for-polarimetry">Measure Star Position for Polarimetry</a></li>
<li><a class="reference internal" href="#measure-satellite-spot-peak-fluxes">Measure satellite spot peak fluxes</a></li>
<li><a class="reference internal" href="#correct-for-atmospheric-differential-refraction">Correct for Atmospheric Differential Refraction</a></li>
<li><a class="reference internal" href="#divide-by-telluric-transmission">Divide by Telluric Transmission</a></li>
<li><a class="reference internal" href="#interpolate-bad-pixels-in-cube">Interpolate bad pixels in cube</a></li>
<li><a class="reference internal" href="#extract-one-spectrum">Extract one spectrum</a></li>
<li><a class="reference internal" href="#calibrate-photometric-flux">Calibrate Photometric Flux</a></li>
<li><a class="reference internal" href="#extract-one-spectrum-plots">Extract one spectrum, plots</a></li>
<li><a class="reference internal" href="#extract-1d-spectrum-from-a-datacube">Extract 1D spectrum from a datacube</a></li>
<li><a class="reference internal" href="#collapse-datacube">Collapse datacube</a></li>
<li><a class="reference internal" href="#calibrate-astrometry-from-binary-using-separation-and-pa">Calibrate astrometry from binary (using separation and PA)</a></li>
<li><a class="reference internal" href="#simple-spectral-differential-imaging">Simple Spectral Differential Imaging</a></li>
<li><a class="reference internal" href="#speckle-alignment">Speckle alignment</a></li>
<li><a class="reference internal" href="#calibrate-astrometry-from-binary-using-6th-orbit-catalog">Calibrate astrometry from binary (using 6th orbit catalog)</a></li>
<li><a class="reference internal" href="#measure-contrast">Measure Contrast</a></li>
<li><a class="reference internal" href="#plot-the-satellite-spot-locations-vs-the-expected-location-from-wavelength-scaling">Plot the satellite spot locations vs. the expected location from wavelength scaling</a></li>
<li><a class="reference internal" href="#klip-algorithm-spectral-differential-imaging">KLIP algorithm Spectral Differential Imaging</a></li>
<li><a class="reference internal" href="#update-world-coordinates">Update World Coordinates</a></li>
<li><a class="reference internal" href="#stores-calibration-in-dataset">Stores calibration in dataset</a></li>
<li><a class="reference internal" href="#normalize-polarimetry-flat-field">Normalize polarimetry flat field</a></li>
<li><a class="reference internal" href="#create-lenslet-flat-field">Create Lenslet Flat Field</a></li>
<li><a class="reference internal" href="#smooth-a-3d-cube">Smooth a 3D Cube</a></li>
<li><a class="reference internal" href="#divide-by-polarized-flat-field">Divide by Polarized Flat Field</a></li>
<li><a class="reference internal" href="#subtract-mean-stellar-polarization-from-podc">Subtract Mean Stellar Polarization from podc</a></li>
<li><a class="reference internal" href="#rotate-north-up">Rotate North Up</a></li>
<li><a class="reference internal" href="#rotate-field-of-view-square">Rotate Field of View Square</a></li>
<li><a class="reference internal" href="#accumulate-images">Accumulate Images</a></li>
<li><a class="reference internal" href="#create-2d-low-frequency-flat">Create 2D Low Frequency Flat</a></li>
<li><a class="reference internal" href="#create-high-resolution-microlens-psf-model">Create High-Resolution Microlens PSF Model</a></li>
<li><a class="reference internal" href="#create-microphonics-noise-model">Create microphonics noise model</a></li>
<li><a class="reference internal" href="#combine-2d-dark-images">Combine 2D dark images</a></li>
<li><a class="reference internal" href="#creates-a-thermal-sky-background-datacube">Creates a thermal/sky background datacube</a></li>
<li><a class="reference internal" href="#find-hot-bad-pixels-from-darks">Find Hot Bad Pixels from Darks</a></li>
<li><a class="reference internal" href="#find-cold-bad-pixels-from-flats">Find Cold Bad Pixels from Flats</a></li>
<li><a class="reference internal" href="#generate-combined-bad-pixel-map">Generate Combined Bad Pixel Map</a></li>
<li><a class="reference internal" href="#clean-polarization-pairs-via-double-difference">Clean Polarization Pairs via Double Difference</a></li>
<li><a class="reference internal" href="#basic-adi">Basic ADI</a></li>
<li><a class="reference internal" href="#primitive-to-interface-tloci-code-and-dependecies-for-psf-subtraction-with-the-gpi-pipeline">Primitive to interface TLOCI code and dependecies for PSF subtraction with the GPI pipeline.</a></li>
<li><a class="reference internal" href="#adi-with-loci">ADI with LOCI</a></li>
<li><a class="reference internal" href="#combine-wavelength-calibrations-locations">Combine Wavelength Calibrations locations</a></li>
<li><a class="reference internal" href="#combine-wavelength-calibrations">Combine Wavelength Calibrations</a></li>
<li><a class="reference internal" href="#klip-adi-for-pol-mode">KLIP ADI for Pol Mode</a></li>
<li><a class="reference internal" href="#populate-flexure-shifts-vs-elevation-table">Populate Flexure Shifts vs Elevation Table</a></li>
<li><a class="reference internal" href="#klip-algorithm-angular-differential-imaging-with-center-forced">KLIP algorithm Angular Differential Imaging With Center Forced</a></li>
<li><a class="reference internal" href="#klip-algorithm-angular-differential-imaging">KLIP algorithm Angular Differential Imaging</a></li>
<li><a class="reference internal" href="#advanced-klip-adi-for-pol-mode">Advanced KLIP ADI for Pol Mode</a></li>
<li><a class="reference internal" href="#klip-algorithm-adi-sdi">KLIP algorithm ADI + SDI</a></li>
<li><a class="reference internal" href="#simple-sdi-of-post-adi-residual">Simple SDI of post ADI residual</a></li>
<li><a class="reference internal" href="#combine-polarization-sequence-via-double-difference">Combine Polarization Sequence via Double Difference</a></li>
<li><a class="reference internal" href="#combine-polarization-sequence">Combine Polarization Sequence</a></li>
<li><a class="reference internal" href="#quality-check-wavelength-calibration">Quality Check Wavelength Calibration</a></li>
<li><a class="reference internal" href="#combine-3d-datacubes">Combine 3D Datacubes</a></li>
<li><a class="reference internal" href="#median-combine-adi-datacubes">Median Combine ADI datacubes</a></li>
<li><a class="reference internal" href="#pad-wavelength-calibration-edges">Pad Wavelength Calibration Edges</a></li>
<li><a class="reference internal" href="#flexure-quicklook-for-spectra-lsqr-microlens-psf">Flexure Quicklook for Spectra (Lsqr, microlens psf)</a></li>
<li><a class="reference internal" href="#flexure-quicklook-for-pol-lsqr-microlens-psf">Flexure Quicklook for Pol (Lsqr, microlens psf)</a></li>
<li><a class="reference internal" href="#subtract-mean-stellar-polarization">Subtract Mean Stellar Polarization</a></li>
<li><a class="reference internal" href="#assemble-polarization-datacube-lsqr-microlens-psf">Assemble Polarization Datacube (Lsqr, microlens psf)</a></li>
<li><a class="reference internal" href="#assemble-spectral-datacube-lsqr-microlens-psf">Assemble Spectral Datacube (Lsqr, microlens psf)</a></li>
<li><a class="reference internal" href="#flexure-2d-x-correlation-with-wavecal-model-perpendicular">Flexure 2D x correlation with wavecal model (perpendicular)</a></li>
<li><a class="reference internal" href="#flexure-2d-x-correlation-with-ulens-and-polcal-models">Flexure 2D x correlation with ulens and polcal models</a></li>
<li><a class="reference internal" href="#flexure-2d-x-correlation-with-wavecal-model">Flexure 2D x correlation with wavecal model</a></li>
<li><a class="reference internal" href="#insert-planet-into-datacube">Insert Planet into datacube</a></li>
<li><a class="reference internal" href="#save-accumulated-stack">Save Accumulated Stack</a></li>
<li><a class="reference internal" href="#save-output">Save Output</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="calibdb.html"
                        title="previous chapter">The Calibration Database</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="logging.html"
                        title="next chapter">Logging and History records</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/usage/primitives.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="primitives-listed-by-category">
<span id="primitives"></span><h1>Primitives, Listed by Category<a class="headerlink" href="#primitives-listed-by-category" title="Permalink to this headline">Â¶</a></h1>
<p>This page documents all available pipeline primitives, currently 125 in total.</p>
<p>First we list the available primitives in each mode, and below provide the detailed documentation including the
parameter arguments for each.</p>
<p>Primitives each have an &#8220;<strong>order</strong>&#8221;, which is a floating point number that defines the <em>default</em> ordering when added to a recipe. Smaller numbers come earlier in the execution sequence. You can change the order arbitrarily in the Recipe Editor of course. Notionally, orders &lt;2 are actions on 2D files, orders between 2-4 are actions on datacubes, and orders &gt; 4 are actions on entire observation sequences, but these are not strictly enforced.</p>
<p>(<em>Note: For simplicity, some engineering and software testing related primitives not intended for end users are not listed in the following tables.</em>)</p>
<p><a class="reference internal" href="#spectralscience"><em>SpectralScience</em></a>
<a class="reference internal" href="#polarimetricscience"><em>PolarimetricScience</em></a>
<a class="reference internal" href="#calibration"><em>Calibration</em></a></p>
<div class="section" id="spectralscience">
<span id="id1"></span><h2>SpectralScience<a class="headerlink" href="#spectralscience" title="Permalink to this headline">Â¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="96%" />
<col width="1%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Order</th>
<th class="head">Primitives relevant to SpectralScience     (60 total)</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0.01</td>
<td><a class="reference internal" href="#displayrawimagewithgpitv"><em>Display raw image with GPItv</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.10</td>
<td><a class="reference internal" href="#flagquicklook"><em>Flag Quicklook</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>0.10</td>
<td><a class="reference internal" href="#correctforinterpixelcapacitance"><em>Correct for Interpixel Capacitance</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.50</td>
<td><a class="reference internal" href="#loadhigh-respsfs"><em>Load High-Res PSFs</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>0.50</td>
<td><a class="reference internal" href="#loadwavelengthcalibration"><em>Load Wavelength Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.90</td>
<td><a class="reference internal" href="#checkdataquality"><em>Check Data Quality</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.10</td>
<td><a class="reference internal" href="#subtractdarkbackground"><em>Subtract Dark Background</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.20</td>
<td><a class="reference internal" href="#subtractthermal-skybackgroundifkband"><em>Subtract Thermal/Sky Background if K band</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.20</td>
<td><a class="reference internal" href="#removepersistencefrompreviousimages"><em>Remove Persistence from Previous Images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.25</td>
<td><a class="reference internal" href="#applyreferencepixelcorrection"><em>Apply Reference Pixel Correction</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.34</td>
<td><a class="reference internal" href="#updatespotshiftsforflexure"><em>Update Spot Shifts for Flexure</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.35</td>
<td><a class="reference internal" href="#destripescienceimage"><em>Destripe science image</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.40</td>
<td><a class="reference internal" href="#interpolatebadpixelsin2dframe"><em>Interpolate bad pixels in 2D frame</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.50</td>
<td><a class="reference internal" href="#combine2dimages"><em>Combine 2D images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.00</td>
<td><a class="reference internal" href="#assemblespectraldatacube"><em>Assemble Spectral Datacube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.00</td>
<td><a class="reference internal" href="#assemblespectraldatacubeusingepsf"><em>Assemble Spectral Datacube using ePSF</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.00</td>
<td><a class="reference internal" href="#assemblespectraldatacubeusingmlenspsf"><em>Assemble Spectral Datacube using mlens PSF</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.10</td>
<td><a class="reference internal" href="#filterdatacubespatially"><em>Filter datacube spatially</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.20</td>
<td><a class="reference internal" href="#dividebylensletflatfield"><em>Divide by Lenslet Flat Field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.20</td>
<td><a class="reference internal" href="#dividebyspectralflatfield"><em>Divide by Spectral Flat Field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.30</td>
<td><a class="reference internal" href="#interpolatewavelengthaxis"><em>Interpolate Wavelength Axis</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.35</td>
<td><a class="reference internal" href="#subtractthermal-skybackgroundcubeifkband"><em>Subtract Thermal/Sky Background Cube if K band</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.41</td>
<td><a class="reference internal" href="#checkforclosed-loopcoronagraphicimage"><em>Check for closed-loop coronagraphic image</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.44</td>
<td><a class="reference internal" href="#measuresatellitespotlocations"><em>Measure satellite spot locations</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.44</td>
<td><a class="reference internal" href="#correctdistortion"><em>Correct Distortion</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.45</td>
<td><a class="reference internal" href="#correctforatmosphericdifferentialrefraction"><em>Correct for Atmospheric Differential Refraction</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.45</td>
<td><a class="reference internal" href="#measuresatellitespotpeakfluxes"><em>Measure satellite spot peak fluxes</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.50</td>
<td><a class="reference internal" href="#interpolatebadpixelsincube"><em>Interpolate bad pixels in cube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.50</td>
<td><a class="reference internal" href="#dividebytellurictransmission"><em>Divide by Telluric Transmission</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.51</td>
<td><a class="reference internal" href="#extractonespectrum"><em>Extract one spectrum</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.51</td>
<td><a class="reference internal" href="#calibratephotometricflux"><em>Calibrate Photometric Flux</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.51</td>
<td><a class="reference internal" href="#extractonespectrum-plots"><em>Extract one spectrum, plots</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.52</td>
<td><a class="reference internal" href="#extract1dspectrumfromadatacube"><em>Extract 1D spectrum from a datacube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.60</td>
<td><a class="reference internal" href="#collapsedatacube"><em>Collapse datacube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.61</td>
<td><a class="reference internal" href="#specklealignment"><em>Speckle alignment</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.61</td>
<td><a class="reference internal" href="#simplespectraldifferentialimaging"><em>Simple Spectral Differential Imaging</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.70</td>
<td><a class="reference internal" href="#plotthesatellitespotlocationsvs-theexpectedlocationfromwavelengthscaling"><em>Plot the satellite spot locations vs. the expected location from wavelength scaling</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.70</td>
<td><a class="reference internal" href="#measurecontrast"><em>Measure Contrast</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.80</td>
<td><a class="reference internal" href="#klipalgorithmspectraldifferentialimaging"><em>KLIP algorithm Spectral Differential Imaging</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.90</td>
<td><a class="reference internal" href="#updateworldcoordinates"><em>Update World Coordinates</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>3.50</td>
<td><a class="reference internal" href="#smootha3dcube"><em>Smooth a 3D Cube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>3.90</td>
<td><a class="reference internal" href="#rotatefieldofviewsquare"><em>Rotate Field of View Square</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>3.90</td>
<td><a class="reference internal" href="#rotatenorthup"><em>Rotate North Up</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.00</td>
<td><a class="reference internal" href="#accumulateimages"><em>Accumulate Images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.10</td>
<td><a class="reference internal" href="#basicadi"><em>Basic ADI</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.11</td>
<td><a class="reference internal" href="#primitivetointerfacetlocicodeanddependeciesforpsfsubtractionwiththegpipipeline"><em>Primitive to interface TLOCI code and dependecies for PSF subtraction with the GPI pipeline.</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.11</td>
<td><a class="reference internal" href="#adiwithloci"><em>ADI with LOCI</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.20</td>
<td><a class="reference internal" href="#klipalgorithmangulardifferentialimagingwithcenterforced"><em>KLIP algorithm Angular Differential Imaging With Center Forced</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.20</td>
<td><a class="reference internal" href="#klipalgorithmangulardifferentialimaging"><em>KLIP algorithm Angular Differential Imaging</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.20</td>
<td><a class="reference internal" href="#klipalgorithmadi-sdi"><em>KLIP algorithm ADI + SDI</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.30</td>
<td><a class="reference internal" href="#simplesdiofpostadiresidual"><em>Simple SDI of post ADI residual</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.50</td>
<td><a class="reference internal" href="#mediancombineadidatacubes"><em>Median Combine ADI datacubes</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.50</td>
<td><a class="reference internal" href="#combine3ddatacubes"><em>Combine 3D Datacubes</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>5.00</td>
<td><a class="reference internal" href="#insertplanetintodatacube"><em>Insert Planet into datacube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>5.00</td>
<td><a class="reference internal" href="#flexurequicklookforspectra-lsqr-microlenspsf"><em>Flexure Quicklook for Spectra (Lsqr, microlens psf)</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>5.00</td>
<td><a class="reference internal" href="#assemblespectraldatacube-lsqr-microlenspsf"><em>Assemble Spectral Datacube (Lsqr, microlens psf)</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>5.00</td>
<td><a class="reference internal" href="#flexure2dxcorrelationwithwavecalmodel"><em>Flexure 2D x correlation with wavecal model</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>5.00</td>
<td><a class="reference internal" href="#flexure2dxcorrelationwithwavecalmodel-perpendicular"><em>Flexure 2D x correlation with wavecal model (perpendicular)</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>10.00</td>
<td><a class="reference internal" href="#saveaccumulatedstack"><em>Save Accumulated Stack</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>10.00</td>
<td><a class="reference internal" href="#saveoutput"><em>Save Output</em></a></td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="polarimetricscience">
<span id="id2"></span><h2>PolarimetricScience<a class="headerlink" href="#polarimetricscience" title="Permalink to this headline">Â¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="94%" />
<col width="1%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Order</th>
<th class="head">Primitives relevant to PolarimetricScience     (45 total)</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0.01</td>
<td><a class="reference internal" href="#displayrawimagewithgpitv"><em>Display raw image with GPItv</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.10</td>
<td><a class="reference internal" href="#correctforinterpixelcapacitance"><em>Correct for Interpixel Capacitance</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>0.10</td>
<td><a class="reference internal" href="#flagquicklook"><em>Flag Quicklook</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.50</td>
<td><a class="reference internal" href="#loadhigh-respsfs"><em>Load High-Res PSFs</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>0.51</td>
<td><a class="reference internal" href="#loadpolarimetryspotcalibration"><em>Load Polarimetry Spot Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.52</td>
<td><a class="reference internal" href="#loadinstrumentalpolarizationcalibration"><em>Load Instrumental Polarization Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>0.90</td>
<td><a class="reference internal" href="#checkdataquality"><em>Check Data Quality</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.10</td>
<td><a class="reference internal" href="#subtractdarkbackground"><em>Subtract Dark Background</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.20</td>
<td><a class="reference internal" href="#subtractthermal-skybackgroundifkband"><em>Subtract Thermal/Sky Background if K band</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.20</td>
<td><a class="reference internal" href="#removepersistencefrompreviousimages"><em>Remove Persistence from Previous Images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.25</td>
<td><a class="reference internal" href="#applyreferencepixelcorrection"><em>Apply Reference Pixel Correction</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.34</td>
<td><a class="reference internal" href="#updatespotshiftsforflexure"><em>Update Spot Shifts for Flexure</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.34</td>
<td><a class="reference internal" href="#flexure2dxcorrelationwithpolcal"><em>Flexure 2D x correlation with polcal</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.35</td>
<td><a class="reference internal" href="#destripescienceimage"><em>Destripe science image</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.40</td>
<td><a class="reference internal" href="#interpolatebadpixelsin2dframe"><em>Interpolate bad pixels in 2D frame</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.50</td>
<td><a class="reference internal" href="#combine2dimages"><em>Combine 2D images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.00</td>
<td><a class="reference internal" href="#assemblepolarizationcube"><em>Assemble Polarization Cube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.10</td>
<td><a class="reference internal" href="#filterdatacubespatially"><em>Filter datacube spatially</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.20</td>
<td><a class="reference internal" href="#dividebylensletflatfield"><em>Divide by Lenslet Flat Field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.35</td>
<td><a class="reference internal" href="#subtractthermal-skybackgroundcubeifkband"><em>Subtract Thermal/Sky Background Cube if K band</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.41</td>
<td><a class="reference internal" href="#checkforclosed-loopcoronagraphicimage"><em>Check for closed-loop coronagraphic image</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.44</td>
<td><a class="reference internal" href="#correctdistortion"><em>Correct Distortion</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.44</td>
<td><a class="reference internal" href="#measurestarpositionforpolarimetry"><em>Measure Star Position for Polarimetry</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.50</td>
<td><a class="reference internal" href="#interpolatebadpixelsincube"><em>Interpolate bad pixels in cube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.60</td>
<td><a class="reference internal" href="#collapsedatacube"><em>Collapse datacube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.70</td>
<td><a class="reference internal" href="#measurecontrast"><em>Measure Contrast</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.90</td>
<td><a class="reference internal" href="#updateworldcoordinates"><em>Update World Coordinates</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>3.50</td>
<td><a class="reference internal" href="#smootha3dcube"><em>Smooth a 3D Cube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>3.50</td>
<td><a class="reference internal" href="#dividebypolarizedflatfield"><em>Divide by Polarized Flat Field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>3.85</td>
<td><a class="reference internal" href="#subtractmeanstellarpolarizationfrompodc"><em>Subtract Mean Stellar Polarization from podc</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>3.90</td>
<td><a class="reference internal" href="#rotatefieldofviewsquare"><em>Rotate Field of View Square</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>3.90</td>
<td><a class="reference internal" href="#rotatenorthup"><em>Rotate North Up</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.00</td>
<td><a class="reference internal" href="#accumulateimages"><em>Accumulate Images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.05</td>
<td><a class="reference internal" href="#cleanpolarizationpairsviadoubledifference"><em>Clean Polarization Pairs via Double Difference</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.20</td>
<td><a class="reference internal" href="#klipadiforpolmode"><em>KLIP ADI for Pol Mode</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.20</td>
<td><a class="reference internal" href="#advancedklipadiforpolmode"><em>Advanced KLIP ADI for Pol Mode</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.40</td>
<td><a class="reference internal" href="#combinepolarizationsequence"><em>Combine Polarization Sequence</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.40</td>
<td><a class="reference internal" href="#combinepolarizationsequenceviadoubledifference"><em>Combine Polarization Sequence via Double Difference</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.50</td>
<td><a class="reference internal" href="#combine3ddatacubes"><em>Combine 3D Datacubes</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>5.00</td>
<td><a class="reference internal" href="#flexure2dxcorrelationwithulensandpolcalmodels"><em>Flexure 2D x correlation with ulens and polcal models</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>5.00</td>
<td><a class="reference internal" href="#subtractmeanstellarpolarization"><em>Subtract Mean Stellar Polarization</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>5.00</td>
<td><a class="reference internal" href="#flexurequicklookforpol-lsqr-microlenspsf"><em>Flexure Quicklook for Pol (Lsqr, microlens psf)</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>5.00</td>
<td><a class="reference internal" href="#assemblepolarizationdatacube-lsqr-microlenspsf"><em>Assemble Polarization Datacube (Lsqr, microlens psf)</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>10.00</td>
<td><a class="reference internal" href="#saveoutput"><em>Save Output</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>10.00</td>
<td><a class="reference internal" href="#saveaccumulatedstack"><em>Save Accumulated Stack</em></a></td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="calibration">
<span id="id3"></span><h2>Calibration<a class="headerlink" href="#calibration" title="Permalink to this headline">Â¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="95%" />
<col width="1%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Order</th>
<th class="head">Primitives relevant to Calibration     (70 total)</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0.01</td>
<td><a class="reference internal" href="#displayrawimagewithgpitv"><em>Display raw image with GPItv</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.10</td>
<td><a class="reference internal" href="#correctforinterpixelcapacitance"><em>Correct for Interpixel Capacitance</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>0.10</td>
<td><a class="reference internal" href="#flagquicklook"><em>Flag Quicklook</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.50</td>
<td><a class="reference internal" href="#loadwavelengthcalibration"><em>Load Wavelength Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>0.50</td>
<td><a class="reference internal" href="#loadhigh-respsfs"><em>Load High-Res PSFs</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.51</td>
<td><a class="reference internal" href="#loadpolarimetryspotcalibration"><em>Load Polarimetry Spot Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>0.52</td>
<td><a class="reference internal" href="#loadinstrumentalpolarizationcalibration"><em>Load Instrumental Polarization Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>0.90</td>
<td><a class="reference internal" href="#checkdataquality"><em>Check Data Quality</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.10</td>
<td><a class="reference internal" href="#subtractdarkbackground"><em>Subtract Dark Background</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.20</td>
<td><a class="reference internal" href="#removepersistencefrompreviousimages"><em>Remove Persistence from Previous Images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.20</td>
<td><a class="reference internal" href="#subtractthermal-skybackgroundifkband"><em>Subtract Thermal/Sky Background if K band</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.25</td>
<td><a class="reference internal" href="#applyreferencepixelcorrection"><em>Apply Reference Pixel Correction</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.34</td>
<td><a class="reference internal" href="#updatespotshiftsforflexure"><em>Update Spot Shifts for Flexure</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.35</td>
<td><a class="reference internal" href="#destripefordarksonly"><em>Destripe for Darks Only</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.35</td>
<td><a class="reference internal" href="#destripescienceimage"><em>Destripe science image</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.40</td>
<td><a class="reference internal" href="#interpolatebadpixelsin2dframe"><em>Interpolate bad pixels in 2D frame</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.50</td>
<td><a class="reference internal" href="#combine2dimages"><em>Combine 2D images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.51</td>
<td><a class="reference internal" href="#combine2dthermal-skybackgrounds"><em>Combine 2D Thermal/Sky Backgrounds</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.70</td>
<td><a class="reference internal" href="#dwavelengthsolutiondeveloper"><em>2D Wavelength Solution Developer</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.70</td>
<td><a class="reference internal" href="#measurewavelengthcalibration"><em>Measure Wavelength Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.70</td>
<td><a class="reference internal" href="#quickwavelengthsolutionupdate"><em>Quick Wavelength Solution Update</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>1.70</td>
<td><a class="reference internal" href="#dwavelengthsolution"><em>2D Wavelength Solution</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>1.80</td>
<td><a class="reference internal" href="#measurepolarizationspotcalibration"><em>Measure Polarization Spot Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.00</td>
<td><a class="reference internal" href="#assemblespectraldatacubeusingmlenspsf"><em>Assemble Spectral Datacube using mlens PSF</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.00</td>
<td><a class="reference internal" href="#assemblepolarizationcube"><em>Assemble Polarization Cube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.00</td>
<td><a class="reference internal" href="#assemblespectraldatacube"><em>Assemble Spectral Datacube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.00</td>
<td><a class="reference internal" href="#assembleundispersedimage"><em>Assemble Undispersed Image</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.00</td>
<td><a class="reference internal" href="#assemblespectraldatacubeusingepsf"><em>Assemble Spectral Datacube using ePSF</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.10</td>
<td><a class="reference internal" href="#filterdatacubespatially"><em>Filter datacube spatially</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.20</td>
<td><a class="reference internal" href="#dividebyspectralflatfield"><em>Divide by Spectral Flat Field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.20</td>
<td><a class="reference internal" href="#dividebylensletflatfield"><em>Divide by Lenslet Flat Field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.25</td>
<td><a class="reference internal" href="#removeflatlampspectrum"><em>Remove Flat Lamp spectrum</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.30</td>
<td><a class="reference internal" href="#interpolatewavelengthaxis"><em>Interpolate Wavelength Axis</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.35</td>
<td><a class="reference internal" href="#subtractthermal-skybackgroundcubeifkband"><em>Subtract Thermal/Sky Background Cube if K band</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.41</td>
<td><a class="reference internal" href="#checkforclosed-loopcoronagraphicimage"><em>Check for closed-loop coronagraphic image</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.44</td>
<td><a class="reference internal" href="#measuresatellitespotlocations"><em>Measure satellite spot locations</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.44</td>
<td><a class="reference internal" href="#measuregpidistortionfromgridpattern"><em>Measure GPI distortion from grid pattern</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.44</td>
<td><a class="reference internal" href="#measurestarpositionforpolarimetry"><em>Measure Star Position for Polarimetry</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.45</td>
<td><a class="reference internal" href="#measuresatellitespotpeakfluxes"><em>Measure satellite spot peak fluxes</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.45</td>
<td><a class="reference internal" href="#correctforatmosphericdifferentialrefraction"><em>Correct for Atmospheric Differential Refraction</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.50</td>
<td><a class="reference internal" href="#interpolatebadpixelsincube"><em>Interpolate bad pixels in cube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.50</td>
<td><a class="reference internal" href="#dividebytellurictransmission"><em>Divide by Telluric Transmission</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.60</td>
<td><a class="reference internal" href="#calibrateastrometryfrombinary-usingseparationandpa"><em>Calibrate astrometry from binary (using separation and PA)</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.60</td>
<td><a class="reference internal" href="#collapsedatacube"><em>Collapse datacube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>2.61</td>
<td><a class="reference internal" href="#calibrateastrometryfrombinary-using6thorbitcatalog"><em>Calibrate astrometry from binary (using 6th orbit catalog)</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>2.90</td>
<td><a class="reference internal" href="#updateworldcoordinates"><em>Update World Coordinates</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>3.00</td>
<td><a class="reference internal" href="#storescalibrationindataset"><em>Stores calibration in dataset</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>3.20</td>
<td><a class="reference internal" href="#normalizepolarimetryflatfield"><em>Normalize polarimetry flat field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>3.20</td>
<td><a class="reference internal" href="#createlensletflatfield"><em>Create Lenslet Flat Field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>3.50</td>
<td><a class="reference internal" href="#dividebypolarizedflatfield"><em>Divide by Polarized Flat Field</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>3.50</td>
<td><a class="reference internal" href="#smootha3dcube"><em>Smooth a 3D Cube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.00</td>
<td><a class="reference internal" href="#accumulateimages"><em>Accumulate Images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.01</td>
<td><a class="reference internal" href="#combine2ddarkimages"><em>Combine 2D dark images</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.01</td>
<td><a class="reference internal" href="#createhigh-resolutionmicrolenspsfmodel"><em>Create High-Resolution Microlens PSF Model</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.01</td>
<td><a class="reference internal" href="#findhotbadpixelsfromdarks"><em>Find Hot Bad Pixels from Darks</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.01</td>
<td><a class="reference internal" href="#createmicrophonicsnoisemodel"><em>Create microphonics noise model</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.01</td>
<td><a class="reference internal" href="#createsathermal-skybackgrounddatacube"><em>Creates a thermal/sky background datacube</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.01</td>
<td><a class="reference internal" href="#findcoldbadpixelsfromflats"><em>Find Cold Bad Pixels from Flats</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.02</td>
<td><a class="reference internal" href="#generatecombinedbadpixelmap"><em>Generate Combined Bad Pixel Map</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.05</td>
<td><a class="reference internal" href="#cleanpolarizationpairsviadoubledifference"><em>Clean Polarization Pairs via Double Difference</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.20</td>
<td><a class="reference internal" href="#combinewavelengthcalibrationslocations"><em>Combine Wavelength Calibrations locations</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.20</td>
<td><a class="reference internal" href="#populateflexureshiftsvselevationtable"><em>Populate Flexure Shifts vs Elevation Table</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.20</td>
<td><a class="reference internal" href="#combinewavelengthcalibrations"><em>Combine Wavelength Calibrations</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.40</td>
<td><a class="reference internal" href="#combinepolarizationsequenceviadoubledifference"><em>Combine Polarization Sequence via Double Difference</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.40</td>
<td><a class="reference internal" href="#combinepolarizationsequence"><em>Combine Polarization Sequence</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.50</td>
<td><a class="reference internal" href="#qualitycheckwavelengthcalibration"><em>Quality Check Wavelength Calibration</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>4.50</td>
<td><a class="reference internal" href="#combine3ddatacubes"><em>Combine 3D Datacubes</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>4.60</td>
<td><a class="reference internal" href="#padwavelengthcalibrationedges"><em>Pad Wavelength Calibration Edges</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-even"><td>10.00</td>
<td><a class="reference internal" href="#saveaccumulatedstack"><em>Save Accumulated Stack</em></a></td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>10.00</td>
<td><a class="reference internal" href="#saveoutput"><em>Save Output</em></a></td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="primitive-detailed-documentation">
<h1>Primitive Detailed Documentation<a class="headerlink" href="#primitive-detailed-documentation" title="Permalink to this headline">Â¶</a></h1>
<div class="section" id="display-raw-image-with-gpitv">
<span id="displayrawimagewithgpitv"></span><span id="index-0"></span><h2>Display raw image with GPItv<a class="headerlink" href="#display-raw-image-with-gpitv" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Display, with GPItv, raw data to be processed</div></blockquote>
<p><strong>Category</strong>:  ALL, HIDDEN      <strong>Order</strong>: 0.01</p>
<p><strong>Inputs</strong>:  A raw 2D file.</p>
<p><strong>Outputs</strong>:  No change to data</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Display in GPITV the current raw <span class="nb">image</span>, before any processing


KEYWORDS<span class="o">:</span>
       gpitv<span class="o">=</span>          session number <span class="k">for</span> the GPITV <span class="nb">window</span> to display in.
                               set to &#39;<span class="m">0</span>&#39; <span class="k">for</span> no display, <span class="k">or</span> <span class="o">&gt;=</span><span class="m">1</span> <span class="k">for</span> a display.




HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2007<span class="o">-</span>11
  2008<span class="o">-</span>04<span class="o">-</span>02 JM<span class="o">:</span> spatial summation <span class="nb">window</span> centered on pixel <span class="k">and</span> interpolation on the zem. comm. wav. <span class="nb">vector</span>
       2008<span class="o">-</span>06<span class="o">-</span>06 JM<span class="o">:</span> adapted to pipeline inputs
       2009<span class="o">-</span>04<span class="o">-</span>15 MDP<span class="o">:</span> Documentation updated
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="69%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>1</td>
<td>1-500: choose gpitv session for displaying output, 0 for no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_display_raw_image_with_gpitv.pro</p>
</div>
<div class="section" id="correct-for-interpixel-capacitance">
<span id="correctforinterpixelcapacitance"></span><span id="index-1"></span><h2>Correct for Interpixel Capacitance<a class="headerlink" href="#correct-for-interpixel-capacitance" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Correct image for interpixel capacitance using Fourier deconvolution.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 0.1</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>:  &#8216;ipccor&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="7%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
<tr class="row-even"><td>alpha</td>
<td>float</td>
<td>[0,1]</td>
<td>0.014</td>
<td>Fraction of charge in adjacent pixels along columns</td>
</tr>
<tr class="row-odd"><td>beta</td>
<td>float</td>
<td>[0,1]</td>
<td>0.014</td>
<td>Fraction of charge in adjacent pixels along rows</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_correct_interpixel_capacitance.pro</p>
</div>
<div class="section" id="flag-quicklook">
<span id="flagquicklook"></span><span id="index-2"></span><h2>Flag Quicklook<a class="headerlink" href="#flag-quicklook" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Flag a given reduction output as &#8216;quicklook&#8217; quality rather than science grade.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 0.1</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>:  The FITS file header in memory gets added a keyword QUIKLOOK=True</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Writes a QUIKLOOK<span class="o">=</span>True keyword to the current header.
       Also updates some FITS history <span class="nb">text</span> to indicate the quicklook status.



HISTORY<span class="o">:</span>
   Marshall Perrin 2013<span class="o">-</span>10<span class="o">-</span>29  Started based on gpi_add_missingkeyword
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="9%" />
<col width="10%" />
<col width="13%" />
<col width="58%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: Save output to disk, 0: Don&#8217;t save</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_flag_quicklook.pro</p>
</div>
<div class="section" id="load-wavelength-calibration">
<span id="loadwavelengthcalibration"></span><span id="index-3"></span><h2>Load Wavelength Calibration<a class="headerlink" href="#load-wavelength-calibration" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reads a wavelength calibration file from disk. This primitive is required for any data-cube extraction.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,Calibration      <strong>Order</strong>: 0.5</p>
<p><strong>Inputs</strong>:  none</p>
<p><strong>Outputs</strong>:  none; wavecal is loaded into memory</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       <span class="nb">Reads</span> a wavelength calibration file from disk.
       The wavelength calibration is stored using pointers into the <span class="k">common</span> block.



HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2008<span class="o">-</span>07
       Documentation updated <span class="o">-</span> Marshall Perrin, 2009<span class="o">-</span>04
  2009<span class="o">-</span>09<span class="o">-</span>02 JM<span class="o">:</span> hist added in header
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2010<span class="o">-</span>03<span class="o">-</span>15 JM<span class="o">:</span> added automatic detection
  2010<span class="o">-</span>08<span class="o">-</span>19 JM<span class="o">:</span> fixed bug which created new pointer everytime this primitive was called
  2010<span class="o">-</span>10<span class="o">-</span>19 JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2013<span class="o">-</span>03<span class="o">-</span>28 JM<span class="o">:</span> added manual shifts <span class="k">of</span> the wavecal
  2013<span class="o">-</span>04                 manual shifts code moved to new update_shifts_for_flexure
  2013<span class="o">-</span>07<span class="o">-</span>10 MP<span class="o">:</span> Documentation update <span class="k">and</span> code cleanup
  2013<span class="o">-</span>07<span class="o">-</span>16 MP<span class="o">:</span> Rename file <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>02 JM<span class="o">:</span> get ELEVATIO <span class="k">and</span> INPORT <span class="k">for</span> later flexure correction
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="7%" />
<col width="7%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired wavelength calibration file to be read</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_load_wavelength_calibration.pro</p>
</div>
<div class="section" id="load-high-res-psfs">
<span id="loadhigh-respsfs"></span><span id="index-4"></span><h2>Load High-Res PSFs<a class="headerlink" href="#load-high-res-psfs" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reads a high-res PSF file from disk. This primitive is required for PSF cube extraction.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience,SpectralScience,Calibration      <strong>Order</strong>: 0.5</p>
<p><strong>Inputs</strong>:  none</p>
<p><strong>Outputs</strong>:  none; mlens psf is loaded into memory</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       <span class="nb">Reads</span> a high<span class="o">-</span>res psf file.
       The high<span class="o">-</span>res psf is stored using pointers into the <span class="k">common</span> block.



HISTORY<span class="o">:</span>
       Originally by Zachary Draper <span class="m">2</span><span class="o">-</span>28<span class="o">-</span>14
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="7%" />
<col width="7%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired wavelength calibration file to be read</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_load_highres_psfs.pro</p>
</div>
<div class="section" id="load-polarimetry-spot-calibration">
<span id="loadpolarimetryspotcalibration"></span><span id="index-5"></span><h2>Load Polarimetry Spot Calibration<a class="headerlink" href="#load-polarimetry-spot-calibration" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reads a pol spot calibration file from disk. This primitive is required for any polarimetry data-cube extraction.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience,Calibration      <strong>Order</strong>: 0.51</p>
<p><strong>Inputs</strong>:  Not used directly</p>
<p><strong>Outputs</strong>:  none; polarimetry spot cal file is loaded into memory</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  <span class="nb">Reads</span> a polarimetry spot calibration file from disk.
  The spot calibration is stored using pointers into the <span class="k">common</span> block.



HISTORY<span class="o">:</span>
  2013<span class="o">-</span>01<span class="o">-</span>28 MMB<span class="o">:</span> Adapted to pol extraction (based on readwavcal.<span class="k">pro</span>)
  2013<span class="o">-</span>02<span class="o">-</span>07 MP<span class="o">:</span>  Updated logging <span class="k">and</span> docs a little bit.
                  Added efficiently <span class="k">not</span> reloading the same file multiple times.
  2013<span class="o">-</span>06<span class="o">-</span>04 JBR<span class="o">:</span> shifts <span class="k">for</span> flexure code is now moved to
                  update_shifts_for_flexure.<span class="k">pro</span> <span class="k">and</span> commented out here.
  2013<span class="o">-</span>07<span class="o">-</span>10 MP<span class="o">:</span>  Documentation update <span class="k">and</span> code cleanup.
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span>  Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span>  CalibrationFile argument syntax update.
  2014<span class="o">-</span>03<span class="o">-</span>21 MP<span class="o">:</span>  Remove &#39;efficient&#39; code <span class="k">for</span> avoiding reloading, since
                                       this doesn&#39;t play well with flexure updates that <span class="nb">shift</span>
                                       the calibrations all around.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="7%" />
<col width="7%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired wavelength calibration file to be read</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_load_polarimetry_spot_calibration.pro</p>
</div>
<div class="section" id="load-instrumental-polarization-calibration">
<span id="loadinstrumentalpolarizationcalibration"></span><span id="index-6"></span><h2>Load Instrumental Polarization Calibration<a class="headerlink" href="#load-instrumental-polarization-calibration" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Load a calibration file for the instrumental polarization.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience,Calibration      <strong>Order</strong>: 0.52</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>:   Instrumental polarization calibration is loaded into memory</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>HISTORY<span class="o">:</span>
       2010<span class="o">-</span>05<span class="o">-</span>22 MDP<span class="o">:</span> started
  2010<span class="o">-</span>10<span class="o">-</span>19 JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2011<span class="o">-</span>07<span class="o">-</span>30 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
  2013<span class="o">-</span>07<span class="o">-</span>16 MP<span class="o">:</span> Renamed <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="6%" />
<col width="10%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired instrumental polarization file to be read</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_load_instrumental_polarization_calibration.pro</p>
</div>
<div class="section" id="check-data-quality">
<span id="checkdataquality"></span><span id="index-7"></span><h2>Check Data Quality<a class="headerlink" href="#check-data-quality" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Check quality of data based on header keywords. For bad data, can fail the reduction or simply alert the user.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 0.9</p>
<p><strong>Inputs</strong>:  2D image file</p>
<p><strong>Outputs</strong>:  No change in data; reduction either continues or is terminated.</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This routine looks at various FITS header keywords to
  assess whether the data should be considered usable <span class="k">or</span> <span class="k">not</span>.

  The keywords checked include GPIHEALT, AVGRNOT, RMSERR.
  You can set the acceptable minimum data quality with the r0 <span class="k">and</span> rmserr
  parameters to this primitive.

  <span class="k">If</span> a file <span class="k">of</span> unacceptable quality is detected, the action taken will
  be determined by the &quot;action&quot; parameter<span class="o">:</span>
    <span class="m">0</span>. Alert the user with a <span class="nb">message</span> printed to screen
       but allow reduction to <span class="k">continue</span>
    <span class="m">1</span>. Halt the reduction <span class="k">and</span> fail the receipe.

 TODO<span class="o">:</span> implement pop<span class="o">-</span>up <span class="nb">window</span> <span class="k">for</span> alerts rather than just
       printing a <span class="nb">message</span> on screen


GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>AVRGNOT,GPIHEALT,RMSERR


HISTORY<span class="o">:</span>
  JM 2010<span class="o">-</span>10 <span class="o">:</span> created
  MP 2013<span class="o">-</span>01 <span class="o">:</span> Docs updated
  2013<span class="o">-</span>07<span class="o">-</span>16 MP<span class="o">:</span> Documentation cleanup. Rename &#39;control_data_quality&#39; <span class="o">-&gt;</span> &#39;check_data_quality&#39;
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="8%" />
<col width="11%" />
<col width="10%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Action</td>
<td>int</td>
<td>[0,10]</td>
<td>1</td>
<td>0:Simple alert and continue reduction, 1:Reduction fails</td>
</tr>
<tr class="row-odd"><td>r0</td>
<td>float</td>
<td>[0,2]</td>
<td>0.08</td>
<td>critical r0 [m] at lambda=0.5microns</td>
</tr>
<tr class="row-even"><td>rmserr</td>
<td>float</td>
<td>[0,1000]</td>
<td><ol class="first last arabic simple" start="10">
<li></li>
</ol>
</td>
<td>Critical rms wavefront error in microns.</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_check_data_quality.pro</p>
</div>
<div class="section" id="subtract-dark-background">
<span id="subtractdarkbackground"></span><span id="index-8"></span><h2>Subtract Dark Background<a class="headerlink" href="#subtract-dark-background" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Subtract a dark frame.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 1.1</p>
<p><strong>Inputs</strong>:  raw 2D image file</p>
<p><strong>Outputs</strong>:  2D image corrected for dark current      <strong>Output Suffix</strong>:  &#8216;darksub&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>        Subtract background from an <span class="nb">image</span> using a dark file.

        <span class="k">If</span> CalibrationFile<span class="o">=</span>AUTOMATIC, the best available dark is
        obtained from the calibration database.
   &quot;Best dark&quot; generally means a dark file that has the most similar
   integration time <span class="k">and</span> is closest in date &amp; time <span class="k">of</span> observation
   to the data in question.

   Specifically, in the Calibration Database code <span class="k">for</span> darks,
   the algorithm first looks <span class="k">for</span> dark files which are between
   <span class="m">0</span>.<span class="m">3</span> <span class="k">and</span> 3x <span class="k">of</span> the desired integration time. It takes all such
   darks which are on the closest date <span class="k">of</span> observation to the
   science data, <span class="k">and</span> from those finds the one that is closest in
   integration time to the science data.

   This dark is <span class="nb">read</span> in, rescaled by the appropriate ratio <span class="k">of</span>
   integration times, <span class="k">and</span> <span class="k">then</span> subtracted from the data.



        Empirically, rescaling darks by too large a factor does <span class="k">not</span>
        result in very high quality subtractions, due to various nonlinear
        behaviors such as saturation <span class="k">of</span> hot pixels <span class="k">and</span> the so<span class="o">-</span>called
        &#39;reset anomaly&#39; effect which biases the readout background level.
        Hence we impose a limit <span class="k">for</span> scaling the dark integration time
        up <span class="k">or</span> down, semi<span class="o">-</span>arbitrarily chosen to be 3x because it seems to
        work reasonably well.  The standard set <span class="k">of</span> darks planned to be
        taken routinely at Gemini should ensure that there are always available
        darks within this range.

        <span class="k">If</span> you desire different behavior, simply set the CalibrationFile manually
        <span class="k">of</span> course.


        Note<span class="o">:</span> <span class="k">If</span> the RequireExactMatch setting is <span class="m">1</span>, <span class="k">then</span> only dark files
               exactly matching in integration time will be used. <span class="k">If</span> there is no
               such file, the data is returned without any subtraction.





HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2008<span class="o">-</span>06
       2009<span class="o">-</span>04<span class="o">-</span>20 MDP<span class="o">:</span> Updated to pipeline format, added docs.
                                   Some code lifted from OSIRIS subtradark_000.<span class="k">pro</span>
  2009<span class="o">-</span>09<span class="o">-</span>02 JM<span class="o">:</span> hist added in header
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2010<span class="o">-</span>10<span class="o">-</span>19 JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2012<span class="o">-</span>07<span class="o">-</span>20 MP<span class="o">:</span> added DRPDARK keyword
  2012<span class="o">-</span>12<span class="o">-</span>13 MP<span class="o">:</span> Remove &quot;Sky&quot; from primitve discription since it&#39;s inaccurate
  2013<span class="o">-</span>07<span class="o">-</span>11 MP<span class="o">:</span> rename &#39;applydarkcorrection&#39; <span class="o">-&gt;</span> &#39;subtract_dark_background&#39; <span class="k">for</span> consistency
       2013<span class="o">-</span>10<span class="o">-</span>03 MP<span class="o">:</span> Add RequireExactMatch option, enable scaling <span class="k">for</span> non<span class="o">-</span>matching exptimes
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
  2014<span class="o">-</span>03<span class="o">-</span>22 MP<span class="o">:</span> Adding experimental interpolation option.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="5%" />
<col width="5%" />
<col width="7%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Name of dark file to subtract</td>
</tr>
<tr class="row-odd"><td>RequireExactMatch</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Must dark calibration file exactly match in integration time, or is scaling from a different exposure time allowed?</td>
</tr>
<tr class="row-even"><td>Interpolate</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Interpolate based on JD between prior and subsequent available darks</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_subtract_dark_background.pro</p>
</div>
<div class="section" id="remove-persistence-from-previous-images">
<span id="removepersistencefrompreviousimages"></span><span id="index-9"></span><h2>Remove Persistence from Previous Images<a class="headerlink" href="#remove-persistence-from-previous-images" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Determines/Removes persistence of previous images</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 1.2</p>
<p><strong>Inputs</strong>:  Raw or destriped 2D image</p>
<p><strong>Outputs</strong>:  2D image corrected for persistence of previous non-saturated images      <strong>Output Suffix</strong>:  &#8216;-nopersis&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  The removal <span class="k">of</span> persistence from previous non<span class="o">-</span>saturated images
  incorporates a model developed <span class="k">for</span> Hubble Space Telescopes Wide
  Field Camera <span class="m">3</span> (WFC3,
  www.stsci.edu<span class="o">/</span>hst<span class="o">/</span>wfc3<span class="o">/</span>ins_performance<span class="o">/</span>persistence).
  Persistence is proportional to the intensity <span class="k">of</span> the illuminating
  source, <span class="k">and</span> is observed to fade exponentially with time. The
  parameters <span class="k">of</span> the mathematical model <span class="k">for</span> the persistence, found
  in the pipeline&#39;s configuration directory were determined
  during integration <span class="k">and</span> test at UCSC.

  This primitive searches <span class="k">for</span> all files in the raw data directory
  taken within 600 seconds (10 <span class="nb">min</span>) <span class="k">of</span> the beginning <span class="k">of</span> the exposure
  <span class="k">of</span> interest. It <span class="k">then</span> calculates the persistence from each <span class="nb">image</span>,
  using the maximum <span class="k">of</span> the stack, <span class="k">and</span> subtracts it from the
  frame. Note that <span class="k">if</span> the detector is exposed to light, but no
  exposures are being taken, persistence will still build up on the
  detector that cannot be subtracted.

  Ideally, this program should be run after the destriping algorithm
  as readnoise does <span class="k">not</span> induce persistence. However, due to limitation
  that a pipeline primitive cannot call another primitive, this has
  <span class="k">not</span> been implemented. Future developement will involve moving the
  destriping algorithm into a idl <span class="k">function</span>, <span class="k">and</span> <span class="k">then</span> calling the
  <span class="k">function</span> from the destriping primitive. This will enable the ability
  <span class="k">for</span> this primitive to destripe the previous images. The user should
  note that the destriping is at a level that is low enough to <span class="k">not</span>
  leave a significant persistence, so this detail will <span class="k">not</span>
  significantly affect science data.

  At this time, the persistence is removed at the <span class="o">~</span>75% level due to
  inaccuracies in the model caused by an insufficient time sampling <span class="k">of</span>
  the initial falloff <span class="k">and</span> readnoise. A new dataset will be taken prior to shipping,
  <span class="k">and</span> new model parameters will be derived prior to commissioning.

       WARNING<span class="o">:</span> Persistence removal does <span class="k">not</span> (yet) work with COADDED images!

       The manual_UTEND keyword allows the user to manually set the UTEND keyword in the <span class="nb">image</span> that is taken in closest time to the <span class="nb">image</span> being reduced. This is important because sometimes when changing modes <span class="k">or</span> exposure times quickly, the CAL <span class="nb">exit</span> shutter can remain <span class="nb">open</span> so light will <span class="k">continue</span> hitting the detector past the UTEND time. Users should note that this is generally occurs only <span class="k">for</span> polarimetry snapshots after <span class="nb">long</span> spectral sequences (taken in the same band).


Requires the persistence_model_parameters.fits calibration file.



HISTORY<span class="o">:</span>

  Wed May 22 15<span class="o">:</span>11<span class="o">:</span>10 2013, LAB <span class="o">&lt;</span>LAB@localhost.localdomain<span class="o">&gt;</span>
  2013<span class="o">-</span>05<span class="o">-</span>14 PI<span class="o">:</span> Started
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="6%" />
<col width="7%" />
<col width="8%" />
<col width="66%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the persistence_parameter file to be read</td>
</tr>
<tr class="row-odd"><td>manual_dt</td>
<td>float</td>
<td>[0,600]</td>
<td>0</td>
<td>Manual input for time (in seconds) since last persisting file - see help for details</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_remove_persistence_from_previous_images.pro</p>
</div>
<div class="section" id="subtract-thermal-sky-background-if-k-band">
<span id="subtractthermal-skybackgroundifkband"></span><span id="index-10"></span><h2>Subtract Thermal/Sky Background if K band<a class="headerlink" href="#subtract-thermal-sky-background-if-k-band" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Subtract a dark frame.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 1.2</p>
<p><strong>Inputs</strong>:  2D image file</p>
<p><strong>Outputs</strong>:  2D image file, unchanged if YJH, background subtracted if K1 or K2.      <strong>Output Suffix</strong>:  &#8216;bkgndsub&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> Subtract thermal background emission, <span class="k">for</span> K band data only

       <span class="o">**</span> special note<span class="o">:</span> <span class="o">**</span>

       This is a new kind <span class="k">of</span> &quot;data dependent optional primitive&quot;. <span class="k">If</span> the filter <span class="k">of</span>
       the current data is YJH, <span class="nb">return</span> without doing <span class="o">*</span>anything<span class="o">*</span>, even logging the
       start<span class="o">/</span><span class="k">end</span> <span class="k">of</span> this primitive.  It becomes a complete no<span class="o">-</span>op <span class="k">for</span> non<span class="o">-</span>K<span class="o">-</span>band
       cases.

Algorithm<span class="o">:</span>

       Get the best available thermal background calibration file from CalDB
       Scale it to current exposure time
       Subtract it.
  The name <span class="k">of</span> the calibration file used is saved to the DRPBKGND header keyword.

ALGORITHM TODO<span class="o">:</span> Deal with uncertainty <span class="k">and</span> pixel mask frames too.





HISTORY<span class="o">:</span>
  2012<span class="o">-</span>12<span class="o">-</span>13 MP<span class="o">:</span> Initial implementation
  2013<span class="o">-</span>01<span class="o">-</span>16 MP<span class="o">:</span> Documentation cleanup.
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>15 MP<span class="o">:</span> Add override_scaling option, remove erroneous hard<span class="o">-</span>coded
                                       constant non<span class="o">-</span><span class="m">1</span> scaling.
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="5%" />
<col width="6%" />
<col width="7%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Name of thermal background file to subtract</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>Override_scaling</td>
<td>float</td>
<td>[0,10]</td>
<td>1.0</td>
<td>Set to value other than 1 to manually adjust the background image flux scaling to better match the science data</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_subtract_thermal_sky_background_if_k_band.pro</p>
</div>
<div class="section" id="clean-cosmic-rays">
<span id="cleancosmicrays"></span><span id="index-11"></span><h2>Clean Cosmic Rays<a class="headerlink" href="#clean-cosmic-rays" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Placeholder for cosmic ray rejection (if needed; not currently implemented!)</div></blockquote>
<p><strong>Category</strong>:  HIDDEN      <strong>Order</strong>: 1.23</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>: Not specified</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  Placeholder; does <span class="k">not</span> actually <span class="k">do</span> anything yet.
  Empirically, cosmic rays <span class="k">do</span> <span class="k">not</span> appear to be a significant noise source
  <span class="k">for</span> the GPI IFS. It&#39;s a substrate<span class="o">-</span>removed H2RG so the level is quite low.
  Furthermore, realtime identification <span class="k">and</span> removal <span class="k">of</span> CRs is included as
  part <span class="k">of</span> the up<span class="o">-</span>the<span class="o">-</span>ramp readout <span class="k">and</span> slope fitting, which handles the
  majority <span class="k">of</span> CRs.


  There are still occasional noticeable residual CRs, particularly in <span class="nb">long</span>
  duration exposures <span class="k">or</span> darks, but they&#39;ve <span class="k">not</span> yet proven annoying enough to
  implement an algorithm here...



HISTORY<span class="o">:</span>
2010<span class="o">-</span>01<span class="o">-</span>28 MDP<span class="o">:</span> Created Templae.
2011<span class="o">-</span>07<span class="o">-</span>30 MDP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
2013<span class="o">-</span>07<span class="o">-</span>16 MDP<span class="o">:</span> Renamed as part <span class="k">of</span> code cleanup.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_clean_cosmic_rays.pro</p>
</div>
<div class="section" id="apply-reference-pixel-correction">
<span id="applyreferencepixelcorrection"></span><span id="index-12"></span><h2>Apply Reference Pixel Correction<a class="headerlink" href="#apply-reference-pixel-correction" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Subtract channel bias levels and bias drift stripes using H2RG reference pixels.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 1.25</p>
<p><strong>Inputs</strong>:  2D image file</p>
<p><strong>Outputs</strong>:  2D image corrected for background using reference pixels      <strong>Output Suffix</strong>:  &#8216;refpixcorr&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Correct <span class="k">for</span> fluctuations in the bias<span class="o">/</span>dark level using the rows <span class="k">of</span>
       reference pixels in the H2RG detectors.

  Note that <span class="o">*</span>vertical<span class="o">*</span> reference pixel subtraction to <span class="nb">fix</span> offsets between
  the 32 readout channels is done in real time during the readout process by
  the IFS Detector Server software. The Detector Server does <span class="k">not</span> currently
  apply any horizontal reference pixel subtraction, so we need to <span class="k">do</span> that in
  the pipeline. See the HRPSTYPE <span class="k">and</span> VRPSTYPE FITS keywords in the SCI
  extension headers.

       Also note that <span class="k">if</span> you use one <span class="k">of</span> the specialized Destriping primitives,
       you <span class="k">do</span> <span class="k">not</span> also need to use this one as well.


  Algorithm choices include<span class="o">:</span>
   <span class="m">1</span>) simple_channels          in this <span class="k">case</span>, just use the <span class="nb">median</span> <span class="k">of</span> each
                                           vertical channel to remove offsets between
                                           the channels. (deprecated, now done by the IFS
                                           detector server in real time during readout)
   <span class="m">2</span>) simple_horizontal        take the <span class="nb">median</span> <span class="k">of</span> the <span class="m">8</span> ref pix <span class="k">for</span> each row,
                                               <span class="k">and</span> subtract that from each row.
   <span class="m">3</span>) smoothed_horizontal      Like the above, but smoothed by N pixels vertically
                                                       <span class="k">for</span> better S<span class="o">/</span>N. N is adjustable using the smoothing_size
                                                       parameter. Empirically values <span class="o">&lt;</span> 20 <span class="k">or</span> 30 seem to be
                                                       <span class="k">not</span> enough smoothing, so the <span class="nb">read</span> noise fluctuations
                                                       give spurious biases to the ref pix model.
   <span class="m">3</span>) interpolated             In this <span class="k">case</span>, use James Larkin&#39;s interpolation
                                               algorithm to remove linear variation with time
                                               in the horizontal direction. This gives the highest
                                               spatial frequency correction but is more affected
                                               by <span class="nb">read</span> noise.

       See discussion in section <span class="m">3</span>.<span class="m">1</span> <span class="k">of</span> Rauscher et al. 2008 Prof SPIE 7021 p 63.





ALGORITHM TODO<span class="o">:</span> Deal with uncertainty <span class="k">and</span> pixel mask frames too.


HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2008<span class="o">-</span>06
       2009<span class="o">-</span>04<span class="o">-</span>20 MDP<span class="o">:</span> Updated to pipeline format, added docs.
                                   Some code lifted from OSIRIS subtradark_000.<span class="k">pro</span>
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2012<span class="o">-</span>07<span class="o">-</span>27 MP<span class="o">:</span> Added Method parameter, James Larkin&#39;s improved algorithm
  2012<span class="o">-</span>10<span class="o">-</span>14 MP<span class="o">:</span> debugging <span class="k">and</span> code cleanup.
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>03 MP<span class="o">:</span> Some docs updates <span class="k">and</span> added SMOOTHED_HORIZONTAL algorithm <span class="k">and</span> smoothing_size parameter
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="3%" />
<col width="39%" />
<col width="8%" />
<col width="39%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
<tr class="row-even"><td>smoothing_size</td>
<td>int</td>
<td>[0,500]</td>
<td>31</td>
<td>Smoothing kernel size for smoothed_horizontal method.</td>
</tr>
<tr class="row-odd"><td>before_and_after</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Show the before-and-after images for the user to see?</td>
</tr>
<tr class="row-even"><td>Method</td>
<td>enum</td>
<td>SIMPLE_CHANNELS|SIMPLE_HORIZONTAL|SMOOTHED_HORIZONTAL|INTERPOLATED</td>
<td>INTERPOLATED</td>
<td>Algorithm for reference pixel subtraction.</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_apply_reference_pixel_correction.pro</p>
</div>
<div class="section" id="update-spot-shifts-for-flexure">
<span id="updatespotshiftsforflexure"></span><span id="index-13"></span><h2>Update Spot Shifts for Flexure<a class="headerlink" href="#update-spot-shifts-for-flexure" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Extract a 3D datacube from a 2D image. Spatial integration (3 pixels) along the dispersion axis</div></blockquote>
<p><strong>Category</strong>:  SpectralScience, Calibration, PolarimetricScience      <strong>Order</strong>: 1.34</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> This primitive updates the wavelength calibration <span class="k">and</span> spot location table
 to account <span class="k">for</span> shifts in the apparent position <span class="k">of</span> each spectrum due to
 elevation<span class="o">-</span>dependent flexure within the IFS.  The observed <span class="nb">image</span> motion is
 about <span class="m">0</span>.<span class="m">7</span> pixels in X <span class="k">and</span> <span class="m">0</span>.<span class="m">5</span> pixels in Y between <span class="m">0</span> <span class="k">and</span> 90 degrees

 By updating the X <span class="k">and</span> Y coordinates <span class="k">of</span> each lenslet across the field <span class="k">of</span> view,
 this primitive enables the extraction <span class="k">of</span> well behaved data cubes
 regardless <span class="k">of</span> the orientation.

 There are several options <span class="k">for</span> how to determine the shifts, set by the
 method keyword<span class="o">:</span>

   method<span class="o">=</span>&quot;None&quot;     No correction applied.
   method<span class="o">=</span>&#39;Manual&#39;   Apply shifts provided by the user via the
                     manual_dx <span class="k">and</span> manual_dy arguments.
   method<span class="o">=</span>&#39;Lookup&#39;   Correction applied based on a lookup table <span class="k">of</span> shifts
                     precomputed based on arc lamp data at multiple
                     orientations, obtained from the calibration
                     database.
   method<span class="o">=</span>&#39;BandShift&#39;Estimate the flexure values by comparing to the
                     most recent wavecal regardless <span class="k">of</span> the band <span class="k">and</span>
                     interpolating.
   method<span class="o">=</span>&#39;Auto&#39;     [work in progress, use at your own risk]
                     Attempt to determine the shifts on<span class="o">-</span>the<span class="o">-</span>fly from each
                     individual exposure via model fitting.

<span class="k">If</span> the &#39;gpitv&#39; argument to this primitive is used to send the output
<span class="nb">image</span> to a gpitv session, it will be displayed <span class="o">*</span>with the updated
wavelength calibration information overplotted<span class="o">*</span>.



HISTORY<span class="o">:</span>
  2013<span class="o">-</span>03<span class="o">-</span>08 MP<span class="o">:</span> Started based on extractcube, initial attempts at automated
                  on<span class="o">-</span>the<span class="o">-</span>fly measurements.
  2013<span class="o">-</span>03<span class="o">-</span>25 JM<span class="o">:</span> Implemented lookup table version.
  2013<span class="o">-</span>04<span class="o">-</span>22 PI<span class="o">:</span> A few bug fixes to lookup table code.
  2013<span class="o">-</span>04<span class="o">-</span>25 MP<span class="o">:</span> Documentation improvements.
  2013<span class="o">-</span>06<span class="o">-</span>04 JBR<span class="o">:</span> Now compatible with polarimetry.
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>02 JM<span class="o">:</span> new way <span class="k">of</span> dealing with the lookup table <span class="k">for</span> flexure effect correction, independent <span class="k">of</span> the reference wavelength solution used to calculate the shifts
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="5%" />
<col width="24%" />
<col width="6%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>method</td>
<td>string</td>
<td>[None|Manual|Lookup|BandShift|Auto]</td>
<td>None</td>
<td>How to correct spot shifts due to flexure? [None|Manual|Lookup|BandShift|Auto]</td>
</tr>
<tr class="row-odd"><td>manual_dx</td>
<td>float</td>
<td>[-10,10]</td>
<td>0</td>
<td>If method=Manual, the X shift of spectra at the center of the detector</td>
</tr>
<tr class="row-even"><td>manual_dy</td>
<td>float</td>
<td>[-10,10]</td>
<td>0</td>
<td>If method=Manual, the Y shift of spectra at the center of the detector</td>
</tr>
<tr class="row-odd"><td>Display</td>
<td>int</td>
<td>[-1,100]</td>
<td>-1</td>
<td>-1 = No display; 0 = New (unused) window; else = Window number to display diagnostic plot.</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_update_spot_shifts_for_flexure.pro</p>
</div>
<div class="section" id="flexure-2d-x-correlation-with-polcal">
<span id="flexure2dxcorrelationwithpolcal"></span><span id="index-14"></span><h2>Flexure 2D x correlation with polcal<a class="headerlink" href="#flexure-2d-x-correlation-with-polcal" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive uses the relevent pol cal file to generate a model detector image to cross correlate with a science image and find the flexure offset.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience      <strong>Order</strong>: 1.34</p>
<p><strong>Inputs</strong>:  Science image, polcal</p>
<p><strong>Outputs</strong>:  Flexure offset in xy detector coordinates.      <strong>Output Suffix</strong>: &#8216;&#8217;      ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This primitive uses the relevent microlense PSF <span class="k">and</span> pol cal to generate a model detector <span class="nb">image</span> to cross <span class="nb">correlate</span> with a science <span class="nb">image</span>.
  The resulting output can be used as a flexure offset prior to flux extraction.



  The resulting output can be used as a flexure offset prior to flux extraction.


<span class="nb">where</span> in the order <span class="k">of</span> the primitives should this go by default?

pick one <span class="k">of</span> the following options <span class="k">for</span> the primitive type<span class="o">:</span>

HISTORY<span class="o">:</span>
   Began 2014<span class="o">-</span>01<span class="o">-</span>13 by Zachary Draper
         2014<span class="o">-</span>09<span class="o">-</span>12 MMB<span class="o">:</span> Branched to a version that cross correlates with the polcal file rather than use the microlens
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="8%" />
<col width="11%" />
<col width="11%" />
<col width="55%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>range</td>
<td>float</td>
<td>[0,5]</td>
<td>0.3</td>
<td>Range of cross corrleation search in pixels.</td>
</tr>
<tr class="row-odd"><td>resolution</td>
<td>float</td>
<td>[0,1]</td>
<td>0.1</td>
<td>Subpixel resolution of cross correlation</td>
</tr>
<tr class="row-even"><td>psf_sep</td>
<td>float</td>
<td>[0,1]</td>
<td>0.1</td>
<td>PSF separation in pixels</td>
</tr>
<tr class="row-odd"><td>stopidl</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: stop IDL, 0: dont stop IDL</td>
</tr>
<tr class="row-even"><td>x_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>initial guess for large offsets</td>
</tr>
<tr class="row-odd"><td>y_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>initial guess for large offsets</td>
</tr>
<tr class="row-even"><td>badpix</td>
<td>float</td>
<td>[0,1]</td>
<td>1</td>
<td>Weight by bad pixel map?</td>
</tr>
<tr class="row-odd"><td>iterate</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Take the first result? Or iterate</td>
</tr>
<tr class="row-even"><td>max_iter</td>
<td>int</td>
<td>[1,100]</td>
<td>15</td>
<td>The maximum number of iterations</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_img_xcorr_polcal.pro</p>
</div>
<div class="section" id="destripe-for-darks-only">
<span id="destripefordarksonly"></span><span id="index-15"></span><h2>Destripe for Darks Only<a class="headerlink" href="#destripe-for-darks-only" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Subtract readout pickup noise using median across all channels. This is an aggressive destriping algorithm suitable only for use on images that have no light. Also includes microphonics noise removal.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 1.35</p>
<p><strong>Inputs</strong>:  A 2D dark image</p>
<p><strong>Outputs</strong>:  2D image corrected for stripe noise      <strong>Output Suffix</strong>:  &#8216;destripe&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Correct <span class="k">for</span> fluctuations in the background bias level
       (i.e. horizontal stripes in     the raw data) using a pixel<span class="o">-</span>by<span class="o">-</span>pixel
       <span class="nb">median</span> across all channels, taking into account the alternating readout
       directions <span class="k">for</span> every other channel.

       This provides a very high level <span class="k">of</span> rejection <span class="k">for</span> stripe noise, but <span class="k">of</span> course
       it assumes that there&#39;s no signal anywhere in your <span class="nb">image</span>. So it&#39;s only
       good <span class="k">for</span> darks.


  A second noise source that can be removed by this routine is the
  so<span class="o">-</span>called microphonics noise induced by high frequency vibrational modes <span class="k">of</span>
  the H2RG. This noise has a characteristic frequenct both temporally <span class="k">and</span>
  spatially, which lends itself to removal via Fourier filtering. After
  destriping, the <span class="nb">image</span> is Fourier transformed, masked to select only the
  Fourier frequencies <span class="k">of</span> interest, <span class="k">and</span> transformed back to yield a model <span class="k">for</span>
  the microphonics striping that can be subtracted from the data. Empirically
  this correction works quite well. Set the &quot;remove_microphonics&quot; option to
  enable this, <span class="k">and</span> set &quot;display&quot; to show on screen a
  diagnostic <span class="nb">plot</span> that lets you see the stripe &amp; microphonics removal in
  action.

SEE ALSO<span class="o">:</span> Destripe science frame




HISTORY<span class="o">:</span>
  2012<span class="o">-</span>10<span class="o">-</span>16 Patrick<span class="o">:</span> fixed syntax error (<span class="k">function</span> name)
  2012<span class="o">-</span>10<span class="o">-</span>13 MP<span class="o">:</span> Started
  2013<span class="o">-</span>01<span class="o">-</span>16 MP<span class="o">:</span> Documentation cleanup
  2012<span class="o">-</span>03<span class="o">-</span>13 MP<span class="o">:</span> Added Fourier filtering to remove microphonics noise
  2013<span class="o">-</span>04<span class="o">-</span>25 MP<span class="o">:</span> Improved documentation, display <span class="k">for</span> microphonics removal.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="6%" />
<col width="7%" />
<col width="6%" />
<col width="65%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>remove_microphonics</td>
<td>string</td>
<td>[yes|no]</td>
<td>yes</td>
<td>Attempt to remove microphonics noise via Fourier filtering?</td>
</tr>
<tr class="row-odd"><td>Display</td>
<td>int</td>
<td>[-1,100]</td>
<td>-1</td>
<td>-1 = No display; 0 = New (unused) window else = Window number to display diagonostics in.</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_destripe_for_darks_only.pro</p>
</div>
<div class="section" id="destripe-science-image">
<span id="destripescienceimage"></span><span id="index-16"></span><h2>Destripe science image<a class="headerlink" href="#destripe-science-image" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Subtract detector striping using measurements between the microspectra</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,Calibration, PolarimetricScience      <strong>Order</strong>: 1.35</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>This primitive was originally developed to remove striping <span class="k">and</span> microphonics
noise in IFS images. The noise level <span class="k">of</span> the detector has since decreased
significantly <span class="k">and</span> therefore this primitive is generally only useful <span class="k">for</span>
short exposures. Note that without proper examination, this primitive may
INTRODUCE a systematic noise into the <span class="nb">image</span>. Users should consult the
IFS handbook destriping section when using this primitive.


 This primitive subtracts horizontal striping from the background <span class="k">of</span> a 2d raw IFS <span class="nb">image</span>
 by masking spectra <span class="k">and</span> using the remaining regions to obtain a
 sampling <span class="k">of</span> the striping.

 The masking can be performed by using the wavelength calibration to mask the
 spectra (recommended) <span class="k">or</span> by thresholding (<span class="k">not</span> recommended).

 WARNING<span class="o">:</span> This destriping algorithm will <span class="k">not</span> work correctly on flat fields <span class="k">or</span>
 any <span class="nb">image</span> <span class="nb">where</span> there is very large amounts <span class="k">of</span> signal covering the entire
 field. <span class="k">If</span> called on such data, it will <span class="nb">print</span> a warning <span class="nb">message</span> <span class="k">and</span> <span class="nb">return</span>
 without modifying the data array.

 Summary <span class="k">of</span> the primitive<span class="o">:</span>
 The principle idea is to build models <span class="k">of</span> the different source <span class="k">of</span> noise
 you want to treat <span class="k">and</span> <span class="k">then</span> subtract them from the real <span class="nb">image</span> at the <span class="k">end</span>.
  <span class="m">1</span><span class="o">/</span> mask computation
  <span class="m">2</span><span class="o">/</span> Channels offset model based on im <span class="o">=</span> <span class="nb">image</span> <span class="o">=&gt;</span> chan_offset
  <span class="m">3</span><span class="o">/</span> Microphonics computation based on im <span class="o">=</span> <span class="nb">image</span> <span class="o">-</span> chan_offset <span class="o">=&gt;</span> microphonics_model
  <span class="m">4</span><span class="o">/</span> Destriping model based on im <span class="o">=</span> <span class="nb">image</span> <span class="o">-</span> chan_offset <span class="o">-</span> microphonics_model <span class="o">=&gt;</span> stripes
  <span class="m">5</span><span class="o">/</span> Output<span class="o">:</span> imout <span class="o">=</span> <span class="nb">image</span> <span class="o">-</span> chan_offset <span class="o">-</span> microphonics_model <span class="o">-</span> stripes

Destriping Algorithm Details<span class="o">:</span>
   Generate a mask <span class="k">of</span> <span class="nb">where</span> the spectra are located, based on the
     already<span class="o">-</span>loaded wavelength <span class="k">or</span> pol spots solutions.
   Mask out those pixels.
 <span class="k">Break</span> the <span class="nb">image</span> up into the 32 readout channels
 Flip the odd channels to account <span class="k">for</span> the alternating readout direction.
 Generate a <span class="nb">median</span> <span class="nb">image</span> across the 32 readout channels
 <span class="nb">Smooth</span> by 20 pixels to generate the broad variations
 mask out any pixels that are <span class="o">&gt;</span><span class="m">3</span> sigma discrepant vs the broad variations
 Generate a better <span class="nb">median</span> <span class="nb">image</span> across the 32 readout channels post masking
 Perform some sanity checks <span class="k">for</span> model validity <span class="k">and</span> <span class="nb">interpolate</span> NaNs as needed
 <span class="nb">Expand</span> to a 2D <span class="nb">image</span> model <span class="k">of</span> the detector


OPTIONAL<span class="o">/</span>EXPERIMENTAL<span class="o">:</span>
 The microphonics noise attenuation can be activitated by setting the parameter remove_microphonics to <span class="m">1</span> <span class="k">or</span> <span class="m">2</span>.
 The microphonics from the <span class="nb">image</span> can be saved in a file using the parameter save_microphonics.
 <span class="k">If</span> Plot_micro_peaks equal &#39;yes&#39;, <span class="k">then</span> it will <span class="nb">open</span> <span class="m">3</span> <span class="nb">plot</span> windows with the peaks aera <span class="k">of</span> the
 microphonics in Fourier space (Before microphonics subtraction, the
 microphonics to be removed <span class="k">and</span> the final result). Used <span class="k">for</span> debugging purposes.

 <span class="k">If</span> remove_microphonics <span class="o">=</span> <span class="m">1</span><span class="o">:</span>
   The algorithm is always applied.

 <span class="k">If</span> remove_microphonics <span class="o">=</span> <span class="m">2</span><span class="o">:</span>
   The algorithm is applied only <span class="k">of</span> the quantity <span class="k">of</span> noise is greater than the micro_threshold parameter.
   A default empirical value <span class="k">of</span> <span class="m">0</span>.01 has been set based on the experience <span class="k">of</span> the author <span class="k">of</span> the algorithm.
   The quantity <span class="k">of</span> microphonics noise is measured with the ratio <span class="k">of</span> the dot_product <span class="k">and</span> the <span class="nb">norm</span> <span class="k">of</span> the
   <span class="nb">image</span><span class="o">:</span> dot_product<span class="o">/</span><span class="nb">sqrt</span>(sum(<span class="nb">abs</span>(<span class="nb">fft</span>(<span class="nb">image</span>))<span class="o">^</span><span class="m">2</span>)).
   With dot_product <span class="o">=</span> sum(<span class="nb">abs</span>(<span class="nb">fft</span>(<span class="nb">image</span>))<span class="o">*</span><span class="nb">abs</span>(<span class="nb">fft</span>(noise_model))) which
   correspond to the projection <span class="k">of</span> the <span class="nb">image</span> on the microphonics noise model in the absolute Fourier space.

 There are <span class="m">3</span> implemented methods right now depending on the value <span class="k">of</span> the parameter method_microphonics.

 <span class="k">If</span> method_microphonics <span class="o">=</span> <span class="m">1</span><span class="o">:</span>
   The microphonics noise removal is based on a fixed precomputed model. This model is the
   normalized absolute value <span class="k">of</span> the Fourier coefficients.
   The filtering consist <span class="k">of</span> diminishing the intensity <span class="k">of</span> the frequencies corresponding to the
   noise in the <span class="nb">image</span> proportionaly to the dot <span class="nb">product</span> <span class="k">of</span> the <span class="nb">image</span> witht the noise model.
   The phase remains unchanged.
   The filtered coefficients in Fourier space become (<span class="m">1</span><span class="o">-</span>dot_product<span class="o">*</span>(Amplitude_noise_model<span class="o">/</span>Amplitude_image)).
   With dot_product <span class="o">=</span> sum(<span class="nb">abs</span>(<span class="nb">fft</span>(<span class="nb">image</span>))<span class="o">*</span><span class="nb">abs</span>(<span class="nb">fft</span>(noise_model))) which correspond to the projection <span class="k">of</span> the <span class="nb">image</span> on the microphonics noise model in the absolute Fourier space.

 <span class="k">If</span> method_microphonics <span class="o">=</span> <span class="m">2</span><span class="o">:</span>
   The frequencies around the <span class="m">3</span> identified peaks <span class="k">of</span> the microphonics noise in Fourier space are all set to zero.
   This algorithm is the best one <span class="k">of</span> you are sure that there is no data in this aera but it is probably better <span class="k">not</span> to use it...

 <span class="k">If</span> method_microphonics <span class="o">=</span> <span class="m">3</span><span class="o">:</span>
   A 2d gaussian is fitted <span class="k">for</span> each <span class="k">of</span> the three peaks <span class="k">of</span> the microphonics noise in Fourier space <span class="k">and</span> <span class="k">then</span> removed.
   Only the absolute value is considered <span class="k">and</span> the phase remains unchanged.
   This algorthim is <span class="k">not</span> as efficient as the two others but <span class="k">if</span> you don&#39;t have an accurate model, it can be better than nothing.

Currently, the readnoise <span class="nb">floor</span>, which is what is used to determine the pixel masking <span class="k">for</span> spectral mode, is set to <span class="m">8</span> electrons divided by the
square root <span class="k">of</span> hte number <span class="k">of</span> coadds. Note that <span class="k">for</span> K band (<span class="k">and</span> sometimes H) this often has to be adjusted. The channel
offset correction should also be used when this value is being adjusted. Note that <span class="k">if</span> too much <span class="k">of</span> the <span class="nb">image</span> is masked,
it will surpass the abort_fraction <span class="k">and</span> no destriping will occur. Using an abort_fraction <span class="k">of</span> <span class="m">0</span>.<span class="m">7</span> is the minimum
a user should use <span class="k">for</span> normal cases.




HISTORY<span class="o">:</span>
    Originally by Marshall Perrin, 2011<span class="o">-</span>07<span class="o">-</span>15
  2011<span class="o">-</span>07<span class="o">-</span>30 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
  2012<span class="o">-</span>12<span class="o">-</span>12 PI<span class="o">:</span> Moved from Subtract_2d_background.<span class="k">pro</span>
  2012<span class="o">-</span>12<span class="o">-</span>30 MMB<span class="o">:</span> Updated <span class="k">for</span> pol extraction. Included Cal file, inserted IDL version checking <span class="k">for</span> <span class="nb">smooth</span>() <span class="k">function</span>
  2013<span class="o">-</span>01<span class="o">-</span>16 MP<span class="o">:</span> Documentation cleanup.
  2013<span class="o">-</span>03<span class="o">-</span>12 MP<span class="o">:</span> Code cleanup, some speed enhancements by vectorization
  2013<span class="o">-</span>05<span class="o">-</span>28 JBR<span class="o">:</span> Primitive copy pasted from the destripe_mask_spectra.<span class="k">pro</span> primitive. Microphonics noise enhancement. Microphonics algorithm now applied before the destriping.
  2013<span class="o">-</span>12<span class="o">-</span>04 PI<span class="o">:</span> Removed high_limit<span class="o">-</span> now does masking based on readnoise levels
       2013<span class="o">-</span>12<span class="o">-</span>30 MP<span class="o">:</span> CalibrationFile argument syntax update.
       2014<span class="o">-</span>02<span class="o">-</span>25 MP<span class="o">:</span> flats in polarization mode are OK to destripe
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="3%" />
<col width="9%" />
<col width="5%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>method</td>
<td>string</td>
<td>[threshold|calfile]</td>
<td>calfile</td>
<td>Find background based on image value threshold cut, or calibration file spectra/spot locations?</td>
</tr>
<tr class="row-odd"><td>abort_fraction</td>
<td>float</td>
<td>[0.0,1.0]</td>
<td>0.9</td>
<td>Necessary fraction of pixels in mask to continue - set at 0.9 to ensure quicklook tool is robust</td>
</tr>
<tr class="row-even"><td>chan_offset_correction</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Tries to correct for channel bias offsets - useful when no dark is available</td>
</tr>
<tr class="row-odd"><td>readnoise_floor</td>
<td>float</td>
<td>[0.0,100]</td>
<td>0.0</td>
<td>Readnoise floor in ADU. 0 = default to 8 electrons per CDS image</td>
</tr>
<tr class="row-even"><td>Save_stripes</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Save the striping noise image subtracted from frame?</td>
</tr>
<tr class="row-odd"><td>Display</td>
<td>int</td>
<td>[-1,100]</td>
<td>-1</td>
<td>-1 = No display; 0 = New (unused) window else = Window number to display diagonostics in.</td>
</tr>
<tr class="row-even"><td>remove_microphonics</td>
<td>int</td>
<td>[0,2]</td>
<td>0</td>
<td>Remove microphonics noise based on a precomputed fixed model.0: not applied. 1: applied. 2: the algoritm is applied only if the measured noise is greater than micro_threshold</td>
</tr>
<tr class="row-odd"><td>method_microphonics</td>
<td>int</td>
<td>[1,3]</td>
<td>0</td>
<td>Method applied for microphonics 1: model projection. 2: all to zero 3: gaussian fit</td>
</tr>
<tr class="row-even"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired microphonics model file to be read</td>
</tr>
<tr class="row-odd"><td>Plot_micro_peaks</td>
<td>string</td>
<td>[yes|no]</td>
<td>no</td>
<td>Plot in 3d the peaks corresponding to the microphonics</td>
</tr>
<tr class="row-even"><td>save_microphonics</td>
<td>string</td>
<td>[yes|no]</td>
<td>no</td>
<td>If remove_microphonics = 1 or (auto and micro_threshold overpassed), save the removed microphonics</td>
</tr>
<tr class="row-odd"><td>micro_threshold</td>
<td>float</td>
<td>[0.0,1.0]</td>
<td>0.01</td>
<td>If remove_microphonics = 2, set the threshold. This value is sum(abs(fft(image))*abs(fft(noise_model)))/sqrt(sum(image^2))</td>
</tr>
<tr class="row-even"><td>write_mask</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>write signal mask to reduced directory?</td>
</tr>
<tr class="row-odd"><td>fraction</td>
<td>float</td>
<td>[0.0,1.0]</td>
<td>0.7</td>
<td>Threshold fraction of the total pixels in a row should be masked</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: Save output to disk, 0: Don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>1</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_destripe_science_image.pro</p>
</div>
<div class="section" id="interpolate-bad-pixels-in-2d-frame">
<span id="interpolatebadpixelsin2dframe"></span><span id="index-17"></span><h2>Interpolate bad pixels in 2D frame<a class="headerlink" href="#interpolate-bad-pixels-in-2d-frame" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Repair bad pixels by interpolating between their neighbors. Can optionally just flag as NaNs or else interpolate.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience, PolarimetricScience, Calibration      <strong>Order</strong>: 1.4</p>
<p><strong>Inputs</strong>:  2D image, ideally post dark subtraction and destriping</p>
<p><strong>Outputs</strong>:  2D image with bad pixels marked or cleaned up.      <strong>Output Suffix</strong>: &#8216;-bpfix&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Interpolates between vertical (spectral dispersion) direction neighboring
       pixels to <span class="nb">fix</span> each bad pixel.

  Bad pixels are identified from<span class="o">:</span>
  <span class="m">1</span>. The pixels marked bad in the current bad pixel mask (provided in the
     CalibrationFile parameter.)
  <span class="m">2</span>. Any additional pixels which are marked as bad in the <span class="nb">image</span> extension
     <span class="k">for</span> data quality (DQ).
  <span class="m">3</span>. Any pixels which are <span class="o">&lt;</span> <span class="o">-</span>50 counts (i.e. are <span class="o">&gt;</span> <span class="m">5</span> sigma negative <span class="nb">where</span>
     sigma is the CDS <span class="nb">read</span> noise <span class="k">for</span> a single <span class="nb">read</span>). TODO<span class="o">:</span> This threshhold
     should be evaluated <span class="k">and</span> possibly made adjustible.

 The action taken on those bad pixels is determined from the &#39;method&#39;
 parameter, which can be one <span class="k">of</span><span class="o">:</span>
   &#39;nan&#39;<span class="o">:</span>   Bad pixels are just marked as NaN, with no interpolation
   &#39;vertical&#39;<span class="o">:</span> Bad pixels are repaired by interpolating over their
            immediate neighbors vertically, the pixels above <span class="k">and</span> below.
            This has been shown to work well <span class="k">for</span> spectral mode GPI data
            since vertical is the spectral dispersion direction.
            (The actual algorithm is a bit more complicated than this to
                         handle cases <span class="nb">where</span> the above <span class="k">and</span><span class="o">/</span><span class="k">or</span> below pixels are themselves
                         also bad.)
   &#39;all8&#39;<span class="o">:</span>  Repair by interpolating over all <span class="m">8</span> surrounding pixels.



       TODO<span class="o">:</span> need to evaluate whether that algorithm is still a good approach <span class="k">for</span>
       polarimetry mode files.

       TODO<span class="o">:</span> implement Christian&#39;s suggestion <span class="k">of</span> a 3D interpolation in 2D space,
       using adjacent lenslet spectra as well. See emails <span class="k">of</span> Oct 18, 2012
       (excerpted below)






HISTORY<span class="o">:</span>
       Originally by Marshall Perrin, 2012<span class="o">-</span>10<span class="o">-</span>18
       2012<span class="o">-</span>12<span class="o">-</span>03 MP<span class="o">:</span> debugging<span class="o">/</span>enhancements <span class="k">for</span> the <span class="k">case</span> <span class="k">of</span> multiple adjacent bad
                                       pixels
       2012<span class="o">-</span>12<span class="o">-</span>09 MP<span class="o">:</span> Added support <span class="k">for</span> using information in DQ extension
       2013<span class="o">-</span>01<span class="o">-</span>16 MP<span class="o">:</span> Documentation cleanup
       2013<span class="o">-</span>02<span class="o">-</span>07 MP<span class="o">:</span> Enhanced all8 interpolation to properly handle cases <span class="nb">where</span>
                                       there are bad pixels in the neighboring pixels.
  2013<span class="o">-</span>04<span class="o">-</span>02 JBR<span class="o">:</span> Correction <span class="k">of</span> a sign in the vertical algorithm when reading the bottom adjacent pixel.
  2013<span class="o">-</span>04<span class="o">-</span>22 JBR<span class="o">:</span> In vertical algorithm, condition added <span class="k">if</span> both upper <span class="k">and</span> bottom pixels are good.
       2013<span class="o">-</span>06<span class="o">-</span>26 MP<span class="o">:</span> Added better FITS history logging <span class="k">for</span> the <span class="k">case</span> <span class="k">of</span> <span class="k">not</span> having a bad pixel <span class="nb">map</span>.
       2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> Rename file <span class="k">for</span> consistency
       2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> Update to allow bad pixel <span class="nb">map</span> files to have values other than
                                       <span class="m">1</span>, with any nonzero value being interpreted as bad.
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="5%" />
<col width="12%" />
<col width="6%" />
<col width="65%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>None</td>
<td>None</td>
<td>None</td>
<td>Filename of the desired bad pixel file to be read</td>
</tr>
<tr class="row-odd"><td>method</td>
<td>string</td>
<td>[n4n|vertical|all8]</td>
<td>vertical</td>
<td>Repair bad bix interpolating all 8 neighboring pixels, or just the 2 vertical ones, or just flag as NaN (n4n)?</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>1</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
<tr class="row-even"><td>negative_bad_thresh</td>
<td>float</td>
<td>[-100000,0]</td>
<td>-50</td>
<td>Pixels more negative than this should be considered bad.</td>
</tr>
<tr class="row-odd"><td>before_and_after</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Show the before-and-after images for the user to see? (for debugging/testing)</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_interpolate_bad_pixels_in_2d_frame.pro</p>
</div>
<div class="section" id="combine-2d-images">
<span id="combine2dimages"></span><span id="index-18"></span><h2>Combine 2D images<a class="headerlink" href="#combine-2d-images" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Combine 2D images such as darks into a master file via mean or median.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 1.5</p>
<p><strong>Inputs</strong>:  Multiple 2D images</p>
<p><strong>Outputs</strong>:  a single combined 2D image      <strong>Output Suffix</strong>:  strlowcase(method)</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> Multiple 2D images can be combined into one using either a <span class="nb">mean</span>,
 a sigma<span class="o">-</span>clipped <span class="nb">mean</span>,  <span class="k">or</span> a <span class="nb">median</span>.




HISTORY<span class="o">:</span>
        Jerome Maire 2008<span class="o">-</span>10
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>10<span class="o">-</span>22 MDP<span class="o">:</span> Created from mediancombine_darks, converted to use
                               accumulator.
  2010<span class="o">-</span>01<span class="o">-</span>25 MDP<span class="o">:</span> Added support <span class="k">for</span> multiple methods, <span class="nb">MEAN</span> method.
  2011<span class="o">-</span>07<span class="o">-</span>30 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
  2012<span class="o">-</span>10<span class="o">-</span>10 MP<span class="o">:</span> Minor code cleanup
  2013<span class="o">-</span>07<span class="o">-</span>10 MP<span class="o">:</span> Minor documentation cleanup
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> file rename <span class="k">for</span> consistency
  2014<span class="o">-</span>01<span class="o">-</span>02 MP<span class="o">:</span> Copied SIGMACLIP implementation from gpi_combine_2d_dark_images
  2014<span class="o">-</span>11<span class="o">-</span>04 MP<span class="o">:</span> Avoid trying to run parallelized sigmaclip <span class="k">if</span> in IDL runtime.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="4%" />
<col width="12%" />
<col width="6%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Method</td>
<td>string</td>
<td>MEAN|MEDIAN|SIGMACLIP</td>
<td>SIGMACLIP</td>
<td>How to combine images: median, mean, or mean with outlier rejection?[MEAN|MEDIAN|SIGMACLIP]</td>
</tr>
<tr class="row-odd"><td>Sigma_cut</td>
<td>float</td>
<td>[1,100]</td>
<td>3</td>
<td>If Method=SIGMACLIP, then data points more than this many standard deviations away from the median value of a given pixel will be discarded.</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_combine_2d_images.pro</p>
</div>
<div class="section" id="combine-2d-thermal-sky-backgrounds">
<span id="combine2dthermal-skybackgrounds"></span><span id="index-19"></span><h2>Combine 2D Thermal/Sky Backgrounds<a class="headerlink" href="#combine-2d-thermal-sky-backgrounds" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Combine 2D images with measurement of thermal or sky background</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 1.51</p>
<p><strong>Inputs</strong>:  2D image(s) taken with lamps off.</p>
<p><strong>Outputs</strong>:  thermal background file, saved as calibration file      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Generate a 2D background <span class="nb">image</span> <span class="k">for</span> use in removing e.g. thermal emission
       from lamp images




HISTORY<span class="o">:</span>
  2012<span class="o">-</span>12<span class="o">-</span>13 MP<span class="o">:</span> Forked from combine2dframes
  2013<span class="o">-</span>07<span class="o">-</span>10 MP<span class="o">:</span> Minor documentation cleanup
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
       2014<span class="o">-</span>01<span class="o">-</span>02 MP<span class="o">:</span> Copied SIGMACLIP implementation from gpi_combine_2d_dark_images
  2014<span class="o">-</span>11<span class="o">-</span>10 MP<span class="o">:</span> Avoid trying to run parallelized sigmaclip <span class="k">if</span> in IDL runtime.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="4%" />
<col width="12%" />
<col width="6%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Method</td>
<td>enum</td>
<td>MEAN|MEDIAN|SIGMACLIP</td>
<td>SIGMACLIP</td>
<td>How to combine images: median, mean, or mean with outlier rejection?[MEAN|MEDIAN|SIGMACLIP]</td>
</tr>
<tr class="row-odd"><td>Sigma_cut</td>
<td>float</td>
<td>[1,100]</td>
<td>3</td>
<td>If Method=SIGMACLIP, then data points more than this many standard deviations away from the median value of a given pixel will be discarded.</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_combine_2d_thermal_sky_backgrounds.pro</p>
</div>
<div class="section" id="d-wavelength-solution-developer">
<span id="dwavelengthsolutiondeveloper"></span><span id="index-20"></span><h2>2D Wavelength Solution Developer<a class="headerlink" href="#d-wavelength-solution-developer" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive uses an existing wavelength solution file to construct a new wavelength solution file by simulating the detector image and performing a least squares fit.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 1.7</p>
<p><strong>Inputs</strong>:  An Xe/Ar lamp detector image</p>
<p><strong>Outputs</strong>:  A wavelength solution cube (and a simulated Xe/Ar lamp detector image; to come)      <strong>Output Suffix</strong>: &#8216;wavecal&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This is the main wavelength calibration generation primitive.

  This Wavelength Solution generator models an arclamp spectrum
  <span class="k">for</span> each lenslet <span class="k">and</span> uses mpfit2dfunc to fit the relevant
  wavelength solution variables (ie. xo, yo, lambdao, dispersion,
  tilt). A wavelength solution file is output along with a
  simulated detector <span class="nb">image</span>.

       A previous wavelength calibration file is used to supply the
       initial guess <span class="k">for</span> the fitting process, which is <span class="k">then</span> updated
       by this primitive.

       This is fairly computationally intensive <span class="k">and</span> requires
       relatively high S<span class="o">/</span>N data. See Quick Wavelength Solution <span class="k">if</span>
       you need faster results (albeit more limited <span class="k">and</span> requiring you
       already have a reference wavecal)








HISTORY<span class="o">:</span>
   2013<span class="o">-</span>09<span class="o">-</span>19 SW<span class="o">:</span> <span class="m">2</span><span class="o">-</span>dimensionsal wavelength solution
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="5%" />
<col width="5%" />
<col width="6%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>display</td>
<td>Int</td>
<td>[0,1]</td>
<td>0</td>
<td>Whether or not to plot each lenslet spectrum model in comparison to the detector measured spectrum: 1;display, 0;no display</td>
</tr>
<tr class="row-odd"><td>whichpsf</td>
<td>Int</td>
<td>[0,1]</td>
<td>0</td>
<td>Type of lenslet PSF model, 0: gaussian, 1: microlens</td>
</tr>
<tr class="row-even"><td>parallel</td>
<td>Int</td>
<td>[0,1]</td>
<td>0</td>
<td>Option for Parallelization,  0: none, 1: parallel</td>
</tr>
<tr class="row-odd"><td>numsplit</td>
<td>Int</td>
<td>[0,100]</td>
<td>0</td>
<td>Number of cores for parallelization. Set to 0 for autoselect.</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>Save_model_image</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save 2d detector model fit image to disk, 0:don&#8217;t save</td>
</tr>
<tr class="row-even"><td>CalibrationFile</td>
<td>wavcal</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired reference wavelength calibration file to be read</td>
</tr>
<tr class="row-odd"><td>Save_model_params</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save model nuisance parameters to disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>AutoOffset</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Automatically determine x/yoffset values 0;NO, 1;YES</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_wavelength_solution_2d_developer.pro</p>
</div>
<div class="section" id="quick-wavelength-solution-update">
<span id="quickwavelengthsolutionupdate"></span><span id="index-21"></span><h2>Quick Wavelength Solution Update<a class="headerlink" href="#quick-wavelength-solution-update" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Given an existing wavecal and a new Xe lamp image, this primitive updates the wavecal based on the X,Y positions measured for a subset of the Xe spectra.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 1.7</p>
<p><strong>Inputs</strong>:  An Xe/Ar lamp detector image</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>: &#8216;wavecal&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This is a modified version <span class="k">of</span> the 2D wavelength solution
  algorithm, which fits a small subset <span class="k">of</span> lenslets (set by
  the &#39;spacing&#39; argument) to very quickly provide an estimated
  wavelength solution, based on some prior wavelength solution.

  This differs from the full wavelength solution in that<span class="o">:</span>

   <span class="m">1</span>) Only a subset <span class="k">of</span> lenslets are fit
   <span class="m">2</span>) The <span class="nb">mean</span> shifts in X <span class="k">and</span> Y are derived from those fits
   <span class="m">3</span>) The output wavelength solution is created by taking
      the input wavelength solution <span class="k">and</span> applying those shifts.
      (i.e. only the overall <span class="nb">shift</span> <span class="k">of</span> the wavecal is updated;
      the individual dispersions <span class="k">and</span> tilts <span class="k">of</span> each lenslet&#39;s
      spectrum are <span class="k">not</span> changed).

  This algorithm is both computationally faster than <span class="k">and</span>
  tolerant <span class="k">of</span> lower S<span class="o">/</span>N data than the full wavelength solution
  algorithm. This is because it is in essence only trying to measure
  <span class="m">2</span> parameters, the average shifts in X <span class="k">and</span> Y, rather than the
  <span class="o">~</span> 150,000 parameters measured <span class="k">and</span> saved <span class="k">for</span> the full wavelength
  calibration algorithm.


KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>FILTER,IFSFILT,GCALLAMP
DRP KEYWORDS<span class="o">:</span> FILETYPE,HISTORY,ISCALIB





HISTORY<span class="o">:</span>
       2013<span class="o">-</span>09<span class="o">-</span>19 SW<span class="o">:</span> <span class="m">2</span><span class="o">-</span>dimensionsal wavelength solution
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="4%" />
<col width="5%" />
<col width="6%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Display</td>
<td>int</td>
<td>[-1,100]</td>
<td>-1</td>
<td>-1 = No display; 0 = New (unused) window; else = Window number to display each lenslet in comparison to the detector lenslet in.</td>
</tr>
<tr class="row-odd"><td>spacing</td>
<td>Int</td>
<td>[0,20]</td>
<td>10</td>
<td>Test every Nth lenslet for this value of N.</td>
</tr>
<tr class="row-even"><td>boxsizex</td>
<td>Int</td>
<td>[0,15]</td>
<td>7</td>
<td>x dimension of a lenslet cutout</td>
</tr>
<tr class="row-odd"><td>boxsizey</td>
<td>Int</td>
<td>[0,50]</td>
<td>24</td>
<td>y dimension of a lenslet cutout</td>
</tr>
<tr class="row-even"><td>xoffset</td>
<td>Int</td>
<td>[-10,10]</td>
<td>0</td>
<td>x offset guess from prior wavecal.</td>
</tr>
<tr class="row-odd"><td>yoffset</td>
<td>Int</td>
<td>[-20,20]</td>
<td>0</td>
<td>y offset guess from prior wavecal.</td>
</tr>
<tr class="row-even"><td>whichpsf</td>
<td>Int</td>
<td>[0,1]</td>
<td>0</td>
<td>Type of psf 0;gaussian, 1;microlens</td>
</tr>
<tr class="row-odd"><td>high_order_correction</td>
<td>Int</td>
<td>[0,1]</td>
<td>1</td>
<td>Higher order flexure offsets? 0; No, 1; Yes</td>
</tr>
<tr class="row-even"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired wavelength calibration file to be read</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>AutoOffset</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Automatically determine x/yoffset values 0;NO, 1;YES</td>
</tr>
<tr class="row-odd"><td>gpitvim_dispgrid</td>
<td>int</td>
<td>[0,500]</td>
<td>15</td>
<td>1-500: choose gpitv session for displaying image output and wavcal grid overplotted, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_quick_wavelength_solution_update.pro</p>
</div>
<div class="section" id="d-wavelength-solution">
<span id="dwavelengthsolution"></span><span id="index-22"></span><h2>2D Wavelength Solution<a class="headerlink" href="#d-wavelength-solution" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive uses an existing wavelength solution file to construct a new wavelength solution file by simulating the detector image and performing a least squares fit.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 1.7</p>
<p><strong>Inputs</strong>:  An Xe/Ar lamp detector image</p>
<p><strong>Outputs</strong>:  A wavelength solution cube (and a simulated Xe/Ar lamp detector image; to come)      <strong>Output Suffix</strong>: &#8216;wavecal&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This is the main wavelength calibration generation primitive.

  This Wavelength Solution generator models an arclamp spectrum
  <span class="k">for</span> each lenslet <span class="k">and</span> uses mpfit2dfunc to fit the relevant
  wavelength solution variables (ie. xo, yo, lambdao, dispersion,
  tilt). A wavelength solution file is output along with a
  simulated detector <span class="nb">image</span>.

       A previous wavelength calibration file is used to supply the
       initial guess <span class="k">for</span> the fitting process, which is <span class="k">then</span> updated
       by this primitive.

       This is fairly computationally intensive <span class="k">and</span> requires
       relatively high S<span class="o">/</span>N data. See Quick Wavelength Solution <span class="k">if</span>
       you need faster results (albeit more limited <span class="k">and</span> requiring you
       already have a reference wavecal)








HISTORY<span class="o">:</span>
   2013<span class="o">-</span>09<span class="o">-</span>19 SW<span class="o">:</span> <span class="m">2</span><span class="o">-</span>dimensionsal wavelength solution
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="5%" />
<col width="5%" />
<col width="6%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>display</td>
<td>Int</td>
<td>[0,1]</td>
<td>0</td>
<td>Whether or not to plot each lenslet spectrum model in comparison to the detector measured spectrum: 1;display, 0;no display</td>
</tr>
<tr class="row-odd"><td>parallel</td>
<td>Int</td>
<td>[0,1]</td>
<td>0</td>
<td>Option for Parallelization,  0: none, 1: parallel</td>
</tr>
<tr class="row-even"><td>numsplit</td>
<td>Int</td>
<td>[0,100]</td>
<td>0</td>
<td>Number of cores for parallelization. Set to 0 for autoselect.</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>Save_model_image</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save 2d detector model fit image to disk, 0:don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>CalibrationFile</td>
<td>wavcal</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired reference wavelength calibration file to be read</td>
</tr>
<tr class="row-even"><td>Save_model_params</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save model nuisance parameters to disk, 0: don&#8217;t save</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_wavelength_solution_2d.pro</p>
</div>
<div class="section" id="measure-wavelength-calibration">
<span id="measurewavelengthcalibration"></span><span id="index-23"></span><h2>Measure Wavelength Calibration<a class="headerlink" href="#measure-wavelength-calibration" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Derive wavelength calibration from an arc lamp or flat-field image.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 1.7</p>
<p><strong>Inputs</strong>:  2D image from narrow band arclamp</p>
<p><strong>Outputs</strong>:       <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This primitive positions <span class="k">of</span> spectra in the <span class="nb">image</span> with narrow
       band lamp <span class="nb">image</span>.

       <span class="o">**</span> DEPRECATED<span class="o">**</span> This is the older &#39;first generation&#39; wavelength
       calibration algorith, which is no longer recommended

ALGORITHM<span class="o">:</span>
       gpi_extract_wavcal starts by detecting the central peak <span class="k">of</span> the <span class="nb">image</span>.
       Next, starting with a initial value <span class="k">of</span> w &amp; P, find the nearest peak (with an increment on the microlens coordinates)
       when nearest peak has been detected, it reevaluates w &amp; P <span class="k">and</span> so forth..

<span class="o">*********************************************************************************</span>
<span class="o">*</span>
<span class="o">*</span>  IMPORTANT WARNING <span class="k">for</span> future software maintainers<span class="o">:</span>
<span class="o">*</span>     The complicated algorithms implemented here were originally developed
<span class="o">*</span>     assuming the dispersion direction in GPI would be horizontal. Given data
<span class="o">*</span>     orientation conventions later adopted, it became vertical. Rather than
<span class="o">*</span>     rewriting all <span class="k">of</span> the following <span class="k">and</span> swapping all the indices around,
<span class="o">*</span>     the images are just <span class="o">*</span>transposed<span class="o">*</span> as the first step <span class="k">of</span> this process, <span class="k">and</span>
<span class="o">*</span>     <span class="k">then</span> the original horizontal algorithm applied. This leads to various
<span class="o">*</span>     complexities about index transformations. Be wary when editing the
<span class="o">*</span>     code here <span class="k">and</span> keep that in mind....
<span class="o">*</span>
<span class="o">*</span>
<span class="o">*********************************************************************************</span>



<span class="k">common</span> needed<span class="o">:</span>

KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>FILTER,IFSFILT,GCALLAMP,GCALSHUT,OBSTYPE
DRP KEYWORDS<span class="o">:</span> FILETYPE,HISTORY,ISCALIB


HISTORY<span class="o">:</span>
        Jerome Maire 2008<span class="o">-</span>10
         JM<span class="o">:</span> nlens, w (initial guess), P (initial guess), cenx (<span class="k">or</span> centrXpos), ceny (<span class="k">or</span> centrYpos) as parameters
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>12<span class="o">-</span>10 JM<span class="o">:</span> initiate position at <span class="m">1</span>.5microns so we can take into account several band
  2010<span class="o">-</span>07<span class="o">-</span>14 JM<span class="o">:</span><span class="k">for</span> DRP testing, correct <span class="k">for</span> DST <span class="nb">finite</span> spectral resolution
  2010<span class="o">-</span>08<span class="o">-</span>16 JM<span class="o">:</span> added bad pixel <span class="nb">map</span>
  2011<span class="o">-</span>07<span class="o">-</span>14 MP<span class="o">:</span> Reworked FITS keyword handling to provide more informative
        error messages in <span class="k">case</span> <span class="k">of</span> missing <span class="k">or</span> invalid keywords.
  2011<span class="o">-</span>08<span class="o">-</span>02 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS.
  2012<span class="o">-</span>12<span class="o">-</span>13 MP<span class="o">:</span> Bad pixel <span class="nb">map</span> now taken from DQ extension <span class="k">if</span> present.
                                  <span class="nb">Print</span> more informative logging messages <span class="k">for</span> the user
                                  Various bits <span class="k">of</span> code cleanup.
  2012<span class="o">-</span>12<span class="o">-</span>20 JM<span class="o">:</span> more centroid methods added
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="5%" />
<col width="9%" />
<col width="7%" />
<col width="66%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>nlens</td>
<td>int</td>
<td>[0,400]</td>
<td>281</td>
<td>side length of  the  lenslet array</td>
</tr>
<tr class="row-odd"><td>centrXpos</td>
<td>int</td>
<td>[0,2048]</td>
<td>1024</td>
<td>Initial approximate x-position [pixel] of central peak at 1.5microns</td>
</tr>
<tr class="row-even"><td>centrYpos</td>
<td>int</td>
<td>[0,2048]</td>
<td>1024</td>
<td>Initial approximate y-position [pixel] of central peak at 1.5microns</td>
</tr>
<tr class="row-odd"><td>w</td>
<td>float</td>
<td>[0.,10.]</td>
<td>4.8</td>
<td>Spectral spacing perpendicular to the dispersion axis at the image center [pixel]</td>
</tr>
<tr class="row-even"><td>P</td>
<td>float</td>
<td>[-7.,7.]</td>
<td>-1.8</td>
<td>Ratio of spectral offset parallel to dispersion over spectral spacing perpendicular to dispersion</td>
</tr>
<tr class="row-odd"><td>emissionlinesfile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>File of emission lines.</td>
</tr>
<tr class="row-even"><td>wav_of_centrXYpos</td>
<td>int</td>
<td>[1,2]</td>
<td>2</td>
<td>1 if centrX-Ypos is the smallest-wavelength peak of the band; 2 if centrX-Ypos refer to 1.5microns</td>
</tr>
<tr class="row-odd"><td>maxpos</td>
<td>float</td>
<td>[-7.,7.]</td>
<td><ol class="first last arabic simple" start="2">
<li></li>
</ol>
</td>
<td>Allowed maximum location fluctuation (in pixel) between adjacent mlens</td>
</tr>
<tr class="row-even"><td>maxtilt</td>
<td>float</td>
<td>[-360.,360.]</td>
<td><ol class="first last arabic simple" start="10">
<li></li>
</ol>
</td>
<td>Allowed maximum tilt fluctuation (in degree) between adjacent mlens</td>
</tr>
<tr class="row-odd"><td>centroidmethod</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Centroid method: 0 means barycentric (fast), 1 means gaussian fit (slow)</td>
</tr>
<tr class="row-even"><td>medfilter</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: Median filtering of dispersion coeff and tilts with a (5x5) median filtering</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>iscalib</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save to Calibrations Database, 0: save in regular reduced data dir</td>
</tr>
<tr class="row-odd"><td>lamp_override</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>0,1: override the filter/lamp combinations?</td>
</tr>
<tr class="row-even"><td>gpitvim_dispgrid</td>
<td>int</td>
<td>[0,500]</td>
<td>15</td>
<td>1-500: choose gpitv session for displaying image output and wavcal grid overplotted, 0: no display</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying wavcal file, 0: no display</td>
</tr>
<tr class="row-even"><td>tests</td>
<td>int</td>
<td>[0,3]</td>
<td>0</td>
<td>1 for extensive tests</td>
</tr>
<tr class="row-odd"><td>testsDST</td>
<td>int</td>
<td>[0,3]</td>
<td>0</td>
<td>1 for DST tests</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_measure_wavelength_calibration.pro</p>
</div>
<div class="section" id="measure-polarization-spot-calibration">
<span id="measurepolarizationspotcalibration"></span><span id="index-24"></span><h2>Measure Polarization Spot Calibration<a class="headerlink" href="#measure-polarization-spot-calibration" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Derive polarization calibration files from a flat field image.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 1.8</p>
<p><strong>Inputs</strong>:  2D image from flat field  in polarization mode</p>
<p><strong>Outputs</strong>:  Measured polarization spot locations calibration file      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>   gpi_extract_polcal detects the positions <span class="k">of</span> the polarized spots in a 2D
   <span class="nb">image</span> based on flat field observations.

ALGORITHM<span class="o">:</span>
   gpi_extract_polcal starts by detecting the central peak <span class="k">of</span> the <span class="nb">image</span>.
   Next, starting with a initial value <span class="k">of</span> w &amp; P, it finds the nearest peak (with an increment on the microlens coordinates)
   when nearest peak has been detected, it reevaluates w &amp; P <span class="k">and</span> so forth..

       Like the spectral mode wavelength calibration code, the first part <span class="k">of</span> this
       algorithm is devoted to determining the positions <span class="k">of</span> each spot on the
       detector.

       Unlike the spectral mode calibration, what we store here is in fact a
       weighted <span class="nb">list</span> <span class="k">of</span> pixels <span class="k">for</span> each lenslet PSF. FIXME <span class="o">-</span> this will need some
       revision to accomodate flexure...




HISTORY<span class="o">:</span>
  2009<span class="o">-</span>06<span class="o">-</span>17<span class="o">:</span> Started, based on gpi_extract_wavcal <span class="o">-</span> Marshall Perrin
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2013<span class="o">-</span>01<span class="o">-</span>28 MMB<span class="o">:</span> added some keywords to pass to find_pol_positions_quadrant
  2013<span class="o">-</span>07<span class="o">-</span>11 MDP<span class="o">:</span> Documentation cleanup.
  2013<span class="o">-</span>07<span class="o">-</span>12 MDP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>10<span class="o">-</span>31 MMB<span class="o">:</span> Big update to parallel code,
  2014<span class="o">-</span>11<span class="o">-</span>25 MDP<span class="o">:</span> Merge parallel <span class="k">and</span> single threaded code into one primitive
                                       with an option to <span class="k">switch</span> between them.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="6%" />
<col width="10%" />
<col width="8%" />
<col width="67%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>nlens</td>
<td>int</td>
<td>[0,400]</td>
<td>281</td>
<td>side length of  the  lenslet array</td>
</tr>
<tr class="row-odd"><td>centrXpos</td>
<td>int</td>
<td>[0,2048]</td>
<td>1078</td>
<td>Initial approximate x-position [pixel] of central peak at 1.5microns</td>
</tr>
<tr class="row-even"><td>centrYpos</td>
<td>int</td>
<td>[0,2048]</td>
<td>1028</td>
<td>Initial approximate y-position [pixel] of central peak at 1.5microns</td>
</tr>
<tr class="row-odd"><td>w</td>
<td>float</td>
<td>[0.,10.]</td>
<td>4.4</td>
<td>Spectral spacing perpendicular to the dispersion axis at the detcetor in pixel</td>
</tr>
<tr class="row-even"><td>P</td>
<td>float</td>
<td>[-7.,7.]</td>
<td>2.18</td>
<td>Micro-pupil pattern</td>
</tr>
<tr class="row-odd"><td>maxpos</td>
<td>float</td>
<td>[-7.,7.]</td>
<td>2.5</td>
<td>Allowed maximum location fluctuation (in pixel) between adjacent mlens</td>
</tr>
<tr class="row-even"><td>FitWidth</td>
<td>float</td>
<td>[-10.,10.]</td>
<td>3</td>
<td>Size of box around a spot used to find center</td>
</tr>
<tr class="row-odd"><td>parallel</td>
<td>Int</td>
<td>[0,1]</td>
<td>0</td>
<td>Option for Parallelization,  0: none, 1: parallel</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>None</td>
</tr>
<tr class="row-odd"><td>Display</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>None</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_measure_polarization_spot_calibration.pro</p>
</div>
<div class="section" id="assemble-polarization-cube">
<span id="assemblepolarizationcube"></span><span id="index-25"></span><h2>Assemble Polarization Cube<a class="headerlink" href="#assemble-polarization-cube" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Extract 2 perpendicular polarizations from a 2D image.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience, Calibration      <strong>Order</strong>: 2.0</p>
<p><strong>Inputs</strong>:  detector image in polarimetry mode</p>
<p><strong>Outputs</strong>:  Polarization pair datacube      <strong>Output Suffix</strong>: &#8216;-podc&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>        extract polarization<span class="o">-</span>mode data cube from an <span class="nb">image</span>
       define first suffix &#39;<span class="o">-</span>podc&#39; (polarization data<span class="o">-</span>cube)

       This routine transforms a 2D detector <span class="nb">image</span> in the dataset.currframe input
       structure into a 3D data cube in the dataset.currframe output structure.
       (<span class="k">not</span> much <span class="k">of</span> a data cube <span class="o">-</span> really just 2x 2D images)


ALGORITHM NOTES<span class="o">:</span>

   Ideally this should be done as an optimum weighting
   (see e.g. Naylor et al, 1997 MNRAS)

   That algorithm is as follows<span class="o">:</span> <span class="k">For</span> each lenslet spot,
      <span class="o">-</span>divide each pixel by the expected fraction <span class="k">of</span> the <span class="nb">total</span> lenslet flux
       in that pixel. (this makes each pixel an estimate <span class="k">of</span> the <span class="nb">total</span> lenslet
       flux)
       <span class="o">-</span>Combine these into a weighted average, weighted by the S<span class="o">/</span>N per pixel


<span class="k">common</span> needed<span class="o">:</span> filter, wavcal, tilt, (nlens)



HISTORY<span class="o">:</span>
  2009<span class="o">-</span>04<span class="o">-</span>22 MDP<span class="o">:</span> Created, based on DST&#39;s cubeextract_polarized.
  2009<span class="o">-</span>09<span class="o">-</span>17  JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>10<span class="o">-</span>08  JM<span class="o">:</span> add gpitv display
  2010<span class="o">-</span>10<span class="o">-</span>19  JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2011<span class="o">-</span>07<span class="o">-</span>15  MP<span class="o">:</span> Code cleanup.
  2011<span class="o">-</span>06<span class="o">-</span>07  JM<span class="o">:</span> added FITS<span class="o">/</span>MEF compatibility
  2013<span class="o">-</span>01<span class="o">-</span>02  MP<span class="o">:</span> Updated output file orientation to be consistent with
                   spectral mode <span class="k">and</span> raw data.
  2013<span class="o">-</span>07<span class="o">-</span>17  MP<span class="o">:</span> Renamed <span class="k">for</span> consistency
  2013<span class="o">-</span>11<span class="o">-</span>30  MP<span class="o">:</span> Clear DQ <span class="k">and</span> Uncert pointers
  2014<span class="o">-</span>02<span class="o">-</span>03  MP<span class="o">:</span> Code <span class="k">and</span> docs cleanup
  2014<span class="o">-</span>07<span class="o">-</span>01 MPF<span class="o">:</span> Modified &quot;PSF&quot; extraction <span class="k">for</span> weighting by a Gaussian
                  <span class="k">and</span> a noise <span class="nb">map</span>.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="8%" />
<col width="9%" />
<col width="9%" />
<col width="66%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
<tr class="row-even"><td>Method</td>
<td>String</td>
<td>BOX|PSF</td>
<td>BOX</td>
<td>Method for pol cube reconstruction, simple box or optimal PSF</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_assemble_polarization_cube.pro</p>
</div>
<div class="section" id="assemble-undispersed-image">
<span id="assembleundispersedimage"></span><span id="index-26"></span><h2>Assemble Undispersed Image<a class="headerlink" href="#assemble-undispersed-image" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Extract a 2D image from a raw undispersed mode image. Box integration of the light from each lenslet.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 2.0</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>:</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This routine performs a simple extraction <span class="k">of</span> GPI IFS undispersed
       data. It requires a pair <span class="k">of</span> fits files explicitly named xlocs.fits
       <span class="k">and</span> ylocs.fits located in the current directory. Those files contain
       a 300x300 array <span class="k">of</span> x <span class="k">and</span> y positions <span class="k">for</span> spots. These files are
       produced by the routine identify.<span class="k">pro</span> which examines a flood illuminated
       grid <span class="k">of</span> spots.

       The routine currently assumes the spots in the IFS are shifted by
       <span class="m">2</span>.36 <span class="k">and</span> <span class="m">2</span>.63 pixels from the time the calibration frame was taken
       in the UCLA lab. <span class="k">If</span> your <span class="nb">image</span> has significant flux in the central lenslets
       <span class="k">then</span> you can comment out the fitting portion <span class="k">of</span> the code, <span class="k">and</span> the
       pattern <span class="nb">shift</span> will be determined <span class="k">for</span> you.

       fname <span class="o">=</span> name <span class="k">of</span> the fits file you want to reduce
       outname <span class="o">=</span> name <span class="k">of</span> the output file this routine will produce

       example usage<span class="o">:</span>
            extu, &quot;test0159.fits&quot;, &quot;extu0159.fits&quot;


KEYWORDS<span class="o">:</span>


HISTORY<span class="o">:</span>
  Originally by James Larkin as extu.<span class="k">pro</span>
  2012<span class="o">-</span>02<span class="o">-</span>07 Pipelinified by Marshall Perrin
  2012<span class="o">-</span>03<span class="o">-</span>30 Rotated by 90 deg to match spectral cube orientation. NaNs outside <span class="k">of</span> FOV. <span class="o">-</span> MP
  2013<span class="o">-</span>03<span class="o">-</span>08 JM<span class="o">:</span> added manual shifts <span class="k">of</span> the spot due to flexure
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>11<span class="o">-</span>30 MDP<span class="o">:</span> Clear DQ <span class="k">and</span> Uncert pointers
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="8%" />
<col width="11%" />
<col width="9%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>xshift</td>
<td>float</td>
<td>[-100,100]</td>
<td>-2.363</td>
<td>Shift in X direction</td>
</tr>
<tr class="row-odd"><td>yshift</td>
<td>float</td>
<td>[-100,100]</td>
<td>-2.6134</td>
<td>Shift in Y direction</td>
</tr>
<tr class="row-even"><td>boxsize</td>
<td>float</td>
<td>[0,10]</td>
<td>5</td>
<td>Size of box to use for spectral extraction</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>suffix</td>
<td>string</td>
<td>None</td>
<td>-extu</td>
<td>Enter output suffix</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_assemble_undispersed_image.pro</p>
</div>
<div class="section" id="assemble-spectral-datacube-using-mlens-psf">
<span id="assemblespectraldatacubeusingmlenspsf"></span><span id="index-27"></span><h2>Assemble Spectral Datacube using mlens PSF<a class="headerlink" href="#assemble-spectral-datacube-using-mlens-psf" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Extract a 3D datacube from a 2D image. Spatial integration (3 pixels) along the dispersion axis</div></blockquote>
<p><strong>Category</strong>:  SpectralScience, Calibration      <strong>Order</strong>: 2.0</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>:       <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>               This routine transforms a 2D detector <span class="nb">image</span> in the dataset.currframe input
               structure into a 3D data cube in the dataset.currframe output structure.
  This routine extracts data cube from an <span class="nb">image</span> using an inversion method along the dispersion <span class="nb">axis</span>



KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>


HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2007<span class="o">-</span>11
  2012<span class="o">-</span>02<span class="o">-</span>01 JM<span class="o">:</span> adapted to vertical dispersion
  2012<span class="o">-</span>02<span class="o">-</span>15 JM<span class="o">:</span> adapted as a pipeline module
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the mlens-PSF calibration file to be read</td>
</tr>
<tr class="row-even"><td>ReuseOutput</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: keep output for following primitives, 0: don&#8217;t keep</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_extractcube_mlenspsf.pro</p>
</div>
<div class="section" id="assemble-spectral-datacube-using-epsf">
<span id="assemblespectraldatacubeusingepsf"></span><span id="index-28"></span><h2>Assemble Spectral Datacube using ePSF<a class="headerlink" href="#assemble-spectral-datacube-using-epsf" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Extract a 3D datacube from a 2D image. Spatial integration (3 pixels) along the dispersion axis</div></blockquote>
<p><strong>Category</strong>:  SpectralScience, Calibration      <strong>Order</strong>: 2.0</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>:       <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>               This routine transforms a 2D detector <span class="nb">image</span> in the dataset.currframe input
               structure into a 3D data cube in the dataset.currframe output structure.
  This routine extracts data cube from an <span class="nb">image</span> using an inversion method along the dispersion <span class="nb">axis</span>



KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>


HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2007<span class="o">-</span>11
  2012<span class="o">-</span>02<span class="o">-</span>01 JM<span class="o">:</span> adapted to vertical dispersion
  2012<span class="o">-</span>02<span class="o">-</span>15 JM<span class="o">:</span> adapted as a pipeline module
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
  2014<span class="o">-</span>07<span class="o">-</span>18 JM<span class="o">:</span> implemented ePSF instead <span class="k">of</span> DST simulated PSF
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the mlens-PSF calibration file to be read</td>
</tr>
<tr class="row-even"><td>ReuseOutput</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: keep output for following primitives, 0: don&#8217;t keep</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_extractcube_epsf.pro</p>
</div>
<div class="section" id="assemble-spectral-datacube">
<span id="assemblespectraldatacube"></span><span id="index-29"></span><h2>Assemble Spectral Datacube<a class="headerlink" href="#assemble-spectral-datacube" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Assemble a 3D datacube from a 2D image. Spatial integration (3 pixels box) along the dispersion axis</div></blockquote>
<p><strong>Category</strong>:  SpectralScience, Calibration      <strong>Order</strong>: 2.0</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>: &#8216;-rawspdc&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>               This routine transforms a 2D detector <span class="nb">image</span> in the dataset.currframe input
               structure into a 3D data cube in the dataset.currframe output structure.
  This routine extracts data cube from an <span class="nb">image</span> using spatial summation along the dispersion <span class="nb">axis</span>
    introduced suffix &#39;<span class="o">-</span>rawspdc&#39; (raw spectral data<span class="o">-</span>cube)

KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>IFSFILT


HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2007<span class="o">-</span>11
  2008<span class="o">-</span>04<span class="o">-</span>02 JM<span class="o">:</span> spatial summation <span class="nb">window</span> centered on pixel <span class="k">and</span> interpolation on the zem. comm. wav. <span class="nb">vector</span>
         2008<span class="o">-</span>06<span class="o">-</span>06 JM<span class="o">:</span> adapted to pipeline inputs
  2009<span class="o">-</span>04<span class="o">-</span>15 MDP<span class="o">:</span> Documentation updated.
  2009<span class="o">-</span>06<span class="o">-</span>20 JM<span class="o">:</span> adapted to wavcal input
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2012<span class="o">-</span>02<span class="o">-</span>01 JM<span class="o">:</span> adapted to vertical dispersion
  2012<span class="o">-</span>02<span class="o">-</span>09 DS<span class="o">:</span> offloaded sdpx calculation
  2013<span class="o">-</span>04<span class="o">-</span>02 JBR<span class="o">:</span> Correction on the y coordinate when reading the det array to match centered pixel convention. Removal <span class="k">of</span> the reference pixel area.
  2013<span class="o">-</span>07<span class="o">-</span>17 MDP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>08<span class="o">-</span>06 MDP<span class="o">:</span> Documentation update, code cleanup to relabel X <span class="k">and</span> Y properly
  2013<span class="o">-</span>11<span class="o">-</span>30 MDP<span class="o">:</span> Clear DQ <span class="k">and</span> Uncert pointers
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_assemble_spectral_datacube.pro</p>
</div>
<div class="section" id="filter-datacube-spatially">
<span id="filterdatacubespatially"></span><span id="index-30"></span><h2>Filter datacube spatially<a class="headerlink" href="#filter-datacube-spatially" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Apply spatial filter to datacubes</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 2.1</p>
<p><strong>Inputs</strong>:  raw 2D image file</p>
<p><strong>Outputs</strong>:  2D image corrected for dark current      <strong>Output Suffix</strong>:  &#8216;sfilt&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>High<span class="o">-</span>pass <span class="k">or</span> Low<span class="o">-</span>pass filter each slice <span class="k">of</span> a GPI datacube using a <span class="nb">median</span> box filter.

This is useful <span class="k">for</span> removing the halos created by uncorrected atmospheric turbulence. This is a tad slow but a useful tool.

Skip_parallelization should be used when opening IDL sessions is slow (when using floating license servers etc)

Other filters may be added later.





HISTORY<span class="o">:</span>
       Originally by Patrick Ingraham Apr <span class="m">2</span>, 2014
       MMB updated to include lowpass <span class="k">and</span> polarimetry.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="11%" />
<col width="8%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>hp_boxsize</td>
<td>int</td>
<td>[0,50]</td>
<td>0</td>
<td>0: no filter, 1+: Filter box size</td>
</tr>
<tr class="row-odd"><td>high_or_lowpass</td>
<td>string</td>
<td>[high|low]</td>
<td>high</td>
<td>High pass or lowpass filter?</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_filter_datacube_spatially.pro</p>
</div>
<div class="section" id="noise-and-flux-analysis">
<span id="noiseandfluxanalysis"></span><span id="index-31"></span><h2>Noise and Flux Analysis<a class="headerlink" href="#noise-and-flux-analysis" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Store a few key values as fits keywords in the file. It can generate anciliary files too.</div></blockquote>
<p><strong>Category</strong>:  HIDDEN      <strong>Order</strong>: 2.1</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>:  Changes is the header of the file without changing the data and saving a fits file report with the value of the sliding median/standard deviation computation.      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  <span class="o">/</span>!\ HIDDEN <span class="o">/</span>!\ It was a primitive used by JB <span class="k">for</span> debug but I don&#39;t think it is gonna be used by anyone <span class="k">else</span>.

  This routine quantifies the noise <span class="k">and</span> the flux in an <span class="nb">image</span> without changing it. It generates fits keyword <span class="k">for</span> this values <span class="k">for</span> further easy <span class="nb">image</span> sorting.
  <span class="k">If</span> asked, it can generate a fits files too.

  <span class="k">If</span> Flux <span class="o">=</span> <span class="m">1</span><span class="o">:</span> Generate fits keywords related with <span class="nb">total</span> flux in the <span class="nb">image</span>
    DN, <span class="nb">total</span> data number <span class="k">of</span> the <span class="nb">image</span>.
    DNLENS, <span class="nb">total</span> data number in the lenslets aera (<span class="k">if</span> <span class="k">not</span> a dark <span class="k">and</span> <span class="k">not</span> a cube)
    DNBACK, <span class="nb">total</span> data number outside the lenslets aera (<span class="k">if</span> <span class="k">not</span> a dark <span class="k">and</span> <span class="k">not</span> a cube)


  <span class="k">If</span> StddevMed <span class="o">&gt;</span> <span class="m">1</span><span class="o">:</span> Generate fits keywords related with the standard deviation in the <span class="nb">image</span>

  <span class="k">If</span> StddevMed <span class="o">=</span> <span class="m">2</span><span class="o">:</span>
    Compute the local <span class="nb">median</span> <span class="k">and</span> the local standard deviation by moving a square <span class="k">of</span> <span class="nb">size</span> Width.
    Because it is time consuming, you can skip pixels using the parameter PixelsSkipped.
    In the output, the <span class="nb">finite</span> value pixels correspond to pixels <span class="nb">where</span> the media <span class="k">and</span> the standard deviation were computed.
    <span class="k">If</span> 2d <span class="nb">image</span><span class="o">:</span> Generate a file with the suffix &#39;<span class="o">-</span>stddevmed&#39; containing an 3d array. [<span class="o">*</span>,<span class="o">*</span>,<span class="m">0</span>] is the <span class="nb">median</span> <span class="k">and</span> [<span class="o">*</span>,<span class="o">*</span>,<span class="m">1</span>] is the standard deviation.
    <span class="k">If</span> 3d <span class="nb">image</span><span class="o">:</span> Generate two files &#39;<span class="o">-</span><span class="nb">stddev</span>&#39; <span class="k">and</span> &#39;<span class="o">-</span><span class="nb">median</span>&#39;. Both same <span class="nb">size</span> <span class="k">of</span> the original <span class="nb">image</span>.


  <span class="k">If</span> microNoise <span class="o">=</span> <span class="m">1</span><span class="o">:</span>
    Estimate the quantity <span class="k">of</span> microphonics noise in the <span class="nb">image</span> based on a model stored as a calibration file.
    The quantity <span class="k">of</span> microphonics noise is measured with the ratio <span class="k">of</span> the dot_product <span class="k">and</span> the <span class="nb">norm</span> <span class="k">of</span> the <span class="nb">image</span><span class="o">:</span> dot_product<span class="o">/</span><span class="nb">sqrt</span>(sum(<span class="nb">abs</span>(<span class="nb">fft</span>(<span class="nb">image</span>))<span class="o">^</span><span class="m">2</span>)).
    With dot_product <span class="o">=</span> sum(<span class="nb">abs</span>(<span class="nb">fft</span>(<span class="nb">image</span>))<span class="o">*</span><span class="nb">abs</span>(<span class="nb">fft</span>(noise_model))) which correspond to the projection <span class="k">of</span> the <span class="nb">image</span> on the microphonics noise model in the absolute Fourier space.
    The fits keyword associated is MICRONOI.


  <span class="k">If</span> FourierTransf <span class="o">=</span> <span class="m">1</span> <span class="k">or</span> <span class="m">2</span><span class="o">:</span>
    Build <span class="k">and</span> <span class="nb">save</span> the Fourier transform <span class="k">of</span> the <span class="nb">image</span>.
    <span class="k">If</span> <span class="m">1</span>, the output is the one directly from the idl <span class="k">function</span> (<span class="nb">fft</span>). Therefore, the Fourier <span class="nb">image</span> is <span class="k">not</span> centered.
    <span class="k">If</span> <span class="m">2</span>, the output will be centered.
    In the <span class="k">case</span> <span class="k">of</span> a cube it is <span class="k">not</span> a 3d <span class="nb">fft</span> that is performed but several 2d ffts.
    suffix<span class="o">=</span>&#39;<span class="o">-</span>absfft&#39; <span class="k">or</span> suffix<span class="o">=</span>&#39;<span class="o">-</span>absfftdc&#39; <span class="k">if</span> it is a cube.



HISTORY<span class="o">:</span>
  Originally by Jean<span class="o">-</span>Baptiste Ruffio 2013<span class="o">-</span>05
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="3%" />
<col width="4%" />
<col width="4%" />
<col width="82%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Flux</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Trigger flux analysis</td>
</tr>
<tr class="row-odd"><td>StddevMed</td>
<td>int</td>
<td>[0,2]</td>
<td>1</td>
<td>Trigger the standard deviation (and median) analysis of the image. if StddevMed=1, only keywords and log are produced. If StddevMed=2, fits files are generated with a sliding median and standard deviation.</td>
</tr>
<tr class="row-even"><td>Width</td>
<td>int</td>
<td>[3,2048]</td>
<td>101</td>
<td>If Stddev = 2, Width of the moving rectangle. It has to be odd.</td>
</tr>
<tr class="row-odd"><td>PixelsSkipped</td>
<td>int</td>
<td>[0,2047]</td>
<td>100</td>
<td>If Stddev = 2, Pixels skipped between two points</td>
</tr>
<tr class="row-even"><td>MicroNoise</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Trigger the microphonics noise analysis</td>
</tr>
<tr class="row-odd"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired microphonics model file to be read</td>
</tr>
<tr class="row-even"><td>FourierTransf</td>
<td>int</td>
<td>[0,1,2]</td>
<td>1</td>
<td>1: frequency 0 on the bottom left. 2: frequencies 0 will be centered on the image.</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_noise_and_flux_analysis.pro</p>
</div>
<div class="section" id="divide-by-lenslet-flat-field">
<span id="dividebylensletflatfield"></span><span id="index-32"></span><h2>Divide by Lenslet Flat Field<a class="headerlink" href="#divide-by-lenslet-flat-field" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Divides a spectral data-cube by a flat field data-cube.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,PolarimetricScience,Calibration      <strong>Order</strong>: 2.2</p>
<p><strong>Inputs</strong>:  Spectral or polarization datacube</p>
<p><strong>Outputs</strong>:  Each slice of the input datacube is divided by the lenslet flat.</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>HISTORY<span class="o">:</span>
  2014<span class="o">-</span>01<span class="o">-</span>02 MP<span class="o">:</span> New primitive
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired wavelength calibration file to be read</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_divide_by_lenslet_flat_field.pro</p>
</div>
<div class="section" id="divide-by-spectral-flat-field">
<span id="dividebyspectralflatfield"></span><span id="index-33"></span><h2>Divide by Spectral Flat Field<a class="headerlink" href="#divide-by-spectral-flat-field" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Divides a spectral data-cube by a flat field data-cube.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,Calibration      <strong>Order</strong>: 2.2</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:  Flat fielded datacube</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  <span class="o">**</span> Needs additional work, will <span class="k">not</span> produce high qualty results yet <span class="o">**</span>



HISTORY<span class="o">:</span>
  2009<span class="o">-</span>08<span class="o">-</span>27<span class="o">:</span> JM created
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>10<span class="o">-</span>09 JM added gpitv display
  2010<span class="o">-</span>10<span class="o">-</span>19 JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2011<span class="o">-</span>07 JM<span class="o">:</span> added check <span class="k">for</span> NAN &amp; zero
  2012<span class="o">-</span>10<span class="o">-</span>11 MP<span class="o">:</span> Added <span class="nb">min</span><span class="o">/</span><span class="nb">max</span> wavelength checks
  2012<span class="o">-</span>10<span class="o">-</span>17 MP<span class="o">:</span> Removed deprecated suffix<span class="o">=</span> keyword
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
       2013<span class="o">-</span>12<span class="o">-</span>30 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired wavelength calibration file to be read</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_divide_by_spectral_flat_field.pro</p>
</div>
<div class="section" id="remove-flat-lamp-spectrum">
<span id="removeflatlampspectrum"></span><span id="index-34"></span><h2>Remove Flat Lamp spectrum<a class="headerlink" href="#remove-flat-lamp-spectrum" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Fit the lamp spectrum and remove it (for delivering flat field cubes)</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 2.25</p>
<p><strong>Inputs</strong>:  Flat field data-cube</p>
<p><strong>Outputs</strong>:   Flat datacube normalized to remove lamp spectrum      <strong>Output Suffix</strong>: &#8216;specflat&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>          Rescale flat<span class="o">-</span>field (keep large scale variations)

          <span class="o">**</span>CAUTION needs additional improvement <span class="o">**</span>





GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>
DRP KEYWORDS<span class="o">:</span> FILETYPE, ISCALIB



HISTORY<span class="o">:</span>
       2009<span class="o">-</span>06<span class="o">-</span>20 JM<span class="o">:</span> created
       2009<span class="o">-</span>07<span class="o">-</span>22 MP<span class="o">:</span> added doc header keywords
       2012<span class="o">-</span>10<span class="o">-</span>11 MP<span class="o">:</span> added <span class="nb">min</span><span class="o">/</span><span class="nb">max</span> wavelength checks
       2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>03 MP<span class="o">:</span> Add check <span class="k">for</span> GCALLAMP<span class="o">=</span>QH on input images
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="6%" />
<col width="25%" />
<col width="9%" />
<col width="54%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
<tr class="row-even"><td>method</td>
<td>string</td>
<td>polyfit|linfit|blackbody|none</td>
<td>blackbody</td>
<td>Method to use for removing lamp spectrum</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_remove_flat_lamp_spectrum.pro</p>
</div>
<div class="section" id="interpolate-wavelength-axis">
<span id="interpolatewavelengthaxis"></span><span id="index-35"></span><h2>Interpolate Wavelength Axis<a class="headerlink" href="#interpolate-wavelength-axis" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Interpolate spectral datacube onto regular wavelength sampling.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,Calibration      <strong>Order</strong>: 2.3</p>
<p><strong>Inputs</strong>:   A raw irregularly-sampled spectral datacube</p>
<p><strong>Outputs</strong>:  Spectral datacube with slices at a regular wavelength sampling      <strong>Output Suffix</strong>: &#8216;spdc&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>               <span class="nb">Interpolate</span> datacube to have each slice at the same wavelength.
               This is a necessary step <span class="k">of</span> creating datacubes in spectral mode
               <span class="k">and</span> should always be used right after Assemble Spectral Datacube.

               Also adds wavelength keywords to the FITS header.



HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2008<span class="o">-</span>06
       2009<span class="o">-</span>04<span class="o">-</span>15 MDP<span class="o">:</span> Documentation improved.
  2009<span class="o">-</span>06<span class="o">-</span>20 JM<span class="o">:</span> adapted to wavcal
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2010<span class="o">-</span>03<span class="o">-</span>15 JM<span class="o">:</span> added error handling
  2012<span class="o">-</span>12<span class="o">-</span>09 MP<span class="o">:</span> Updates to WCS output
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="6%" />
<col width="8%" />
<col width="8%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Spectralchannels</td>
<td>int</td>
<td>[0,100]</td>
<td>37</td>
<td>Choose how many spectral channels for output datacube</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_interpolate_wavelength_axis.pro</p>
</div>
<div class="section" id="subtract-thermal-sky-background-cube-if-k-band">
<span id="subtractthermal-skybackgroundcubeifkband"></span><span id="index-36"></span><h2>Subtract Thermal/Sky Background Cube if K band<a class="headerlink" href="#subtract-thermal-sky-background-cube-if-k-band" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Subtract a thermal/sky cube</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 2.35</p>
<p><strong>Inputs</strong>:  3D image file</p>
<p><strong>Outputs</strong>:  3D image file, unchanged if YJH, background subtracted if K1 or K2.      <strong>Output Suffix</strong>:  &#8216;bkgnd_cube_sub&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> Subtract thermal background emission in the datacube, <span class="k">for</span> K band data only

 This is identical to the gpi_subtact_thermal_sky_if_k_band primtive except the subtraction
 is done in cube space instead <span class="k">of</span> detector space. It also uses sky cubes rather than the 2d sky images.

       <span class="o">**</span> special note<span class="o">:</span> <span class="o">**</span>

       This is a new kind <span class="k">of</span> &quot;data dependent optional primitive&quot;. <span class="k">If</span> the filter <span class="k">of</span>
       the current data is YJH, <span class="nb">return</span> without doing <span class="o">*</span>anything<span class="o">*</span>, even logging the
       start<span class="o">/</span><span class="k">end</span> <span class="k">of</span> this primitive.  It becomes a complete no<span class="o">-</span>op <span class="k">for</span> non<span class="o">-</span>K<span class="o">-</span>band
       cases.

Algorithm<span class="o">:</span>

       Get the best available thermal<span class="o">/</span>sky background cube calibration file from CalDB
       Scale it to current exposure time
       Subtract it.
  The name <span class="k">of</span> the calibration file used is saved to the DRPBKGND header keyword.

ALGORITHM TODO<span class="o">:</span> Deal with uncertainty <span class="k">and</span> pixel mask frames too.





HISTORY<span class="o">:</span>
  2013<span class="o">-</span>12<span class="o">-</span>23 PI<span class="o">:</span> Initial implementation
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="5%" />
<col width="6%" />
<col width="7%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Name of thermal/sky background cube to subtract</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>Override_scaling</td>
<td>float</td>
<td>[0,10]</td>
<td>1.0</td>
<td>Set to value other than 1 to manually adjust the background image flux scaling to better match the science data</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_subtract_thermal_sky_background_cube_if_k_band.pro</p>
</div>
<div class="section" id="check-for-closed-loop-coronagraphic-image">
<span id="checkforclosed-loopcoronagraphicimage"></span><span id="index-37"></span><h2>Check for closed-loop coronagraphic image<a class="headerlink" href="#check-for-closed-loop-coronagraphic-image" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Check whether file represents a closed-loop  coronagraphic image.</div></blockquote>
<p><strong>Category</strong>:  Calibration,SpectralScience,PolarimetricScience      <strong>Order</strong>: 2.41</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>: Not specified</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This primitive checks that the input file is in fact a coronagraphic <span class="nb">image</span>.
       It is intended to be used in quicklook recipes that may encounter all sorts
       <span class="k">of</span> different data.

       Any following primitives will only be executed <span class="k">if</span> the
       <span class="nb">image</span> is in fact coronagraphic data. This is useful so the quicklook
       recipe can include satellite spots <span class="k">or</span> contrast measurement primitives,
       which would generally cause the recipe to fail <span class="k">if</span> they receive any
       unocculted data. With this primitive added in the recipe before those
       steps, they will just be skipped without producing any error messages.



HISTORY<span class="o">:</span>
  2013<span class="o">-</span>08<span class="o">-</span>02 ds <span class="o">-</span> initial version
  2013<span class="o">-</span>11<span class="o">-</span>12 MP <span class="o">-</span> add check <span class="k">for</span> PUPVIEWR inserted
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>err_on_false</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>If false, 0: continue to next image; 1: Throw error</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_check_coronagraph_status.pro</p>
</div>
<div class="section" id="correct-distortion">
<span id="correctdistortion"></span><span id="index-38"></span><h2>Correct Distortion<a class="headerlink" href="#correct-distortion" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Correct GPI distortion</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,PolarimetricScience      <strong>Order</strong>: 2.44</p>
<p><strong>Inputs</strong>:  spectral or polarimetric datacube</p>
<p><strong>Outputs</strong>:   Distortion-corrected datacube      <strong>Output Suffix</strong>: &#8216;_distorcorr&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Corrects distortion by <span class="nb">bilinear</span> resampling <span class="k">of</span> the
       input datacube according to a predetermined distortion solution.

       Note that this primitive can go <span class="o">*</span>either<span class="o">*</span> before <span class="k">or</span> after
       Accumulate Images.
       As a Level <span class="m">1</span> primitive, it will undistort one cube at a time;
       As a Level <span class="m">2</span> primitive it will undistort the whole stack <span class="k">of</span>
       accumulated images all at once.

       This primitive <span class="o">*</span>MUST<span class="o">*</span> be run before the &#39;Measure Satellite spot locations&#39; primitive







HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2009<span class="o">-</span>12
  2013<span class="o">-</span>04<span class="o">-</span>23 Major change <span class="k">of</span> the code, now based on Quinn&#39;s routine <span class="k">for</span> distortion correction <span class="o">-</span> JM
  2013<span class="o">-</span>07<span class="o">-</span>16 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
       2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
       2014<span class="o">-</span>05<span class="o">-</span>10 MP<span class="o">:</span> Update to enable this to work before <span class="k">or</span> after accumulate
       images.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired distortion calibration file to be read</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>10</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_correct_distortion.pro</p>
</div>
<div class="section" id="measure-satellite-spot-locations">
<span id="measuresatellitespotlocations"></span><span id="index-39"></span><h2>Measure satellite spot locations<a class="headerlink" href="#measure-satellite-spot-locations" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Measure the locations of the satellite spots in the datacube, and save the results to the FITS keyword headers.</div></blockquote>
<p><strong>Category</strong>:  Calibration,SpectralScience      <strong>Order</strong>: 2.44</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>: Not specified</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> Measures the locations <span class="k">of</span> the satellite spots; saves to FITS keywords.
 The sat spots locations are saved to SATS1_1, SATS1_2, <span class="k">and</span> so on.
 The inferred location <span class="k">of</span> the star is saved to PSFCENTX <span class="k">and</span> PSFCENTY
 (this is the <span class="nb">mean</span> location <span class="k">of</span> all the locations at each wavelength)

 By default, the sat spots information are saved to the FITS header keywords
 <span class="k">of</span> the current file in <span class="nb">memory</span>, <span class="k">and</span> will only be saved <span class="k">if</span> you subsequently
 <span class="nb">save</span> that datacube (i.e. using &#39;<span class="nb">save</span><span class="o">=</span><span class="m">1</span>&#39; on this primitive <span class="k">or</span> a subsequent
 one). The &#39;update_prev_fits_header&#39; option will, in addition, also let you
 write the same keyword information to the header <span class="k">of</span> the most recently saved
 file. This is useful <span class="k">if</span> you have just already saved the datacube, <span class="k">and</span> you
 only now want to update this metadata.



HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2009<span class="o">-</span>12
  2012<span class="o">-</span>09<span class="o">-</span>18 Offloaded functionality to <span class="k">common</span> backend <span class="o">-</span> ds
  2013<span class="o">-</span>07<span class="o">-</span>17 MP Documentation updated, rename <span class="k">for</span> consistency.
  2014<span class="o">-</span>10<span class="o">-</span>11 DS Added PSFCEN_XX <span class="k">for</span> all slices to header
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="4%" />
<col width="7%" />
<col width="7%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>refine_fits</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>0: Use wavelength scaling only; 1: Fit each slice</td>
</tr>
<tr class="row-odd"><td>reference_index</td>
<td>int</td>
<td>[-1,50]</td>
<td>-1</td>
<td>Index of slice to use for initial satellite detection. -1 for Auto.</td>
</tr>
<tr class="row-even"><td>search_window</td>
<td>int</td>
<td>[1,50]</td>
<td>20</td>
<td>Radius of aperture used for locating satellite spots.</td>
</tr>
<tr class="row-odd"><td>highpass</td>
<td>int</td>
<td>[0,25]</td>
<td>1</td>
<td>1: Use high pass filter (default size) 0: don&#8217;t 2+: size of highpass filter box</td>
</tr>
<tr class="row-even"><td>constrain</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: Constrain distance between sat spots by band; 0: Unconstrained search.</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>update_prev_fits_header</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Update FITS metadata in the most recently saved datacube?</td>
</tr>
<tr class="row-odd"><td>loc_input</td>
<td>int</td>
<td>[0,2]</td>
<td>0</td>
<td>0: Find spots automatically; 1: Use values below as initial satellite spot location</td>
</tr>
<tr class="row-even"><td>x1</td>
<td>int</td>
<td>[0,300]</td>
<td>0</td>
<td>approx x-location of top left spot on reference slice of the datacube in pixels</td>
</tr>
<tr class="row-odd"><td>y1</td>
<td>int</td>
<td>[0,300]</td>
<td>0</td>
<td>approx y-location of top left spot on reference slice of the datacube in pixels</td>
</tr>
<tr class="row-even"><td>x2</td>
<td>int</td>
<td>[0,300]</td>
<td>0</td>
<td>approx x-location of bottom left spot on reference slice of the datacube in pixels</td>
</tr>
<tr class="row-odd"><td>y2</td>
<td>int</td>
<td>[0,300]</td>
<td>0</td>
<td>approx y-location of bottom left spot on reference slice of the datacube in pixels</td>
</tr>
<tr class="row-even"><td>x3</td>
<td>int</td>
<td>[0,300]</td>
<td>0</td>
<td>approx x-location of top right spot on reference slice of the datacube in pixels</td>
</tr>
<tr class="row-odd"><td>y3</td>
<td>int</td>
<td>[0,300]</td>
<td>0</td>
<td>approx y-location of top right spot on reference slice of the datacube in pixels</td>
</tr>
<tr class="row-even"><td>x4</td>
<td>int</td>
<td>[0,300]</td>
<td>0</td>
<td>approx x-location of bottom right spot on reference slice of the datacube in pixels</td>
</tr>
<tr class="row-odd"><td>y4</td>
<td>int</td>
<td>[0,300]</td>
<td>0</td>
<td>approx y-location of bottom right spot on reference slice of the datacube in pixels</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_measure_satellite_spot_locations.pro</p>
</div>
<div class="section" id="measure-gpi-distortion-from-grid-pattern">
<span id="measuregpidistortionfromgridpattern"></span><span id="index-40"></span><h2>Measure GPI distortion from grid pattern<a class="headerlink" href="#measure-gpi-distortion-from-grid-pattern" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Measure GPI distortion from grid pattern</div></blockquote>
<p><strong>Category</strong>:  Calibration HIDDEN      <strong>Order</strong>: 2.44</p>
<p><strong>Inputs</strong>:  Not used</p>
<p><strong>Outputs</strong>:  Distortion correction coefficients file (values are currently hard-coded)      <strong>Output Suffix</strong>: &#8216;-distor&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       CAUTION <span class="o">-</span> <span class="k">NOT</span> IMPLEMENTED

       the distortion in lab was analyzed with other non<span class="o">-</span>pipeline tools by Quinn,
       <span class="k">and</span> the result is just hard coded here to output it in the GPI pipeline
       format.





HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2009<span class="o">-</span>12
      Switched sxaddpar to backbone<span class="o">-&gt;</span>set_keyword 01.31.2012 Dmitry Savransky
  2013  Added hard<span class="o">-</span>coded measurement <span class="k">of</span> the distortion made by Quinn   JM
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="9%" />
<col width="10%" />
<col width="13%" />
<col width="58%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_measure_distortion.pro</p>
</div>
<div class="section" id="measure-star-position-for-polarimetry">
<span id="measurestarpositionforpolarimetry"></span><span id="index-41"></span><h2>Measure Star Position for Polarimetry<a class="headerlink" href="#measure-star-position-for-polarimetry" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Finds the location of the occulted star in polarimetry mode, and save the results to the FITS keyword headers.</div></blockquote>
<p><strong>Category</strong>:  Calibration, PolarimetricScience      <strong>Order</strong>: 2.445</p>
<p><strong>Inputs</strong>:  Polarimetric mode datacube</p>
<p><strong>Outputs</strong>:  Polarimetric mode datacube with star location recorded in header</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> Finds the location <span class="k">of</span> the occulted star (i.e. <span class="nb">image</span> center); saves center to FITS keywords.
 The algorithm used is a version <span class="k">of</span> the <span class="nb">radon</span> transform, used to find <span class="nb">where</span>
 all the broadband extended speckles intersect in the <span class="nb">image</span>.

 The inferred star location is saved to PSFCENTX, PSFCENY keywords in the
 header





HISTORY<span class="o">:</span>
       2014<span class="o">-</span>01<span class="o">-</span>31 JW<span class="o">:</span> Created. Accurary is subpixel <span class="o">-</span> hopefully.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="6%" />
<col width="15%" />
<col width="8%" />
<col width="57%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>x0</td>
<td>int</td>
<td>[0,300]</td>
<td>147</td>
<td>initial guess for image center x-coordinate</td>
</tr>
<tr class="row-odd"><td>y0</td>
<td>int</td>
<td>[0,300]</td>
<td>147</td>
<td>inital guess ofr image center y-coordinate</td>
</tr>
<tr class="row-even"><td>search_window</td>
<td>int</td>
<td>[1,50]</td>
<td>5</td>
<td>Radius of search window to search for the center</td>
</tr>
<tr class="row-odd"><td>mask_radius</td>
<td>int</td>
<td>[0,100]</td>
<td>50</td>
<td>Radius of center of image to mask (centered on x0, y0 inputs)</td>
</tr>
<tr class="row-even"><td>highpass</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: Use high pass filter 0: don&#8217;t</td>
</tr>
<tr class="row-odd"><td>lower_threshold</td>
<td>float</td>
<td>[-100000,100000]</td>
<td>-100</td>
<td>Lower pixel values will be converted to this value</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_measure_star_position_for_polarimetry.pro</p>
</div>
<div class="section" id="measure-satellite-spot-peak-fluxes">
<span id="measuresatellitespotpeakfluxes"></span><span id="index-42"></span><h2>Measure satellite spot peak fluxes<a class="headerlink" href="#measure-satellite-spot-peak-fluxes" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Calculate peak fluxes of satellite spots in datacubes</div></blockquote>
<p><strong>Category</strong>:  Calibration,SpectralScience      <strong>Order</strong>: 2.45</p>
<p><strong>Inputs</strong>:  spectral datacube with spot locations in the header</p>
<p><strong>Outputs</strong>:   datacube with measured spot fluxes</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  Measure the fluxes <span class="k">of</span> the satellite spots.
  You must run &#39;Measure Satellite Spot Locations&#39; before you can use this
  one.

       Spot fluxes are measured <span class="k">and</span> <span class="k">then</span> saved to SATF1_1, SATF1_2 etc keywords
       in the header.

  By default, the sat spots information are saved to the FITS header keywords
  <span class="k">of</span> the current file in <span class="nb">memory</span>, <span class="k">and</span> will only be saved <span class="k">if</span> you subsequently
  <span class="nb">save</span> that datacube (i.e. using &#39;<span class="nb">save</span><span class="o">=</span><span class="m">1</span>&#39; on this primitive <span class="k">or</span> a subsequent
  one). The &#39;update_prev_fits_header&#39; option will, in addition, also let you
  write the same keyword information to the header <span class="k">of</span> the most recently saved
  file. This is useful <span class="k">if</span> you have just already saved the datacube, <span class="k">and</span> you
  only now want to update this metadata.





HISTORY<span class="o">:</span>
       Written 09<span class="o">-</span>18<span class="o">-</span>2012 savransky1@llnl.gov
       2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Renamed <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="22%" />
<col width="5%" />
<col width="7%" />
<col width="8%" />
<col width="58%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>gauss_fit</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>0: Extract maximum pixel; 1: Correlate with Gaussian to find peak</td>
</tr>
<tr class="row-odd"><td>reference_index</td>
<td>int</td>
<td>[0,50]</td>
<td>0</td>
<td>Index of slice to use for initial satellite detection.</td>
</tr>
<tr class="row-even"><td>ap_rad</td>
<td>int</td>
<td>[1,50]</td>
<td>7</td>
<td>Radius of aperture used for finding peaks.</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>update_prev_fits_header</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Update FITS metadata in the most recently saved datacube?</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_measure_satellite_spot_peak_fluxes.pro</p>
</div>
<div class="section" id="correct-for-atmospheric-differential-refraction">
<span id="correctforatmosphericdifferentialrefraction"></span><span id="index-43"></span><h2>Correct for Atmospheric Differential Refraction<a class="headerlink" href="#correct-for-atmospheric-differential-refraction" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Interpolates the cube to undo any shifts due to ADR (or leftover ADC).</div></blockquote>
<p><strong>Category</strong>:  Calibration, SpectralScience      <strong>Order</strong>: 2.45</p>
<p><strong>Inputs</strong>:  Spectral Cube</p>
<p><strong>Outputs</strong>:  Interpolated Spectral Cube that is shifted for ADR compensation</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Uses the location <span class="k">of</span> the satellite spots to calculate a center at each
       wavelength, fits the x <span class="k">and</span> y drift as a <span class="k">function</span> <span class="k">of</span> wavelenght to a straight
       line <span class="k">and</span> compensates <span class="k">for</span> it.





HISTORY<span class="o">:</span>
       2014<span class="o">-</span>01<span class="o">-</span>31 JW<span class="o">:</span> Created. Accurary is subpixel <span class="o">-</span> hopefully.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="66%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>refslice</td>
<td>int</td>
<td>[0,36]</td>
<td>20</td>
<td>reference slice to perform relative shifts</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_correct_adr_shift.pro</p>
</div>
<div class="section" id="divide-by-telluric-transmission">
<span id="dividebytellurictransmission"></span><span id="index-44"></span><h2>Divide by Telluric Transmission<a class="headerlink" href="#divide-by-telluric-transmission" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Divides a spectral data-cube by a telluric transmission calibration file.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,Calibration      <strong>Order</strong>: 2.5</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:   datacube calibrated for telluric absorption</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Corrects <span class="k">for</span> telluric transmission based on a telluric calibrator
       target reduced <span class="k">and</span> provied as the CalibrationFile.



HISTORY<span class="o">:</span>
       2009<span class="o">-</span>08<span class="o">-</span>27<span class="o">:</span> JM created
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>10<span class="o">-</span>09 JM added gpitv display
  2010<span class="o">-</span>10<span class="o">-</span>19 JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2011<span class="o">-</span>08<span class="o">-</span>01 MP<span class="o">:</span> Update <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS files
  2012<span class="o">-</span>10<span class="o">-</span>10 MP<span class="o">:</span> Minor code cleanup; remove deprecated suffix<span class="o">=</span> parameter
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired wavelength calibration file to be read</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_divide_by_telluric_transmission.pro</p>
</div>
<div class="section" id="interpolate-bad-pixels-in-cube">
<span id="interpolatebadpixelsincube"></span><span id="index-45"></span><h2>Interpolate bad pixels in cube<a class="headerlink" href="#interpolate-bad-pixels-in-cube" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Repair bad pixels by interpolating between their neighbors.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience, PolarimetricScience, Calibration      <strong>Order</strong>: 2.5</p>
<p><strong>Inputs</strong>:  Cube in either spectral or polarization mode</p>
<p><strong>Outputs</strong>:  Cube with bad pixels potentially found and cleaned up.      <strong>Output Suffix</strong>: &#8216;-bpfix&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Searches <span class="k">for</span> statistical outlier bad pixels in a cube <span class="k">and</span> replace them
       by interpolating between their neighbors.

 CAUTION<span class="o">:</span>
       Heuristic <span class="k">and</span> <span class="k">not</span> guaranteed <span class="k">or</span> tested in any way; this is more a
       convenience <span class="k">function</span> than a rigorous statistcally justified repair tool





HISTORY<span class="o">:</span>
       2013<span class="o">-</span>12<span class="o">-</span>14 MP<span class="o">:</span> Created as a convenience <span class="k">function</span> cleanup tool. Almost
       certainly <span class="k">not</span> the best algorithm <span class="o">-</span> just something quick <span class="k">and</span> good enough <span class="k">for</span>
       now?
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="5%" />
<col width="7%" />
<col width="7%" />
<col width="65%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>1</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
<tr class="row-even"><td>before_and_after</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Show the before-and-after images for the user to see? (for debugging/testing)</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_interpolate_bad_pixels_in_cube.pro</p>
</div>
<div class="section" id="extract-one-spectrum">
<span id="extractonespectrum"></span><span id="index-46"></span><h2>Extract one spectrum<a class="headerlink" href="#extract-one-spectrum" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Extract one spectrum from a datacube somewhere in the FOV specified by the user.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience HIDDEN      <strong>Order</strong>: 2.51</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:       <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>KEYWORDS<span class="o">:</span>
       <span class="o">/</span><span class="nb">Save</span>   Set to <span class="m">1</span> to <span class="nb">save</span> the output <span class="nb">image</span> to a disk file.

GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>FILTER,IFSUNIT
DRP KEYWORDS<span class="o">:</span> CUNIT,DATAFILE,SPECCENX,SPECCENY


HISTORY<span class="o">:</span>

  JM 2010<span class="o">-</span>03 <span class="o">:</span> created module.
  2012<span class="o">-</span>10<span class="o">-</span>17 MP<span class="o">:</span> Removed deprecated suffix keyword. FIXME this needs major
  cleanup!
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="6%" />
<col width="16%" />
<col width="7%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>xcenter</td>
<td>int</td>
<td>[0,1000]</td>
<td>141</td>
<td>x-locations in pixel on datacube where extraction will be made</td>
</tr>
<tr class="row-odd"><td>ycenter</td>
<td>int</td>
<td>[0,1000]</td>
<td>141</td>
<td>y-locations in pixel on datacube where extraction will be made</td>
</tr>
<tr class="row-even"><td>radius</td>
<td>float</td>
<td>[0,1000]</td>
<td><ol class="first last arabic simple" start="5">
<li></li>
</ol>
</td>
<td>Aperture radius (in pixel i.e. mlens) to extract photometry for each wavelength.</td>
</tr>
<tr class="row-odd"><td>method</td>
<td>string</td>
<td>[median|mean|total]</td>
<td>total</td>
<td>method of photometry extraction:median,mean,total</td>
</tr>
<tr class="row-even"><td>ps_figure</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose # of saved fig suffix name, 0: no ps figure</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output (fits) on disk, 0: don&#8217;t save</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: extract_one_spectrum.pro</p>
</div>
<div class="section" id="calibrate-photometric-flux">
<span id="calibratephotometricflux"></span><span id="index-47"></span><h2>Calibrate Photometric Flux<a class="headerlink" href="#calibrate-photometric-flux" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Apply photometric calibration to a single or set of datacubes</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 2.51</p>
<p><strong>Inputs</strong>:</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>: &#8216;-phot&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This primitive applies a spectrophotometric calibrations to the datacube that is determined either
       from the satellite spots <span class="k">of</span> the supplied cube, the satellite spots <span class="k">of</span> a
       user<span class="o">-</span>indicated cube, <span class="k">or</span> any user<span class="o">-</span>supplied spectral response <span class="k">function</span> (e.g. derived
       from an <span class="nb">open</span> loop <span class="nb">image</span> <span class="k">of</span> a standard star). Use <span class="k">of</span> this primitive is complicated, therefore a
       tutorial has been written to guide the user through it&#39;s use. The tutorial can be found as part
       <span class="k">of</span> the online documentation under the User&#39;s guide (http<span class="o">://</span>docs.planetimager.org<span class="o">/</span>pipeline<span class="o">/</span>usage<span class="o">/</span>index.html).

The user may also specify the extraction <span class="k">and</span> sky radii used in performing the aperture photometry. Note that the &#39;annuli&#39; only represent the radial <span class="nb">size</span> <span class="k">of</span> the extraction. The background is extracted by fitting a constant to an annulus surrounding the central star at the same radius as the planet. The inner width <span class="k">of</span> the annulus is equal to the inner_sky_radius, the outer annulus describes the distance from the companion to the edges <span class="k">of</span> the annulus that should be considered when fitting the constant. <span class="k">If</span> the user wishes to examine the section being fit, they should modify line 350 accordingly.

Error bars are calculated <span class="k">and</span> put into the headers to be used with future primitives such as gpi_extract_1d_spectrum. They are determined by convolving the sky annulus with the extraction aperture <span class="k">then</span> taking the standard deviation.


       <span class="m">1</span><span class="o">:</span> (REQUIRED) datacube that requires calibration (loaded as an Input FITS file). The satellites <span class="k">of</span> this <span class="nb">image</span> are used to determine the spectrophotometry calibration unless one <span class="k">of</span> the options below are used.
       2a (OPTIONAL)<span class="o">:</span> datacube <span class="k">or</span> to be used to determine the calibration <span class="o">-</span> this should be entered into the calib_cube_name parameter. This is meant to be used when the satellites <span class="k">of</span> a single cube are too low SNR <span class="k">and</span> a stack <span class="k">of</span> cubes must be used to get the SNR to high enough levels (which naturally blurs the companion).
       <span class="k">OR</span>
       <span class="m">2b</span> (OPTIONAL)<span class="o">:</span> a 2D spectrum (in ADU per COADD, <span class="nb">where</span> the COADD corresponds to input <span class="o">#</span><span class="m">1</span>) <span class="o">-</span> this should be entered using the calib_spectrum keyword. The file format must be three columns, the first being wavelength in microns, the second being flux in ADU per COADD, <span class="k">and</span> third being the uncertainty. THis is useful to calibrate a cube <span class="k">if</span> the calibration uses a different star as a calibrator.

The spectral type <span class="k">of</span> the star <span class="k">and</span> it&#39;s magnitude must be defined in the SPECTYPE <span class="k">and</span> HMAG header keywords. These should be set by default but sometimes are <span class="k">not</span> (<span class="k">or</span> are <span class="k">not</span> set correctly). One can modify them using the <span class="o">&lt;</span>Add Gemini <span class="k">and</span> GPI keywords<span class="o">&gt;</span> primitive. Note that this primitive is only visible when <span class="o">&lt;</span>Show all primitives<span class="o">&gt;</span> is selected from the <span class="o">&lt;</span>Options<span class="o">&gt;</span> dropdown menu in the recipe editor. Based on these values, the primitive searches <span class="k">for</span> an appropropriate pickles model <span class="k">for</span> the spectral type <span class="k">and</span> uses this to perform the calibration. <span class="k">If</span> the pickles model is <span class="k">not</span> appropriate, <span class="k">or</span> you wish to provide a different model, this model can be input using the &#39;calib_model_spectrum&#39; keyword.  The file format must be three columns, the first being wavelength in microns, the second being the flux in erg<span class="o">/</span>s<span class="o">/</span>cm2<span class="o">/</span>A, the third being the uncertainty.


Note that the calib_cube_name,calib_spectrum, <span class="k">and</span> calib_model_spectrum require the entire directory<span class="o">+</span>filename unless they are in the output directory


GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>FILTER,IFSUNIT
DRP KEYWORDS<span class="o">:</span> CUNIT,DATAFILE



HISTORY<span class="o">:</span>

  JM 2010<span class="o">-</span>03 <span class="o">:</span> created module.
  2012<span class="o">-</span>10<span class="o">-</span>17 MP<span class="o">:</span> Removed deprecated suffix keyword. needs major cleanup!
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
       2014<span class="o">-</span>01<span class="o">-</span>07 PI<span class="o">:</span> Created new gpi_calibrate_photometric_flux <span class="o">-</span> big overhaul from the original apply_photometric_calibration
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="5%" />
<col width="6%" />
<col width="6%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
<tr class="row-even"><td>extraction_radius</td>
<td>float</td>
<td>[0,1000]</td>
<td><ol class="first last arabic simple" start="3">
<li></li>
</ol>
</td>
<td>Aperture radius at middle wavelength (in spaxels i.e. mlens) to extract photometry for each wavelength.</td>
</tr>
<tr class="row-odd"><td>inner_sky_radius</td>
<td>float</td>
<td>[1,100]</td>
<td><ol class="first last arabic simple" start="10">
<li></li>
</ol>
</td>
<td>Inner aperture radius at middle wavelength (in spaxels i.e. mlens) to extract sky for each wavelength.</td>
</tr>
<tr class="row-even"><td>outer_sky_radius</td>
<td>float</td>
<td>[1,100]</td>
<td><ol class="first last arabic simple" start="25">
<li></li>
</ol>
</td>
<td>Outer aperture radius at middle wavelength (in spaxels i.e. mlens) to extract sky for each wavelength.</td>
</tr>
<tr class="row-odd"><td>c_ap_scaling</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Perform aperture scaling with wavelength?</td>
</tr>
<tr class="row-even"><td>calib_cube_name</td>
<td>string</td>
<td>None</td>
<td>&nbsp;</td>
<td>Leave blank to use satellites of this cube, or enter a file to use those satellites</td>
</tr>
<tr class="row-odd"><td>calib_model_spectrum</td>
<td>string</td>
<td>None</td>
<td>&nbsp;</td>
<td>Leave blank to use satellites of this cube, or enter a file to use with the spectrum for the satellites</td>
</tr>
<tr class="row-even"><td>calib_spectrum</td>
<td>string</td>
<td>None</td>
<td>&nbsp;</td>
<td>Leave blank to use satellites of this cube, or enter calibrated spectrum file</td>
</tr>
<tr class="row-odd"><td>FinalUnits</td>
<td>int</td>
<td>[0,10]</td>
<td>5</td>
<td>0: ADU per coadd, 1: ADU/s, 2: ph/s/nm/m^2, 3: Jy, 4: &#8216;W/m^2/um, 5: ergs/s/cm^2/A, 6: ergs/s/cm^2/Hz&#8217;</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_calibrate_photometric_flux.pro</p>
</div>
<div class="section" id="extract-one-spectrum-plots">
<span id="extractonespectrum-plots"></span><span id="index-48"></span><h2>Extract one spectrum, plots<a class="headerlink" href="#extract-one-spectrum-plots" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Extract one spectrum from a datacube somewhere in the FOV specified by the user.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience HIDDEN      <strong>Order</strong>: 2.51</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:       <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>KEYWORDS<span class="o">:</span>
       <span class="o">/</span><span class="nb">Save</span>   Set to <span class="m">1</span> to <span class="nb">save</span> the output <span class="nb">image</span> to a disk file.
KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>FILTER,IFSUNIT
DRP KEYWORDS<span class="o">:</span> CUNIT,DATAFILE,SPECCENX,SPECCENY



HISTORY<span class="o">:</span>

  JM 2010<span class="o">-</span>03 <span class="o">:</span> created module.
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="6%" />
<col width="16%" />
<col width="7%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>xcenter</td>
<td>int</td>
<td>[0,1000]</td>
<td>141</td>
<td>x-locations in pixel on datacube where extraction will be made</td>
</tr>
<tr class="row-odd"><td>ycenter</td>
<td>int</td>
<td>[0,1000]</td>
<td>141</td>
<td>y-locations in pixel on datacube where extraction will be made</td>
</tr>
<tr class="row-even"><td>radius</td>
<td>float</td>
<td>[0,1000]</td>
<td><ol class="first last arabic simple" start="5">
<li></li>
</ol>
</td>
<td>Aperture radius (in pixel i.e. mlens) to extract photometry for each wavelength.</td>
</tr>
<tr class="row-odd"><td>method</td>
<td>string</td>
<td>[median|mean|total]</td>
<td>total</td>
<td>method of photometry extraction:median,mean,total</td>
</tr>
<tr class="row-even"><td>ps_figure</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose # of saved fig suffix name, 0: no ps figure</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output (fits) on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>suffix</td>
<td>string</td>
<td>None</td>
<td>-spec</td>
<td>Enter output suffix (fits)</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: extract_one_spectrum2.pro</p>
</div>
<div class="section" id="extract-1d-spectrum-from-a-datacube">
<span id="extract1dspectrumfromadatacube"></span><span id="index-49"></span><h2>Extract 1D spectrum from a datacube<a class="headerlink" href="#extract-1d-spectrum-from-a-datacube" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Extract one spectrum from a datacube somewhere in the FOV specified by the user.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 2.52</p>
<p><strong>Inputs</strong>:  Datacube containing a source that needs extracting, located by the xcenter and ycenter arguments</p>
<p><strong>Outputs</strong>:  1D spectrum      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>WARNING<span class="o">:</span> This primitive will <span class="k">not</span> provide spectra <span class="k">of</span> publishable quality
it is designed to perform a quick extraction <span class="k">of</span> a source.


This primitive extracts a spectrum from a data cube. It is meant to be used
on datacubes that have been calibrated by gpi_apply_photometric_calibration,
but this is <span class="k">not</span> strictly required.

The extraction radius is pulled out <span class="k">of</span> the header such that is uses the
same as what was used to calibrate the cube. <span class="k">If</span> they keyword is <span class="k">not</span> found,
<span class="k">then</span> the extraction_radius keyword is used. The extraction_radius keyword will
also be used <span class="k">if</span> the override keyword is set to <span class="m">1</span>. Note that this is <span class="k">NOT</span>
recommended <span class="k">and</span> will introduce systematics into the data.

The centroiding is performed by fitting a gaussian to the region <span class="k">of</span> interest.
A line is <span class="k">then</span> fit to the centroids <span class="k">and</span> used. In this fit, the first <span class="k">and</span> last
<span class="m">4</span> data points are excluded due to low transmission. The errors <span class="k">for</span> each centroid
are determined by taking the largest <span class="k">of</span> the offsets between the subtraction <span class="k">of</span> adjacent
centroids (e.g. yerr[j]<span class="o">=</span><span class="m">0</span>.<span class="m">1</span><span class="o">&gt;</span><span class="nb">abs</span>(yarr0[j]<span class="o">-</span>yarr0[j<span class="o">+</span><span class="m">1</span>])<span class="o">&gt;</span><span class="nb">abs</span>(yarr0[j]<span class="o">-</span>yarr0[j<span class="o">-</span><span class="m">1</span>]) )

All photometry is done in ADU<span class="o">/</span>coadd. This is performed by converting the cube to
ADU<span class="o">/</span>coadd <span class="k">then</span> converting back to whatever units the cube was input with.

The error bars are determined using the same method as the satellite spots
in gpi_calibrate_photometric_flux primitive. The user specifies the sky radii used in performing the aperture photometry. Note that the &#39;annuli&#39; only represent the radial <span class="nb">size</span> <span class="k">of</span> the extraction. The background is extracted by fitting a constant to an annulus surrounding the central star at the same radius as the planet. The inner width <span class="k">of</span> the annulus is equal to the inner_sky_radius, the outer annulus describes the distance from the companion to the edges <span class="k">of</span> the annulus that should be considered when fitting the constant. <span class="k">If</span> the user wishes to examine the section being fit, they should modify line 350 accordingly.

Highpass filtering the <span class="nb">image</span> is recommended to determine the centroids, note that the highpass filtered <span class="nb">image</span>
is <span class="k">not</span> used when measuring the extracted spectrum.


KEYWORDS<span class="o">:</span>

<span class="nb">Save</span><span class="o">:</span> Set to <span class="m">1</span> to <span class="nb">save</span> the spectrum to a disk file (.fits).
xcenter<span class="o">:</span> x<span class="o">-</span>location <span class="k">of</span> extraction (in pixels)
ycenter<span class="o">:</span> y<span class="o">-</span>location <span class="k">of</span> extraction (in pixels)
highpass<span class="o">:</span> highpass filter the <span class="nb">image</span> when determining centroid?
inner_sky_radius<span class="o">:</span> inner radius used in defining sky subtraction annulus section
outer_sky_radius<span class="o">:</span> outer radius used in defining sky subtraction annulus section
override<span class="o">:</span> allows input <span class="k">of</span> a new extraction radius, <span class="k">and</span> the use<span class="o">/</span>non<span class="o">-</span>use <span class="k">of</span> c_ap_scaling
extraction_radius<span class="o">:</span> Radius used to define annulus <span class="k">for</span> source extraction. This keyword is only active <span class="k">if</span> the override keyword is set, <span class="k">or</span> <span class="k">if</span> the CEXTR_AP keyword, set by the Calibrate Photometric Flux primitive (gpi_calibrate_photometric_flux.<span class="k">pro</span>) is <span class="k">not</span> present in the header
c_ap_scaling<span class="o">:</span> keyword that activates the scaling <span class="k">of</span> the apertures with wavelength. This keyword is only active <span class="k">if</span> the override keyword is set, <span class="k">or</span> <span class="k">if</span> the C_AP_SC keyword, set by the Calibrate Photometric Flux primitive (gpi_calibrate_photometric_flux.<span class="k">pro</span>) is <span class="k">not</span> present in the header
display<span class="o">:</span> <span class="nb">window</span> used to display the extracted spectrum <span class="nb">plot</span>
save_ps_plot<span class="o">:</span> saves a postscript version <span class="k">of</span> the <span class="nb">plot</span> <span class="k">if</span> desired
write_ascii_file<span class="o">:</span> writes as ascii output <span class="k">of</span> the spectra <span class="o">-</span> no header info included

GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>FILTER,IFSUNIT
DRP KEYWORDS<span class="o">:</span> CUNIT,DATAFILE,SPECCENX,SPECCENY


HISTORY<span class="o">:</span>

  2014<span class="o">-</span>01<span class="o">-</span>07 PI<span class="o">:</span> Created Module <span class="o">-</span> big overhaul from the original extract 1d spectrum
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="4%" />
<col width="5%" />
<col width="5%" />
<col width="74%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>xcenter</td>
<td>float</td>
<td>[-1,280]</td>
<td>-1</td>
<td>x-location in pixels on datacube where extraction will be made</td>
</tr>
<tr class="row-even"><td>ycenter</td>
<td>float</td>
<td>[-1,280]</td>
<td>-1</td>
<td>y-location in pixels on datacube where extraction will be made</td>
</tr>
<tr class="row-odd"><td>highpass</td>
<td>int</td>
<td>[0,25]</td>
<td>0</td>
<td>highpass filter box size for centroiding</td>
</tr>
<tr class="row-even"><td>no_centroid_override</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Do not centroid on extraction source?</td>
</tr>
<tr class="row-odd"><td>inner_sky_radius</td>
<td>float</td>
<td>[1,100]</td>
<td><ol class="first last arabic simple" start="10">
<li></li>
</ol>
</td>
<td>Inner aperture radius at middle wavelength slice (in spaxels i.e. mlens) to extract sky</td>
</tr>
<tr class="row-even"><td>outer_sky_radius</td>
<td>float</td>
<td>[1,100]</td>
<td><ol class="first last arabic simple" start="20">
<li></li>
</ol>
</td>
<td>Outer aperture radius at middle wavelength slice (in spaxels i.e. mlens) to extract sky</td>
</tr>
<tr class="row-odd"><td>override</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Override apertures/scaling from spectrophotometric calibration?</td>
</tr>
<tr class="row-even"><td>extraction_radius</td>
<td>float</td>
<td>[0,1000]</td>
<td><ol class="first last arabic simple" start="5">
<li></li>
</ol>
</td>
<td>Aperture radius at middle wavelength (in spaxels i.e. mlens) to extract photometry for each wavelength. (only active if Override is set)</td>
</tr>
<tr class="row-odd"><td>c_ap_scaling</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Perform aperture scaling with wavelength?</td>
</tr>
<tr class="row-even"><td>display</td>
<td>int</td>
<td>[-1,100]</td>
<td>17</td>
<td>-1 = No display; 0 = New (unused) window; else = Window number to display diagnostic plot.</td>
</tr>
<tr class="row-odd"><td>save_ps_plot</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Save PostScript of plot?</td>
</tr>
<tr class="row-even"><td>write_ascii_file</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Save ascii file of spectrum (.dat)?</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_extract_1d_spectrum.pro</p>
</div>
<div class="section" id="collapse-datacube">
<span id="collapsedatacube"></span><span id="index-50"></span><h2>Collapse datacube<a class="headerlink" href="#collapse-datacube" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Collapse the wavelength dimension of a datacube via mean, median or total.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 2.6</p>
<p><strong>Inputs</strong>:  any datacube</p>
<p><strong>Outputs</strong>:  image containing that datacube collapsed to 2D      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> TODO<span class="o">:</span> more advanced collapse methods.


GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>
DRP KEYWORDS<span class="o">:</span> CDELT3, CRPIX3,CRVAL3,CTYPE3,NAXIS3


HISTORY<span class="o">:</span>
 2010<span class="o">-</span>04<span class="o">-</span>23 JM created
 2011<span class="o">-</span>07<span class="o">-</span>30 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
 2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> primitive rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="5%" />
<col width="13%" />
<col width="8%" />
<col width="62%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Method</td>
<td>enum</td>
<td>MEDIAN|TOTAL</td>
<td>TOTAL</td>
<td>How to collapse datacube: total or median (with flux conservation)</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>ReuseOutput</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: keep output for following primitives, 0: don&#8217;t keep</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_collapse_datacube.pro</p>
</div>
<div class="section" id="calibrate-astrometry-from-binary-using-separation-and-pa">
<span id="calibrateastrometryfrombinary-usingseparationandpa"></span><span id="index-51"></span><h2>Calibrate astrometry from binary (using separation and PA)<a class="headerlink" href="#calibrate-astrometry-from-binary-using-separation-and-pa" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Calculate astrometry from unocculted binaries using user-specified separation and PA at DATEOBS</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 2.6</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:   plate scale &amp; orientation      <strong>Output Suffix</strong>: &#8216;astrom&#8217; ; output suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>CRPA
DRP KEYWORDS<span class="o">:</span> FILETYPE,ISCALIB


HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2009<span class="o">-</span>12
       2013<span class="o">-</span>07<span class="o">-</span>19 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="6%" />
<col width="10%" />
<col width="8%" />
<col width="65%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>separation</td>
<td>float</td>
<td>[0.,4.]</td>
<td><ol class="first last arabic simple">
<li></li>
</ol>
</td>
<td>Separation [arcsec] at date DATEOBS of observation of the binaries</td>
</tr>
<tr class="row-odd"><td>pa</td>
<td>float</td>
<td>[0.,360.]</td>
<td>4.8</td>
<td>Position angle [degree] at date DATEOBS of observation of the binaries</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_calibrate_astrometry_from_binary_position.pro</p>
</div>
<div class="section" id="simple-spectral-differential-imaging">
<span id="simplespectraldifferentialimaging"></span><span id="index-52"></span><h2>Simple Spectral Differential Imaging<a class="headerlink" href="#simple-spectral-differential-imaging" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Apply SSDI to create a 2D subtracted image from a cube. Given the user&#8217;s specified wavelength ranges, extract the 3D datacube slices for each of those wavelength ranges. Collapse these down into 2D images by simply averaging the values at each lenslet (ignoring NANs).  Rescale Image1, then compute  diffImage = I1scaled - k* I2</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 2.61</p>
<p><strong>Inputs</strong>:</p>
<p><strong>Outputs</strong>:       <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This recipe rescales <span class="k">and</span> subtracts <span class="m">2</span> frames in different user<span class="o">-</span>defined
  bandwidths. This recipe is used <span class="k">for</span> speckle suppression using the
  Marois et al (2000) algorithm.

  This routine does <span class="k">NOT</span> update the data structures in <span class="nb">memory</span>. You <span class="o">**</span>MUST<span class="o">**</span>
  set the keyword <span class="nb">SAVE</span><span class="o">=</span><span class="m">1</span> <span class="k">or</span> <span class="k">else</span> the output is silently discarded.

    input datacube
    wavelength solution from <span class="k">common</span> block

KEYWORDS<span class="o">:</span>
    L1Min<span class="o">=</span>        Wavelength range <span class="m">1</span>, minimum wavelength [in microns]
    L1Max<span class="o">=</span>        Wavelength range <span class="m">1</span>, maximum wavelength [in microns]
    L2Min<span class="o">=</span>        Wavelength range <span class="m">2</span>, minimum wavelength [in microns]
    L2Max<span class="o">=</span>        Wavelength range <span class="m">2</span>, maximum wavelength [in microns]
    k<span class="o">=</span>            Multiplicative coefficient <span class="k">for</span> multiplying the <span class="nb">image</span> <span class="k">for</span>
                Wavelength Range <span class="o">*</span><span class="m">2</span><span class="o">*</span>. Default value is k<span class="o">=</span><span class="m">1</span>.

    <span class="o">/</span><span class="nb">Save</span>        Set to <span class="m">1</span> to <span class="nb">save</span> the output file to disk

DRP KEYWORDS<span class="o">:</span> CDELT3,CRPIX3,CRVAL3,CTYPE3,NAXIS3

ALGORITHM<span class="o">:</span>
   Given the user&#39;s specified wavelength ranges, extract the 3D datacube slices
   <span class="k">for</span> each <span class="k">of</span> those wavelength ranges. Collapse these down into 2D images by
   simply averaging the values at each lenslet (ignoring NANs).  Rescale Image1
   using fftscale so that the PSF scale matches that <span class="k">of</span> Image2 (as computed
   from the average wavelength <span class="k">for</span> each <span class="nb">image</span>). <span class="k">Then</span> compute
      diffImage <span class="o">=</span> I1scaled <span class="o">-</span> k<span class="o">*</span> I2
   <span class="k">Then</span> hopefully output the <span class="nb">image</span> somewhere <span class="k">if</span> <span class="nb">SAVE</span><span class="o">=</span><span class="m">1</span> is set.



HISTORY<span class="o">:</span>
  2007<span class="o">-</span>11 Jerome Maire
  2009<span class="o">-</span>04<span class="o">-</span>15 MDP<span class="o">:</span> Documentation updated; slight code cleanup
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="6%" />
<col width="10%" />
<col width="8%" />
<col width="65%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>L1Min</td>
<td>float</td>
<td>[0.9,2.5]</td>
<td>1.55</td>
<td>Wavelength range 1, minimum wavelength [in microns]</td>
</tr>
<tr class="row-odd"><td>L1Max</td>
<td>float</td>
<td>[0.9,2.5]</td>
<td>1.57</td>
<td>Wavelength range 1, maximum wavelength [in microns]</td>
</tr>
<tr class="row-even"><td>L2Min</td>
<td>float</td>
<td>[0.9,2.5]</td>
<td>1.60</td>
<td>Wavelength range 2, minimum wavelength [in microns]</td>
</tr>
<tr class="row-odd"><td>L2Max</td>
<td>float</td>
<td>[0.9,2.5]</td>
<td>1.65</td>
<td>Wavelength range 2, maximum wavelength [in microns]</td>
</tr>
<tr class="row-even"><td>k</td>
<td>float</td>
<td>[0,10]</td>
<td>1.0</td>
<td>Scaling factor of Intensity(wav_range2) with diffImage = I1scaled - k* I2</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>ReuseOutput</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: keep output for following primitives, 0: don&#8217;t keep</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_simple_spectral_differential_imaging.pro</p>
</div>
<div class="section" id="speckle-alignment">
<span id="specklealignment"></span><span id="index-53"></span><h2>Speckle alignment<a class="headerlink" href="#speckle-alignment" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This recipe rescales datacube PSF slices with respect to a chosen reference PSF slice.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 2.61</p>
<p><strong>Inputs</strong>:   Spectral datacube</p>
<p><strong>Outputs</strong>:  Resampled and rescaled spectral datacube      <strong>Output Suffix</strong>:  suffix+&#8217;-specalign&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This recipe rescales datacube slices with respect to a chosen reference slice.


ALGORITHM<span class="o">:</span>
       Given the user&#39;s specified wavelength ranges, extract the 3D datacube slices
       <span class="k">for</span> each <span class="k">of</span> those wavelength ranges.   Rescale slices with respect to a reference slice
       using fftscale so that the PSF scale matches that <span class="k">of</span> reference (as computed
       from the average wavelength <span class="k">for</span> each <span class="nb">image</span>).



HISTORY<span class="o">:</span>
       2012<span class="o">-</span>02 JM
  07.30.2012 <span class="o">-</span> offladed backend to speckle_align <span class="o">-</span> ds
  05.10.2012 <span class="o">-</span> updates to match other primitives <span class="k">and</span> to reflect
               changes to backend by Tyler Barker
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>k</td>
<td>int</td>
<td>[0,100]</td>
<td>0</td>
<td>Slice of the reference PSF</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_speckle_alignment.pro</p>
</div>
<div class="section" id="calibrate-astrometry-from-binary-using-6th-orbit-catalog">
<span id="calibrateastrometryfrombinary-using6thorbitcatalog"></span><span id="index-54"></span><h2>Calibrate astrometry from binary (using 6th orbit catalog)<a class="headerlink" href="#calibrate-astrometry-from-binary-using-6th-orbit-catalog" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Calculate astrometry from unocculted binaries; Calculate Separation and PA at date DATEOBS using the sixth orbit catalog.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 2.61</p>
<p><strong>Inputs</strong>:  spectral mode datacube, observing a reference binary</p>
<p><strong>Outputs</strong>:   plate scale &amp; orientation, saved as calibration file      <strong>Output Suffix</strong>: &#8216;astrom&#8217; ; output suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>CRPA,DATE<span class="o">-</span>OBS,OBJECT,TIME<span class="o">-</span>OBS
DRP KEYWORDS<span class="o">:</span> FILETYPE,ISCALIB


HISTORY<span class="o">:</span>
       Originally by Jerome Maire 2009<span class="o">-</span>12
       2013<span class="o">-</span>07<span class="o">-</span>19 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="9%" />
<col width="10%" />
<col width="13%" />
<col width="58%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_calibrate_astrometry_from_binary_orbit_catalog.pro</p>
</div>
<div class="section" id="measure-contrast">
<span id="measurecontrast"></span><span id="index-55"></span><h2>Measure Contrast<a class="headerlink" href="#measure-contrast" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Measure the contrast. Save as PNG or FITS table.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,PolarimetricScience      <strong>Order</strong>: 2.7</p>
<p><strong>Inputs</strong>:  Spectral mode datacube</p>
<p><strong>Outputs</strong>:    Contrast datacube, plot of contrast curve      <strong>Output Suffix</strong>: &#8216;-contr&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Measure, display on screen, <span class="k">and</span> optionally <span class="nb">save</span> the contrast.

  TODO <span class="o">-</span> should we revise this to call the same contrast measurement backend
  as GPItv?

  By default, the sat spots information are saved to the FITS header keywords
  <span class="k">of</span> the current file in <span class="nb">memory</span>, <span class="k">and</span> will only be saved <span class="k">if</span> you subsequently
  <span class="nb">save</span> that datacube (i.e. using &#39;<span class="nb">save</span><span class="o">=</span><span class="m">1</span>&#39; on this primitive <span class="k">or</span> a subsequent
  one). The &#39;update_prev_fits_header&#39; option will, in addition, also let you
  write the same keyword information to the header <span class="k">of</span> the most recently saved
  file. This is useful <span class="k">if</span> you have just already saved the datacube, <span class="k">and</span> you
  only now want to update this metadata.




HISTORY<span class="o">:</span>
       initial version imported GPItv (with definition <span class="k">of</span> contrast corrected) <span class="o">-</span> JM
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="5%" />
<col width="6%" />
<col width="7%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>Display</td>
<td>int</td>
<td>[-1,100]</td>
<td>-1</td>
<td>-1 = No display; 0 = New (unused) window; else = Window number to display in.</td>
</tr>
<tr class="row-even"><td>update_prev_fits_header</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Update FITS metadata in the most recently saved datacube?</td>
</tr>
<tr class="row-odd"><td>SaveProfile</td>
<td>string</td>
<td>None</td>
<td>&nbsp;</td>
<td>Save radial profile to filename as FITS (blank for no save, dir name for default naming, AUTO for auto full path)</td>
</tr>
<tr class="row-even"><td>SavePNG</td>
<td>string</td>
<td>None</td>
<td>&nbsp;</td>
<td>Save plot to filename as PNG (blank for no save, dir name for default naming, AUTO for auto full path)</td>
</tr>
<tr class="row-odd"><td>contrsigma</td>
<td>float</td>
<td>[0.,20.]</td>
<td><ol class="first last arabic simple" start="5">
<li></li>
</ol>
</td>
<td>Contrast sigma limit</td>
</tr>
<tr class="row-even"><td>slice</td>
<td>int</td>
<td>[-1,50]</td>
<td>-1</td>
<td>Slice to plot. -1 for all</td>
</tr>
<tr class="row-odd"><td>DarkHoleOnly</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>0: Plot profile in dark hole only; 1: Plot outer profile as well.</td>
</tr>
<tr class="row-even"><td>contr_yunit</td>
<td>int</td>
<td>[0,2]</td>
<td>0</td>
<td>0: Standard deviation; 1: Median; 2: Mean.</td>
</tr>
<tr class="row-odd"><td>contr_xunit</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>0: Arcsec; 1: lambda/D.</td>
</tr>
<tr class="row-even"><td>yscale</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>0: Auto y axis scaling; 1: Manual scaling.</td>
</tr>
<tr class="row-odd"><td>contr_yaxis_type</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>0: Linear; 1: Log</td>
</tr>
<tr class="row-even"><td>contr_yaxis_min</td>
<td>float</td>
<td>[0.,1.]</td>
<td>0.00000001</td>
<td>Y axis minimum</td>
</tr>
<tr class="row-odd"><td>contr_yaxis_max</td>
<td>float</td>
<td>[0.,1.]</td>
<td><ol class="first last arabic simple">
<li></li>
</ol>
</td>
<td>Y axis maximum</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_measure_contrast.pro</p>
</div>
<div class="section" id="plot-the-satellite-spot-locations-vs-the-expected-location-from-wavelength-scaling">
<span id="plotthesatellitespotlocationsvs-theexpectedlocationfromwavelengthscaling"></span><span id="index-56"></span><h2>Plot the satellite spot locations vs. the expected location from wavelength scaling<a class="headerlink" href="#plot-the-satellite-spot-locations-vs-the-expected-location-from-wavelength-scaling" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Measured vs. wavelength scaled sat spot locations</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 2.7</p>
<p><strong>Inputs</strong>:  Spectral mode datacube</p>
<p><strong>Outputs</strong>:  Plot of results.</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This is a quality check routine that verifies the expected wavelength
       scaling <span class="k">of</span> the datacube, based on how the satellite spots locations
       vary with wavelength.


KEYWORDS<span class="o">:</span>






HISTORY<span class="o">:</span>
       written 12<span class="o">/</span>11<span class="o">/</span>2012 <span class="o">-</span> ds
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="7%" />
<col width="9%" />
<col width="8%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Display</td>
<td>int</td>
<td>[-1,100]</td>
<td>1</td>
<td>-1 = No display; 0 = New (unused) window; else = Window number to display in.</td>
</tr>
<tr class="row-odd"><td>SaveData</td>
<td>string</td>
<td>None</td>
<td>&nbsp;</td>
<td>Save data to filename (blank for no save)</td>
</tr>
<tr class="row-even"><td>SavePNG</td>
<td>string</td>
<td>None</td>
<td>&nbsp;</td>
<td>Save plot to filename as PNG(blank for no save)</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_meas_satspot_dev.pro</p>
</div>
<div class="section" id="klip-algorithm-spectral-differential-imaging">
<span id="klipalgorithmspectraldifferentialimaging"></span><span id="index-57"></span><h2>KLIP algorithm Spectral Differential Imaging<a class="headerlink" href="#klip-algorithm-spectral-differential-imaging" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reduce speckle noise using the KLIP algorithm across the spectral axis of a datacube.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 2.8</p>
<p><strong>Inputs</strong>:  Spectral datacube</p>
<p><strong>Outputs</strong>:   Spectral datacube after SDI PSF subtraction      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This algorithm reduces PSF speckles in a datacube using the
  KLIP algorithm <span class="k">and</span> Spectral Differential Imaging.


ALGORITHM<span class="o">:</span>
      Measure annuli out from the center <span class="k">of</span> the cube <span class="k">and</span> create a
      reference set <span class="k">for</span> each annuli <span class="k">of</span> each slice. Apply KLIP to the
      reference set <span class="k">and</span> project the target slice onto the KL
      transform <span class="nb">vector</span>. Subtract the projected <span class="nb">image</span> from the
      original <span class="k">and</span> <span class="k">repeat</span> <span class="k">for</span> all slices


HISTORY<span class="o">:</span>
       Written 2013. Tyler Barker
       2013<span class="o">-</span>07<span class="o">-</span>18 MP<span class="o">:</span> Renamed <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="7%" />
<col width="11%" />
<col width="9%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>refslice</td>
<td>int</td>
<td>[0,100]</td>
<td>0</td>
<td>Slice of the reference PSF</td>
</tr>
<tr class="row-even"><td>annuli</td>
<td>int</td>
<td>[0,100]</td>
<td>5</td>
<td>Number of annuli used</td>
</tr>
<tr class="row-odd"><td>movement</td>
<td>float</td>
<td>[0.0,5.0]</td>
<td>2.0</td>
<td>Minimum pixel movement for reference set</td>
</tr>
<tr class="row-even"><td>prop</td>
<td>float</td>
<td>[0.8,1.0]</td>
<td>.99999</td>
<td>Proportion of eigenvalues used to truncate KL transform vectors</td>
</tr>
<tr class="row-odd"><td>arcsec</td>
<td>float</td>
<td>[0.0,1.0]</td>
<td>.4</td>
<td>Radius of interest if using 1 annulus</td>
</tr>
<tr class="row-even"><td>signal</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: calculate signal to noise ration, 0: don&#8217;t calculate</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_klip_algorithm_spectral_differential_imaging.pro</p>
</div>
<div class="section" id="update-world-coordinates">
<span id="updateworldcoordinates"></span><span id="index-58"></span><h2>Update World Coordinates<a class="headerlink" href="#update-world-coordinates" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Add wcs info, assuming target star is precisely centered.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 2.9</p>
<p><strong>Inputs</strong>:   Any datacube</p>
<p><strong>Outputs</strong>:  Datacube with updated WCS keywords in header      <strong>Output Suffix</strong>: &#8216;-addwcs&#8217;              ; output suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>   Creates a WCS<span class="o">-</span>compliant header based on the target star&#39;s RA <span class="k">and</span> DEC.
   Currently assumes the target star is precisely centered.


KEYWORDS<span class="o">:</span>
    CalibrationFile<span class="o">=</span>    Name <span class="k">of</span> astrometric binaries calibration file


GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>CRPA,RA,DEC
DRP KEYWORDS<span class="o">:</span> CDELT1,CDELT2,CRPIX1,CRPIX2,CRVAL1,CRVAL2,CTYPE1,CTYPE2,HISTORY,PC1_1,PC2_2,PSFCENTX,PSFCENTY


HISTORY<span class="o">:</span>
  JM 2009<span class="o">-</span>12
  JM 2010<span class="o">-</span>03<span class="o">-</span>10<span class="o">:</span> add handle <span class="k">of</span> x0,y0 ref point with PSFcenter
  2010<span class="o">-</span>10<span class="o">-</span>19 JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2011<span class="o">-</span>08<span class="o">-</span>01 MP<span class="o">:</span> Update <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
  2013<span class="o">-</span>07<span class="o">-</span>18 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
       2013<span class="o">-</span>12<span class="o">-</span>30 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>string</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Name of astrometry offset calibratoin file</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>0</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_update_world_coordinates.pro</p>
</div>
<div class="section" id="stores-calibration-in-dataset">
<span id="storescalibrationindataset"></span><span id="index-59"></span><h2>Stores calibration in dataset<a class="headerlink" href="#stores-calibration-in-dataset" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Stores the current calibration into the dataset structure.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 3.0</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>To be called before an accumulate <span class="nb">image</span>.
It is used <span class="k">for</span> high resolution microlens PSF determination

<span class="k">common</span> needed<span class="o">:</span>

KEYWORDS<span class="o">:</span>


HISTORY<span class="o">:</span>
    Originally by Jean<span class="o">-</span>Baptiste Ruffio 2013<span class="o">-</span>08
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: store_calib_in_dataset.pro</p>
</div>
<div class="section" id="normalize-polarimetry-flat-field">
<span id="normalizepolarimetryflatfield"></span><span id="index-60"></span><h2>Normalize polarimetry flat field<a class="headerlink" href="#normalize-polarimetry-flat-field" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Normalize a polarimetry-mode flat field to unity.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 3.1992</p>
<p><strong>Inputs</strong>:  polarimetry data-cube with flat lamp</p>
<p><strong>Outputs</strong>:   Normalized polarimetry mode flat field      <strong>Output Suffix</strong>: &#8216;polflat&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>
DRP KEYWORDS<span class="o">:</span>NAXES,NAXISi,FILETYPE,ISCALIB


HISTORY<span class="o">:</span>
       2009<span class="o">-</span>06<span class="o">-</span>20<span class="o">:</span> JM created
       2009<span class="o">-</span>07<span class="o">-</span>22<span class="o">:</span> MDP added doc header keywords
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>10<span class="o">-</span>09 JM added gpitv display
  2011<span class="o">-</span>07<span class="o">-</span>30 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_normalize_polarimetry_flat_field.pro</p>
</div>
<div class="section" id="create-lenslet-flat-field">
<span id="createlensletflatfield"></span><span id="index-61"></span><h2>Create Lenslet Flat Field<a class="headerlink" href="#create-lenslet-flat-field" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Create a 2D flat field for wavelength-independent lenslet throughput variations.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 3.1992</p>
<p><strong>Inputs</strong>:  Flat lamp data</p>
<p><strong>Outputs</strong>:  2D lenslet flat field file      <strong>Output Suffix</strong>: &#8216;lensletflat&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Creates a simple derived flat field <span class="k">for</span> non<span class="o">-</span>uniform transmission in the
       lenslets.

       WARNING<span class="o">:</span> experimental code, probably <span class="k">not</span> yet ready <span class="k">for</span> prime time.



HISTORY<span class="o">:</span>
   2014<span class="o">-</span>01<span class="o">-</span>02 MP<span class="o">:</span> Created
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_create_lenslet_flat_field.pro</p>
</div>
<div class="section" id="smooth-a-3d-cube">
<span id="smootha3dcube"></span><span id="index-62"></span><h2>Smooth a 3D Cube<a class="headerlink" href="#smooth-a-3d-cube" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Smooth a cube by convolution with a Gaussian kernel, repeated for each slice of the cube.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience,SpectralScience,Calibration      <strong>Order</strong>: 3.5</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:   smoothed datacube      <strong>Output Suffix</strong>:  &#8216;clean&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>Convolves images with a gaussian kernel

       Note<span class="o">:</span> This primitive will work properly either before <span class="k">or</span> after Accumulate
       Images. <span class="k">If</span> after, it will <span class="nb">smooth</span> all accumulated images.


GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>
DRP KEYWORDS<span class="o">:</span> HISTORY



HISTORY<span class="o">:</span>
  2014<span class="o">-</span>01<span class="o">-</span>09 MMB created
  2014<span class="o">-</span>04<span class="o">-</span>28 MP<span class="o">:</span> Minor documentation updates
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Smooth_FWHM</td>
<td>int</td>
<td>[0,100]</td>
<td>3</td>
<td>FWHM of gaussian kernel</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_smooth_cube.pro</p>
</div>
<div class="section" id="divide-by-polarized-flat-field">
<span id="dividebypolarizedflatfield"></span><span id="index-63"></span><h2>Divide by Polarized Flat Field<a class="headerlink" href="#divide-by-polarized-flat-field" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Divides a 2-slice polarimetry file by a flat field.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience, Calibration      <strong>Order</strong>: 3.5</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:   datacube with slices flat-fielded</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  <span class="o">**</span> Needs additional work, will <span class="k">not</span> produce high qualty results yet <span class="o">**</span>


GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>
DRP KEYWORDS<span class="o">:</span> HISTORY



HISTORY<span class="o">:</span>
  2009<span class="o">-</span>07<span class="o">-</span>22<span class="o">:</span> MDP created
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>10<span class="o">-</span>09 JM added gpitv display
  2010<span class="o">-</span>10<span class="o">-</span>19 JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2011<span class="o">-</span>07<span class="o">-</span>30 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>12<span class="o">-</span>16 MP<span class="o">:</span> CalibrationFile argument syntax update.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="8%" />
<col width="10%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>CalibrationFile</td>
<td>String</td>
<td>None</td>
<td>AUTOMATIC</td>
<td>Filename of the desired wavelength calibration file to be read</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_divide_by_polarized_flat_field.pro</p>
</div>
<div class="section" id="subtract-mean-stellar-polarization-from-podc">
<span id="subtractmeanstellarpolarizationfrompodc"></span><span id="index-64"></span><h2>Subtract Mean Stellar Polarization from podc<a class="headerlink" href="#subtract-mean-stellar-polarization-from-podc" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This description of the processing or calculation will show ; up in the Recipe Editor GUI. This is an example template for creating new ; primitives. It multiples any input cube by a constant value.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience      <strong>Order</strong>: 3.85</p>
<p><strong>Inputs</strong>:  Coronagraphic mode Stokes Datacube</p>
<p><strong>Outputs</strong>:  That datacube with an estimated stellar polarization subtracted off.      <strong>Output Suffix</strong>: &#8216;podc_sub&#8217;      ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  Subtract an estimate <span class="k">of</span> the stellar polarization, measured from
  the <span class="nb">mean</span> polarization inside the occulting spot radius.

  This primitive is simple, but has <span class="k">not</span> been extensively tested.
  Under what circumstances, <span class="k">if</span> any, it is useful on GPI data in practice
  is still TBD.








HISTORY<span class="o">:</span>
   2014<span class="o">-</span>03<span class="o">-</span>23 MP<span class="o">:</span> Started
   2014<span class="o">-</span>05<span class="o">-</span>14 MMB<span class="o">:</span> Rewrote <span class="k">for</span> podc
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="5%" />
<col width="8%" />
<col width="16%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Method</td>
<td>String</td>
<td>Auto|Manual</td>
<td>Auto</td>
<td>Choose where to meausre the inst_pol?</td>
</tr>
<tr class="row-odd"><td>InnerRadius</td>
<td>float</td>
<td>[-1,140]</td>
<td>-1</td>
<td>The inner radius where you start to measure the instrumental polarization. -1 = the size of the FPM</td>
</tr>
<tr class="row-even"><td>OuterRadius</td>
<td>float</td>
<td>[0,140]</td>
<td>20</td>
<td>The radius out to which you measure the instrumental polarization. 0 = inside the occulter</td>
</tr>
<tr class="row-odd"><td>WriteToFile</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: Write the difference to a file, 0: Dont</td>
</tr>
<tr class="row-even"><td>Filename</td>
<td>string</td>
<td>None</td>
<td>Stellar_Polarization.txt</td>
<td>The filename where you write out the stellar polarization</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_subtract_mean_stellar_polarization_podc.pro</p>
</div>
<div class="section" id="rotate-north-up">
<span id="rotatenorthup"></span><span id="index-65"></span><h2>Rotate North Up<a class="headerlink" href="#rotate-north-up" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Rotate datacubes so that north is up and east is left.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,PolarimetricScience      <strong>Order</strong>: 3.9</p>
<p><strong>Inputs</strong>:  Datacube(s) in either spectral or polarimetric mode</p>
<p><strong>Outputs</strong>:  Rotated datacube(s) with north up and east left.      <strong>Output Suffix</strong>:  &#8216;-northup&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  <span class="nb">Rotate</span> so that North is Up, <span class="k">and</span> east is to the left.
  <span class="k">If</span> necessary this will flip handedness as well as <span class="nb">rotate</span>
  to get the right parity in the output <span class="nb">image</span>.

       Note that this primitive can go <span class="o">*</span>either<span class="o">*</span> before <span class="k">or</span> after
       Accumulate Images. As a Level <span class="m">1</span> primitive, it will <span class="nb">rotate</span>
       one cube at a time; as a Level <span class="m">2</span> primitive it will <span class="nb">rotate</span>
       the whole stack <span class="k">of</span> accumulated images all at once (though
       it rotates each one by its own particular rotation angle).


KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>RA,DEC,PAR_ANG
DRP KEYWORDS<span class="o">:</span> CDELT1,CDELT2,CRPIX1,CRPIX2,CRVAL1,CRVAL2,NAXIS1,NAXIS2,PC1_1,PC1_2,PC2_1,PC2_2


HISTORY<span class="o">:</span>
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="7%" />
<col width="13%" />
<col width="7%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Rot_Method</td>
<td>string</td>
<td>CUBIC|FFT</td>
<td>CUBIC</td>
<td>Method to compute the rotation</td>
</tr>
<tr class="row-odd"><td>Center_Method</td>
<td>string</td>
<td>HEADERS|MANUAL</td>
<td>HEADERS</td>
<td>Determine the center of rotation from FITS header keywords, manual entry</td>
</tr>
<tr class="row-even"><td>centerx</td>
<td>int</td>
<td>[0,281]</td>
<td>140</td>
<td>Center X Pixel if Center_Method=Manual</td>
</tr>
<tr class="row-odd"><td>centery</td>
<td>int</td>
<td>[0,281]</td>
<td>140</td>
<td>Center Y Pixel if Center_Method=Manual</td>
</tr>
<tr class="row-even"><td>pivot</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Pivot about the center of the image? 0 = No</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>None</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_rotate_north_up.pro</p>
</div>
<div class="section" id="rotate-field-of-view-square">
<span id="rotatefieldofviewsquare"></span><span id="index-66"></span><h2>Rotate Field of View Square<a class="headerlink" href="#rotate-field-of-view-square" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Rotate datacubes so that the field of view is squarely aligned with the image axes.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience,PolarimetricScience      <strong>Order</strong>: 3.9</p>
<p><strong>Inputs</strong>:  datacube</p>
<p><strong>Outputs</strong>:  Rotated datacube      <strong>Output Suffix</strong>:  &#8216;_rot&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>   <span class="nb">Rotate</span> by the lenslet<span class="o">/</span>field relative angle, so that the GPI IFS
   field <span class="k">of</span> view is roughly square with the pixel coordinate axes.



KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>RA,DEC,PAR_ANG
DRP KEYWORDS<span class="o">:</span> CDELT1,CDELT2,CRPIX1,CRPIX2,CRVAL1,CRVAL2,NAXIS1,NAXIS2,PC1_1,PC1_2,PC2_1,PC2_2


HISTORY<span class="o">:</span>
  2012<span class="o">-</span>04<span class="o">-</span>10 MDP<span class="o">:</span> Created, based on rotate_north_up.<span class="k">pro</span>
  2013<span class="o">-</span>11<span class="o">-</span>07 ds <span class="o">-</span> updated to use gpi_update_wcs_basic
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="6%" />
<col width="11%" />
<col width="9%" />
<col width="66%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Method</td>
<td>enum</td>
<td>CUBIC|FFT</td>
<td>CUBIC</td>
<td>None</td>
</tr>
<tr class="row-odd"><td>crop</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Set to 1 to crop out non-illuminated pixels</td>
</tr>
<tr class="row-even"><td>Show</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>None</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>None</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_rotate_field_of_view_square.pro</p>
</div>
<div class="section" id="accumulate-images">
<span id="accumulateimages"></span><span id="index-67"></span><h2>Accumulate Images<a class="headerlink" href="#accumulate-images" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Stores images for combination by a subsequent primitive. Can buffer on disk for datasets too large to hold just in RAM.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 4.0</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>:       <strong>Output Suffix</strong>:  save_currdata( DataSet,  Modules[thisModuleIndex].OutputDir, suffix, display</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Stores images <span class="k">for</span> later combination.

<span class="k">common</span> needed<span class="o">:</span>

KEYWORDS<span class="o">:</span>


HISTORY<span class="o">:</span>
  2009<span class="o">-</span>07<span class="o">-</span>22 MDP<span class="o">:</span> started
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2013<span class="o">-</span>08, 2013<span class="o">-</span>10 MDP<span class="o">:</span> Minor code formatting cleanup
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="13%" />
<col width="28%" />
<col width="17%" />
<col width="28%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Method</td>
<td>string</td>
<td>OnDisk|InMemory</td>
<td>InMemory</td>
<td>OnDisk|InMemory</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_accumulate_images.pro</p>
</div>
<div class="section" id="create-2d-low-frequency-flat">
<span id="create2dlowfrequencyflat"></span><span id="index-68"></span><h2>Create 2D Low Frequency Flat<a class="headerlink" href="#create-2d-low-frequency-flat" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Create Low Frequency Flat 2D from polarimetry flats</div></blockquote>
<p><strong>Category</strong>:  HIDDEN      <strong>Order</strong>: 4.01</p>
<p><strong>Inputs</strong>:   Polarimetry flats</p>
<p><strong>Outputs</strong>:  Low frequency flat      <strong>Output Suffix</strong>:  &#8216;-LFFlat&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre><span class="o">/</span>!\ PROBABLY OUT <span class="k">OF</span> DATE <span class="o">/</span>!\ <span class="o">/</span>!\ PROBABLY OUT <span class="k">OF</span> DATE <span class="o">/</span>!\
That&#39;s why it is hidden. It was <span class="k">not</span> even ready <span class="k">for</span> release.
See new LF flat determination with the work in progress by JB <span class="k">and</span> Patrick on high resolution microlens PSF.
<span class="o">/</span>!\ PROBABLY OUT <span class="k">OF</span> DATE <span class="o">/</span>!\ <span class="o">/</span>!\ PROBABLY OUT <span class="k">OF</span> DATE <span class="o">/</span>!\

This primitive use a combined <span class="nb">image</span> <span class="k">of</span> polarimetry flat<span class="o">-</span>fields to build a low frequency flat <span class="k">for</span> the detector array.
The idea applied here is to integrate the flux <span class="k">of</span> each single spot using aperture photometry. <span class="k">For</span> every spot, the neighbors are masked <span class="k">and</span> the <span class="k">function</span> aper computes the <span class="nb">total</span> flux <span class="k">and</span> the sky<span class="o">/</span>background correction. The neighbors are masked in order to have enough pixels to compute the sky. <span class="k">Then</span>, we combine the flux <span class="k">of</span> every couple <span class="k">of</span> spots.
<span class="k">Then</span>, we divide all the values with their <span class="nb">median</span>.
Artifacts are removed by computing a local std deviation <span class="k">and</span> <span class="nb">median</span>. Every pixels further than n<span class="o">-</span>sigma were replaced by the value <span class="k">of</span> the smoothed flat.
To finish, we interpolates the resulting values over the 2048x2048 detector array using <span class="nb">triangulate</span><span class="o">/</span><span class="nb">trigrid</span> <span class="k">function</span> (linear interpolation using nearest neighbors).

There are still borders problem. The flat is <span class="k">not</span> valid near the edges.



HISTORY<span class="o">:</span>
    Originally by Jean<span class="o">-</span>Baptiste Ruffio 2013<span class="o">-</span>06
    2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Renamed <span class="k">for</span> consistency
         2013<span class="o">-</span>12<span class="o">-</span>03 MP<span class="o">:</span> Add check <span class="k">for</span> GCALLAMP<span class="o">=</span>QH on input images
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_create_2d_low_frequency_flat.pro</p>
</div>
<div class="section" id="create-high-resolution-microlens-psf-model">
<span id="createhigh-resolutionmicrolenspsfmodel"></span><span id="index-69"></span><h2>Create High-Resolution Microlens PSF Model<a class="headerlink" href="#create-high-resolution-microlens-psf-model" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Create a few calibrations files based on the determination of a high resolution PSF.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.01</p>
<p><strong>Inputs</strong>:   Multiple 2D images with appropriate illumination</p>
<p><strong>Outputs</strong>:  High resolution microlens PSF empirical model      <strong>Output Suffix</strong>:  &#8216;-&#8216;+filter+&#8217;-&#8216;+nrw_filt+&#8217;PSFs&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>This primitive is based on the determination <span class="k">of</span> a high resolution PSF <span class="k">for</span> each lenslet. It uses an adapted none iterative algorithm from the paper <span class="k">of</span> Jay Anderson <span class="k">and</span> Ivan R. King 2000.



HISTORY<span class="o">:</span>
    Originally by Jean<span class="o">-</span>Baptiste Ruffio 2013<span class="o">-</span>06
    2014<span class="o">-</span>01<span class="o">-</span>23 MP<span class="o">:</span> Rename <span class="k">and</span> documentation update
    2014<span class="o">-</span>04<span class="o">-</span>10 PI<span class="o">:</span> overhaul <span class="k">of</span> highres_psf creation
    2014<span class="o">-</span>04<span class="o">-</span>10 PI<span class="o">:</span> overhaul <span class="k">of</span> flexure handling
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="7%" />
<col width="8%" />
<col width="8%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>filter_wavelength</td>
<td>string</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>Narrowband filter wavelength</td>
</tr>
<tr class="row-odd"><td>bad_pixel_mask</td>
<td>string</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>Bad pixel mask</td>
</tr>
<tr class="row-even"><td>flat_field</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Is this a flat field</td>
</tr>
<tr class="row-odd"><td>flat_filename</td>
<td>string</td>
<td>None</td>
<td>&nbsp;</td>
<td>Name of flat field</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_create_highres_microlens_psf_model.pro</p>
</div>
<div class="section" id="create-microphonics-noise-model">
<span id="createmicrophonicsnoisemodel"></span><span id="index-70"></span><h2>Create microphonics noise model<a class="headerlink" href="#create-microphonics-noise-model" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Create a microphonics noise model in Fourier space.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.01</p>
<p><strong>Inputs</strong>:   several dark frames with strong microphonics</p>
<p><strong>Outputs</strong>:  microphonics model in Fourier space, saved as a calibration file      <strong>Output Suffix</strong>:  &#8216;-microModel&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> Create a microphonics noise model in Fourier space. The model consits only on the absolute value <span class="k">of</span> the Fourier coefficients.
 This has to be applied after the accumulate images primitive.

 <span class="k">For</span> each frame, it computes the absolute value <span class="k">of</span> the Fourier coefficients around the <span class="m">3</span> microphonics identified peaks.
 <span class="k">Then</span> it combines the results <span class="k">of</span> all frames based on the method defined by Combining method.
 Either adding all the models <span class="k">or</span> taking the <span class="nb">median</span>.
 <span class="k">Then</span> it normalizes the model.

 <span class="k">If</span> Gauss_Interp <span class="o">=</span> <span class="m">1</span><span class="o">:</span> Each <span class="k">of</span> the <span class="m">3</span> peaks is fitted by a 2d gaussian. Better <span class="k">not</span> using it. It doesn&#39;t give better results.



HISTORY<span class="o">:</span>
    Originally by Jean<span class="o">-</span>Baptiste Ruffio 2013<span class="o">-</span>05
    2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="10%" />
<col width="7%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Gauss_Interp</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: Interpolate each peak by a 2d gaussian, 0: don&#8217;t interpolate</td>
</tr>
<tr class="row-odd"><td>Combining_Method</td>
<td>string</td>
<td>ADD|MEDIAN</td>
<td>ADD</td>
<td>Method to combine the Fourier transforms of the microphonics (ADD|MEDIAN)</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_create_microphonics_noise_model.pro</p>
</div>
<div class="section" id="combine-2d-dark-images">
<span id="combine2ddarkimages"></span><span id="index-71"></span><h2>Combine 2D dark images<a class="headerlink" href="#combine-2d-dark-images" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Combine 2D dark images into a master file via mean or median.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.01</p>
<p><strong>Inputs</strong>:   several dark frames</p>
<p><strong>Outputs</strong>:  master dark frame, saved as a calibration file      <strong>Output Suffix</strong>:  &#8216;-dark&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> Several dark frames are combined to produce a master dark file, which
 is saved to the calibration database. This combination can be done using
 either a <span class="nb">mean</span> <span class="k">or</span> <span class="nb">median</span> algorithm, <span class="k">or</span> a <span class="nb">mean</span> with outlier
 rejection (sigma clipping)

 Also, based on the <span class="nb">variance</span> between the various dark frames, the
 <span class="nb">read</span> noise is estimated, <span class="k">and</span> a <span class="nb">list</span> <span class="k">of</span> hot pixels is derived.
 The <span class="nb">read</span> noise <span class="k">and</span> number <span class="k">of</span> significantly hot pixels are written
 as keywords to the FITS header <span class="k">for</span> use in trending analyses.




HISTORY<span class="o">:</span>
        Jerome Maire 2008<span class="o">-</span>10
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>10<span class="o">-</span>22 MDP<span class="o">:</span> Created from mediancombine_darks, converted to use
                               accumulator.
  2010<span class="o">-</span>01<span class="o">-</span>25 MDP<span class="o">:</span> Added support <span class="k">for</span> multiple methods, <span class="nb">MEAN</span> method.
  2010<span class="o">-</span>03<span class="o">-</span>08 JM<span class="o">:</span> ISCALIB flag <span class="k">for</span> Calib DB
  2011<span class="o">-</span>07<span class="o">-</span>30 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
       2013<span class="o">-</span>12<span class="o">-</span>15 MP<span class="o">:</span> Implemented SIGMACLIP, doc header updates.
  2014<span class="o">-</span>11<span class="o">-</span>04 MP<span class="o">:</span> Avoid trying to run parallelized sigmaclip <span class="k">if</span> in IDL runtime.
  2015<span class="o">-</span>02<span class="o">-</span>05 KBF<span class="o">:</span> <span class="nb">Fix</span> readnoise estimation
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="4%" />
<col width="12%" />
<col width="6%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Method</td>
<td>string</td>
<td>MEAN|MEDIAN|SIGMACLIP</td>
<td>SIGMACLIP</td>
<td>How to combine images: median, mean, or mean with outlier rejection?[MEAN|MEDIAN|SIGMACLIP]</td>
</tr>
<tr class="row-odd"><td>Sigma_cut</td>
<td>float</td>
<td>[1,100]</td>
<td>3</td>
<td>If Method=SIGMACLIP, then data points more than this many standard deviations away from the median value of a given pixel will be discarded.</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_combine_2d_dark_images.pro</p>
</div>
<div class="section" id="creates-a-thermal-sky-background-datacube">
<span id="createsathermal-skybackgrounddatacube"></span><span id="index-72"></span><h2>Creates a thermal/sky background datacube<a class="headerlink" href="#creates-a-thermal-sky-background-datacube" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Create Sky/Thermal background cubes</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.01</p>
<p><strong>Inputs</strong>:   A 2D sky image (should be a combination of several frames)</p>
<p><strong>Outputs</strong>:  A master sky frame, saved as a calibration file      <strong>Output Suffix</strong>:  &#8216;-bkgnd_cube&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>Create a thermal<span class="o">/</span>sky background cube (3D) rather than using 2D detector frames as is done using the Combine 2D Thermal<span class="o">/</span>Sky Backgrounds primitive. This allows a smoothing <span class="k">of</span> the sky frame that will decrease the photon noise.



HISTORY<span class="o">:</span>
  2013<span class="o">-</span>12<span class="o">-</span>23 PI<span class="o">:</span> Created Primitive
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="6%" />
<col width="8%" />
<col width="8%" />
<col width="62%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>smooth_box_size</td>
<td>int</td>
<td>[0,100]</td>
<td>3</td>
<td>Size of box to smooth by (0: No smooth)</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_create_sky_bkgd_cube.pro</p>
</div>
<div class="section" id="find-hot-bad-pixels-from-darks">
<span id="findhotbadpixelsfromdarks"></span><span id="index-73"></span><h2>Find Hot Bad Pixels from Darks<a class="headerlink" href="#find-hot-bad-pixels-from-darks" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Find hot pixels from a stack of dark images (best with deep integration darks)</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.01</p>
<p><strong>Inputs</strong>:  Multiple dark images</p>
<p><strong>Outputs</strong>:    Map of hot bad pixels      <strong>Output Suffix</strong>: &#8216;-hotpix&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>This is a variant <span class="k">of</span> combinedarkframes that (instead <span class="k">of</span> combining darks)
analyzes them to find hot pixels <span class="k">and</span> <span class="k">then</span> writes out a
mask showing <span class="nb">where</span> the various hot pixels are.


The current algorithm determines pixels that are hot according to the
criteria<span class="o">:</span>
       (a) dark count rate must be <span class="o">&gt;</span> <span class="m">1</span> e<span class="o">-/</span>second <span class="k">for</span> that pixel
       (b) that must be measured with <span class="o">&gt;</span><span class="m">5</span> sigma confidence above the estimated
           <span class="nb">read</span> noise <span class="k">of</span> the frames.
The first criterion can be adjusted using the hot_bad_thresh argument.




HISTORY<span class="o">:</span>
  2009<span class="o">-</span>07<span class="o">-</span>20 JM<span class="o">:</span> created
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2012<span class="o">-</span>01<span class="o">-</span>31 Switched sxaddpar to backbone<span class="o">-&gt;</span>set_keyword Dmitry Savransky
  2012<span class="o">-</span>11<span class="o">-</span>15 MP<span class="o">:</span> Algorithm entirely replaced with one based on combinedarkframes.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="6%" />
<col width="9%" />
<col width="8%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>hot_bad_thresh</td>
<td>float</td>
<td>[0,100.]</td>
<td>1.0</td>
<td>Threshhold to consider a hot pixel bad, in electrons/second.</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_find_hot_bad_pixels_from_darks.pro</p>
</div>
<div class="section" id="find-cold-bad-pixels-from-flats">
<span id="findcoldbadpixelsfromflats"></span><span id="index-74"></span><h2>Find Cold Bad Pixels from Flats<a class="headerlink" href="#find-cold-bad-pixels-from-flats" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Find cold pixels from a stack of flat images using all different filters</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.01</p>
<p><strong>Inputs</strong>:  Flat field images, preferably in multiple filters</p>
<p><strong>Outputs</strong>:  Map of cold bad pixels      <strong>Output Suffix</strong>: &#8216;-coldpix&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>This primitive finds cold (nonresponsive) pixels from the combination
<span class="k">of</span> flat fields at multiple wavelengths. This trickier than one might think,
because it&#39;s <span class="k">not</span> possible to illuminate the detector with an actual flat
field. The best you can <span class="k">do</span> is a flat through the lenslet array but that&#39;s
still <span class="k">not</span> very flat overall, with 37,000 spectra all over the place...

Instead, we can use a clever hack<span class="o">:</span> let&#39;s add up flat fields at various
different wavelengths, <span class="k">and</span> in the <span class="k">end</span> we should get something that&#39;s actually
at least got some light into all the pixels. But there&#39;s a huge ripple pattern
everwhere.

We <span class="k">then</span> rely on the symmetry <span class="k">of</span> the lenslet array to compare a given pixel to
one that should have pretty similar flux levels, <span class="k">and</span> we use that to find the
cold pixels.





HISTORY<span class="o">:</span>
 2013<span class="o">-</span>03<span class="o">-</span>08 MP<span class="o">:</span> Implemented in pipeline based on algorithm from Christian
 2013<span class="o">-</span>12<span class="o">-</span>03 MP<span class="o">:</span> Add check <span class="k">for</span> GCALLAMP<span class="o">=</span>QH on input images
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_find_cold_bad_pixels_from_flats.pro</p>
</div>
<div class="section" id="generate-combined-bad-pixel-map">
<span id="generatecombinedbadpixelmap"></span><span id="index-75"></span><h2>Generate Combined Bad Pixel Map<a class="headerlink" href="#generate-combined-bad-pixel-map" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This routine combines various sub-types of bad pixel mask (hot, cold,  anomalous nonlinear pixels) to generate a master bad pixel list.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.02</p>
<p><strong>Inputs</strong>:  bad pixel maps</p>
<p><strong>Outputs</strong>:  Combined bad pixel map      <strong>Output Suffix</strong>:  &#8216;-badpix&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>This routine is used to combine the <span class="m">3</span> individual types <span class="k">of</span> bad pixel maps<span class="o">::</span>

    Hot bad pixels
    Cold bad pixels
    Nonlinear (too nonlinear to be usable) pixels

into one master bad pixel <span class="nb">map</span>.

This is an unusual recipe, in that its input file data is <span class="k">not</span> actually
used in any way. All it does is use the first file to identify the date <span class="k">for</span>
which the bad pixel maps are generated.

<span class="k">For</span> this routine to run, there must be at least Hot <span class="k">and</span> Cold bad pixel maps
already present in the calibration DB. The nonlinear pixels <span class="nb">map</span> is optional.



HISTORY<span class="o">:</span>
   Jerome Maire 2009<span class="o">-</span>08<span class="o">-</span>10
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2012<span class="o">-</span>01<span class="o">-</span>31 Switched sxaddpar to backbone<span class="o">-&gt;</span>set_keyword Dmitry Savransky
  2012<span class="o">-</span>11<span class="o">-</span>19 MP<span class="o">:</span> Complete algorithm overhaul.
  2013<span class="o">-</span>04<span class="o">-</span>29 MP<span class="o">:</span> Better error checking; nonlinearbadpix is optional.
  2013<span class="o">-</span>07<span class="o">-</span>12 MP<span class="o">:</span> rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_generate_combined_bad_pixel_map.pro</p>
</div>
<div class="section" id="clean-polarization-pairs-via-double-difference">
<span id="cleanpolarizationpairsviadoubledifference"></span><span id="index-76"></span><h2>Clean Polarization Pairs via Double Difference<a class="headerlink" href="#clean-polarization-pairs-via-double-difference" title="Permalink to this headline">Â¶</a></h2>
<p><strong>Category</strong>:  PolarimetricScience,Calibration      <strong>Order</strong>: 4.05</p>
<p><strong>Inputs</strong>:   Multiple polarization pair datacubes</p>
<p><strong>Outputs</strong>:  Multiple polarization pair datacubes, hopefully with reduced      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Given a sequence <span class="k">of</span> polarization pair cubes, use a modified <span class="nb">double</span>
       differencing approach to mitigate systematics between the e<span class="o">-</span> <span class="k">and</span> o<span class="o">-</span> ray
       channels <span class="k">of</span> the cubes.

       This must be used after Accumulate Images. Unlike most such primitives, it
       acts on the entire stack <span class="k">of</span> cubes at once without combining them yet.

       This must be used prior to rotating the cubes <span class="k">if</span> it is to have any hope at
       all <span class="k">of</span> working well.

       <span class="o">**</span>Caution<span class="o">**</span> Experimental<span class="o">/</span>Under Development code <span class="o">-</span> algorithms may still be in
       flux.

                  systematics





HISTORY<span class="o">:</span>
       2013<span class="o">-</span>03<span class="o">-</span>20      Started by Marshall, forked from gpi_combine_polarizations_dd.<span class="k">pro</span>
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="4%" />
<col width="6%" />
<col width="6%" />
<col width="72%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>fix_badpix</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Also locate statistical outlier bad pixels and repair via interpolation?</td>
</tr>
<tr class="row-odd"><td>Save_diffbias</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Save the difference image systematic bias estimate subtracted from each pair?</td>
</tr>
<tr class="row-even"><td>gpitv_diffbias</td>
<td>int</td>
<td>[0,500]</td>
<td>10</td>
<td>Display empirical systematic bias in difference frames in a GPITV session 1-500, or 0 for  no display</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>debug</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Stop at breakpoints for debug/test</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_clean_polarization_pairs_dd.pro</p>
</div>
<div class="section" id="basic-adi">
<span id="basicadi"></span><span id="index-77"></span><h2>Basic ADI<a class="headerlink" href="#basic-adi" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Implements the basic ADI algorithm described by Marois et al. (2006).</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 4.1</p>
<p><strong>Inputs</strong>:</p>
<p><strong>Outputs</strong>:       <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>               ADI algorithm based on original Marois et al (2006) paper.



KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>FILTER,PAR_ANG,TELDIAM
DRP KEYWORDS<span class="o">:</span> PSFCENTX,PSFCENTY


HISTORY<span class="o">:</span>
        Adapted <span class="k">for</span> GPI <span class="o">-</span> Jerome Maire 2008<span class="o">-</span>08
   multiwavelength <span class="o">-</span> JM
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
   2010<span class="o">-</span>04<span class="o">-</span>26 JM<span class="o">:</span> verify how many spectral channels to process <span class="k">and</span> adapt ADI <span class="k">for</span> that,
               so we can use ADI on collapsed datacubes <span class="k">or</span> SDI
               outputs
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="6%" />
<col width="8%" />
<col width="8%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>numimmed</td>
<td>int</td>
<td>[1,100]</td>
<td>3</td>
<td>number of images for the calculation of the PSF reference</td>
</tr>
<tr class="row-odd"><td>nfwhm</td>
<td>float</td>
<td>[0,20]</td>
<td>1.5</td>
<td>number of FWHM to calculate the minimal distance for reference calculation</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>10</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_basic_adi.pro</p>
</div>
<div class="section" id="primitive-to-interface-tloci-code-and-dependecies-for-psf-subtraction-with-the-gpi-pipeline">
<span id="primitivetointerfacetlocicodeanddependeciesforpsfsubtractionwiththegpipipeline"></span><span id="index-78"></span><h2>Primitive to interface TLOCI code and dependecies for PSF subtraction with the GPI pipeline.<a class="headerlink" href="#primitive-to-interface-tloci-code-and-dependecies-for-psf-subtraction-with-the-gpi-pipeline" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Implements the TLOCI algorithm (Marois et al. 2014)</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 4.11</p>
<p><strong>Inputs</strong>:  data-cubes <em>spdc</em></p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>: &#8216;-tloci&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>
DRP KEYWORDS<span class="o">:</span>



HISTORY<span class="o">:</span>
        ZHD<span class="o">:</span> Intial creation.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="8%" />
<col width="10%" />
<col width="9%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>np</td>
<td>float</td>
<td>[0,20]</td>
<td>1</td>
<td>Number of processors to use.</td>
</tr>
<tr class="row-odd"><td>badpix</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Clean bad spaxels of each datacube?</td>
</tr>
<tr class="row-even"><td>unshrpmsk</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Use unsharp mask?</td>
</tr>
<tr class="row-odd"><td>register</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Image registration?</td>
</tr>
<tr class="row-even"><td>p_spec</td>
<td>string</td>
<td>[STRING]</td>
<td>None</td>
<td>Planet spectrum filename.</td>
</tr>
<tr class="row-odd"><td>s_spec</td>
<td>string</td>
<td>[STRING]</td>
<td>None</td>
<td>Stellar spectrum filename.</td>
</tr>
<tr class="row-even"><td>coeff</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Positive Coeffecients?</td>
</tr>
<tr class="row-odd"><td>Lambda</td>
<td>float</td>
<td>[0,1]</td>
<td>0.5</td>
<td>Lambda (Spectrum weighting)</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>10</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_tloci.pro</p>
</div>
<div class="section" id="adi-with-loci">
<span id="adiwithloci"></span><span id="index-79"></span><h2>ADI with LOCI<a class="headerlink" href="#adi-with-loci" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Implements the LOCI ADI algorithm (Lafreniere et al. 2007)</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 4.11</p>
<p><strong>Inputs</strong>:  data-cube</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>               ADI algorithm based on Lafreniere et al. 2007.


Code currently only offers the use <span class="k">of</span> positive <span class="k">and</span> negative coefficients.



KEYWORDS<span class="o">:</span>
GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>COADDS,CRFOLLOW,DEC,EXPTIME,HA,PAR_ANG
DRP KEYWORDS<span class="o">:</span> HISTORY,PSFCENTX,PSFCENTY



HISTORY<span class="o">:</span>
        Jerome Maire <span class="o">:-</span> multiwavelength 2008<span class="o">-</span>08
  JM<span class="o">:</span> adapted <span class="k">for</span> GPI<span class="o">-</span>pip
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2010<span class="o">-</span>04<span class="o">-</span>26 JM<span class="o">:</span> verify how many spectral channels to process <span class="k">and</span> adapt LOCI <span class="k">for</span> that,
               so we can use LOCI on collapsed datacubes <span class="k">or</span> SDI outputs
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible, added start_primitive
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="6%" />
<col width="8%" />
<col width="8%" />
<col width="67%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>nfwhm</td>
<td>float</td>
<td>[0,20]</td>
<td>1.5</td>
<td>number of FWHM to calculate the minimal distance for reference calculation</td>
</tr>
<tr class="row-odd"><td>coeff_type</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>0: positive and negative 1: positive only Coefficients in LOCI algorithm</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>10</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_adi_with_loci.pro</p>
</div>
<div class="section" id="combine-wavelength-calibrations-locations">
<span id="combinewavelengthcalibrationslocations"></span><span id="index-80"></span><h2>Combine Wavelength Calibrations locations<a class="headerlink" href="#combine-wavelength-calibrations-locations" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Combine wavelength calibration from  flat and arc</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.2</p>
<p><strong>Inputs</strong>:  3D wavcal</p>
<p><strong>Outputs</strong>:</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>gpi_combine_wavcal_all is a simple <span class="nb">median</span> combination <span class="k">of</span> wav. cal. files obtained with flat <span class="k">and</span> arc images.
 TO <span class="k">DO</span><span class="o">:</span> exclude some mlens from the <span class="nb">median</span> in <span class="k">case</span> <span class="k">of</span>  wavcal


GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>FILTER,IFSFILT
DRP KEYWORDS<span class="o">:</span> DATAFILE, DATE<span class="o">-</span>OBS,TIME<span class="o">-</span>OBS


HISTORY<span class="o">:</span>
   Jerome Maire 2009<span class="o">-</span>08<span class="o">-</span>10
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2012<span class="o">-</span>10<span class="o">-</span>17 MP<span class="o">:</span> Removed deprecated suffix<span class="o">=</span> keyword
  2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="65%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>polydegree</td>
<td>int</td>
<td>[1,2]</td>
<td>1</td>
<td>1: linear wavelength solution, 2: quadratic wav. sol.</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_combine_wavcal_locations_all.pro</p>
</div>
<div class="section" id="combine-wavelength-calibrations">
<span id="combinewavelengthcalibrations"></span><span id="index-81"></span><h2>Combine Wavelength Calibrations<a class="headerlink" href="#combine-wavelength-calibrations" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Performs simple median combination of wavelength calibrations from flat and/or arc lamps</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.2</p>
<p><strong>Inputs</strong>:  Multiple 3D wavcal cubes</p>
<p><strong>Outputs</strong>:  One merged 3D wavecal cube</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>gpi_combine_wavcal_all is a simple <span class="nb">median</span> combination <span class="k">of</span> wav. cal. files obtained with flat <span class="k">and</span> arc images.
 TO <span class="k">DO</span><span class="o">:</span> exclude some mlens from the <span class="nb">median</span> in <span class="k">case</span> <span class="k">of</span>  wavcal

 This is <span class="o">**</span>mostly deprecated<span class="o">**:</span> in general it is recommended to combine
 all 2D images <span class="k">for</span> a given arc lamp <span class="k">and</span> <span class="k">then</span> derive one wavelength solution
 from those, rather than deriving multiple wavecals <span class="k">and</span> <span class="k">then</span> combining them.
 On the other hand <span class="k">if</span> you want to merge wavecals from two different lamps <span class="k">then</span>
 you can indeed use this.


GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>DATE<span class="o">-</span>OBS,FILTER,IFSFILT,TIME<span class="o">-</span>OBS
DRP KEYWORDS<span class="o">:</span> DATAFILE


HISTORY<span class="o">:</span>
   Jerome Maire 2009<span class="o">-</span>08<span class="o">-</span>10
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2012<span class="o">-</span>10<span class="o">-</span>17 MP<span class="o">:</span> Removed deprecated suffix<span class="o">=</span> keyword
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_combine_wavelength_calibrations.pro</p>
</div>
<div class="section" id="klip-adi-for-pol-mode">
<span id="klipadiforpolmode"></span><span id="index-82"></span><h2>KLIP ADI for Pol Mode<a class="headerlink" href="#klip-adi-for-pol-mode" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reduce speckle noise using the KLIP algorithm with ADI data</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience      <strong>Order</strong>: 4.2</p>
<p><strong>Inputs</strong>:  Multiple spectral datacubes</p>
<p><strong>Outputs</strong>:  A reduced datacube with reduced PSF speckle halo      <strong>Output Suffix</strong>:  suffix+&#8217;-klip&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This algorithm reduces PSF speckles in a datacube using the
  KLIP algorithm <span class="k">and</span> Angular Differential Imaging in Pol Mode

ALGORITHM<span class="o">:</span>
      Star location must have been previously measured using satellite spots.
      Measure annuli out from the center <span class="k">of</span> the cube <span class="k">and</span> create a
      reference set <span class="k">for</span> each annuli <span class="k">of</span> each slice. Apply KLIP to the
      reference set <span class="k">and</span> project the target slice onto the KL
      transform <span class="nb">vector</span>. Subtract the projected <span class="nb">image</span> from the
      original <span class="k">and</span> <span class="k">repeat</span> <span class="k">for</span> all slices




HISTORY<span class="o">:</span>
       2013<span class="o">-</span>10<span class="o">-</span>21 <span class="o">-</span> ds
       2014<span class="o">-</span>03<span class="o">-</span>23 <span class="o">-</span> MMB<span class="o">:</span> Started adjusting <span class="k">for</span> pol mode
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="6%" />
<col width="12%" />
<col width="8%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>annuli</td>
<td>int</td>
<td>[0,100]</td>
<td>0</td>
<td>Number of annuli to use</td>
</tr>
<tr class="row-even"><td>MinRotation</td>
<td>float</td>
<td>[0.0,360.0]</td>
<td>1</td>
<td>Minimum rotation between images (degrees)</td>
</tr>
<tr class="row-odd"><td>CollapsePol</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Collapse the pol cube and perform KLIP on the total intensity?</td>
</tr>
<tr class="row-even"><td>prop</td>
<td>float</td>
<td>[0.8,1.0]</td>
<td>.99999</td>
<td>Proportion of eigenvalues used to truncate KL transform vectors</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_klip_algorithm_angular_differential_imaging_pol.pro</p>
</div>
<div class="section" id="populate-flexure-shifts-vs-elevation-table">
<span id="populateflexureshiftsvselevationtable"></span><span id="index-83"></span><h2>Populate Flexure Shifts vs Elevation Table<a class="headerlink" href="#populate-flexure-shifts-vs-elevation-table" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Derive shifts vs elevation lookup table.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.2</p>
<p><strong>Inputs</strong>: Not specified</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>:  &#8216;-shifts&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This <span class="k">function</span> produces the table <span class="k">of</span> spectral position shifts vs
       elevation angle used to compensate <span class="k">for</span> flexure within the GPI IFS.

       It takes as input a series <span class="k">of</span> reduced wavelength calibration
       files. It compares them against a reference wavelength calibration file
       obtained from the calibration database (which must have been taken in
       horizontal orientation).  The <span class="nb">shift</span> in X <span class="k">and</span> Y positions are calculated
       <span class="k">for</span> each lenslet, <span class="k">and</span> <span class="k">then</span> the <span class="nb">mean</span> over the entire field is taken.

       The <span class="nb">median</span> X <span class="nb">shift</span> <span class="k">and</span> Y <span class="nb">shift</span> <span class="k">for</span> each elevation is saved into
       a table in the calibration database.


  Optionally the user can request diagnostic <span class="nb">plots</span> displayed on screen <span class="k">or</span>
  saved to disk as postscript files.




HISTORY<span class="o">:</span>
     Jerome Maire 2013<span class="o">-</span>02
     2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="6%" />
<col width="8%" />
<col width="7%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>Display</td>
<td>int</td>
<td>[-1,100]</td>
<td>1</td>
<td>-1 = No display; 0 = New (unused) window else = Window number to display diagnostic plot in.</td>
</tr>
<tr class="row-even"><td>saveplots</td>
<td>string</td>
<td>[yes|no]</td>
<td>no</td>
<td>Save diagnostic plots to PS files after running? [yes|no]</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_populate_flexure_shifts_vs_elevation_table.pro</p>
</div>
<div class="section" id="klip-algorithm-angular-differential-imaging-with-center-forced">
<span id="klipalgorithmangulardifferentialimagingwithcenterforced"></span><span id="index-84"></span><h2>KLIP algorithm Angular Differential Imaging With Center Forced<a class="headerlink" href="#klip-algorithm-angular-differential-imaging-with-center-forced" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reduce speckle noise using the KLIP algorithm with ADI data with center forced to image center</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 4.2</p>
<p><strong>Inputs</strong>:  Multiple spectral datacubes</p>
<p><strong>Outputs</strong>:  A reduced datacube with reduced PSF speckle halo      <strong>Output Suffix</strong>:  suffix+&#8217;-klip&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This algorithm reduces noise in a datacube using the
  KLIP algorithm <span class="k">and</span> Angular Differential Imaging.
  This is the same as the &#39;KLIP algorithm Angular Differential Imaging&#39;
  primitive, except the location <span class="k">of</span> the star can be entered manually.
  This can be useful <span class="k">for</span> data with low S<span class="o">/</span>N <span class="nb">where</span> the satellite
  spots are <span class="k">not</span> found automatically.

  Used during first light run, may <span class="k">not</span> be that generally applicable
  now that automatic satellite spot finding is much more robust.


ALGORITHM<span class="o">:</span>
      Measure annuli out from the center <span class="k">of</span> the cube <span class="k">and</span> create a
      reference set <span class="k">for</span> each annuli <span class="k">of</span> each slice. Apply KLIP to the
      reference set <span class="k">and</span> project the target slice onto the KL
      transform <span class="nb">vector</span>. Subtract the projected <span class="nb">image</span> from the
      original <span class="k">and</span> <span class="k">repeat</span> <span class="k">for</span> all slices


HISTORY<span class="o">:</span>
       2013<span class="o">-</span>11<span class="o">-</span>16<span class="o">-</span> ds  Developed on the fly during first light run
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="6%" />
<col width="12%" />
<col width="8%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>annuli</td>
<td>int</td>
<td>[0,100]</td>
<td>0</td>
<td>Number of annuli to use</td>
</tr>
<tr class="row-even"><td>centerx</td>
<td>int</td>
<td>[0,281]</td>
<td>140</td>
<td>Center X Pixel</td>
</tr>
<tr class="row-odd"><td>centery</td>
<td>int</td>
<td>[0,281]</td>
<td>140</td>
<td>Center Y Pixel</td>
</tr>
<tr class="row-even"><td>MinRotation</td>
<td>float</td>
<td>[0.0,360.0]</td>
<td>1</td>
<td>Minimum rotation between images (degrees)</td>
</tr>
<tr class="row-odd"><td>prop</td>
<td>float</td>
<td>[0.8,1.0]</td>
<td>.99999</td>
<td>Proportion of eigenvalues used to truncate KL transform vectors</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_klip_algorithm_angular_differential_imaging_forcecent.pro</p>
</div>
<div class="section" id="klip-algorithm-angular-differential-imaging">
<span id="klipalgorithmangulardifferentialimaging"></span><span id="index-85"></span><h2>KLIP algorithm Angular Differential Imaging<a class="headerlink" href="#klip-algorithm-angular-differential-imaging" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reduce speckle noise using the KLIP algorithm with ADI data</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 4.2</p>
<p><strong>Inputs</strong>:  Multiple spectral datacubes</p>
<p><strong>Outputs</strong>:  A reduced datacube with reduced PSF speckle halo      <strong>Output Suffix</strong>:  suffix+&#8217;-klip&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This algorithm reduces PSF speckles in a datacube using the
  KLIP algorithm <span class="k">and</span> Angular Differential Imaging.

ALGORITHM<span class="o">:</span>
      Star location must have been previously measured using satellite spots.
      Measure annuli out from the center <span class="k">of</span> the cube <span class="k">and</span> create a
      reference set <span class="k">for</span> each annuli <span class="k">of</span> each slice. Apply KLIP to the
      reference set <span class="k">and</span> project the target slice onto the KL
      transform <span class="nb">vector</span>. Subtract the projected <span class="nb">image</span> from the
      original <span class="k">and</span> <span class="k">repeat</span> <span class="k">for</span> all slices




HISTORY<span class="o">:</span>
       2013<span class="o">-</span>10<span class="o">-</span>21 <span class="o">-</span> ds
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="6%" />
<col width="12%" />
<col width="8%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>annuli</td>
<td>int</td>
<td>[0,100]</td>
<td>0</td>
<td>Number of annuli to use</td>
</tr>
<tr class="row-even"><td>MinRotation</td>
<td>float</td>
<td>[0.0,360.0]</td>
<td>1</td>
<td>Minimum rotation between images (degrees)</td>
</tr>
<tr class="row-odd"><td>prop</td>
<td>float</td>
<td>[0.8,1.0]</td>
<td>.99999</td>
<td>Proportion of eigenvalues used to truncate KL transform vectors</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_klip_algorithm_angular_differential_imaging.pro</p>
</div>
<div class="section" id="advanced-klip-adi-for-pol-mode">
<span id="advancedklipadiforpolmode"></span><span id="index-86"></span><h2>Advanced KLIP ADI for Pol Mode<a class="headerlink" href="#advanced-klip-adi-for-pol-mode" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reduce speckle noise using the KLIP algorithm with ADI data</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience      <strong>Order</strong>: 4.2</p>
<p><strong>Inputs</strong>:  Multiple spectral datacubes</p>
<p><strong>Outputs</strong>:  A reduced datacube with reduced PSF speckle halo      <strong>Output Suffix</strong>:  suffix+&#8217;-klip&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This algorithm reduces PSF speckles in a datacube using the
  KLIP algorithm <span class="k">and</span> Angular Differential Imaging in Pol Mode

ALGORITHM<span class="o">:</span>
      Star location must have been previously measured using satellite spots.
      Measure annuli out from the center <span class="k">of</span> the cube <span class="k">and</span> create a
      reference set <span class="k">for</span> each annuli <span class="k">of</span> each slice. Apply KLIP to the
      reference set <span class="k">and</span> project the target slice onto the KL
      transform <span class="nb">vector</span>. Subtract the projected <span class="nb">image</span> from the
      original <span class="k">and</span> <span class="k">repeat</span> <span class="k">for</span> all slices




HISTORY<span class="o">:</span>
       2013<span class="o">-</span>10<span class="o">-</span>21 <span class="o">-</span> ds
       2014<span class="o">-</span>03<span class="o">-</span>23 <span class="o">-</span> MMB<span class="o">:</span> Started adjusting <span class="k">for</span> pol mode
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="10%" />
<col width="6%" />
<col width="10%" />
<col width="11%" />
<col width="62%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>annuli</td>
<td>int</td>
<td>[0,100]</td>
<td>0</td>
<td>Number of annuli to use</td>
</tr>
<tr class="row-even"><td>MinRotation</td>
<td>float</td>
<td>[0.0,360.0]</td>
<td>1</td>
<td>Minimum rotation between images (degrees)</td>
</tr>
<tr class="row-odd"><td>CollapsePol</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Collapse the pol cube and perform KLIP on the total intensity?</td>
</tr>
<tr class="row-even"><td>Mask</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Do you want to mask out any area? If so you must provide a list of mask files</td>
</tr>
<tr class="row-odd"><td>MaskList</td>
<td>string</td>
<td>None</td>
<td>MaskList.txt</td>
<td>The name of a file containing a list of mask files corresponding to each</td>
</tr>
<tr class="row-even"><td>prop</td>
<td>float</td>
<td>[0.8,1.0]</td>
<td>.99999</td>
<td>Proportion of eigenvalues used to truncate KL transform vectors</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_klip_adi_pol_adv.pro</p>
</div>
<div class="section" id="klip-algorithm-adi-sdi">
<span id="klipalgorithmadi-sdi"></span><span id="index-87"></span><h2>KLIP algorithm ADI + SDI<a class="headerlink" href="#klip-algorithm-adi-sdi" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Reduce speckle noise using the KLIP algorithm with ADI+SDI data</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 4.2</p>
<p><strong>Inputs</strong>:  Multiple spectral datacubes, ADR corrected</p>
<p><strong>Outputs</strong>:  A reduced datacube with reduced PSF speckle halo      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This algorithm reduces PSF speckles in a datacube using the
  KLIP algorithm <span class="k">and</span> Angular Differential Imaging <span class="o">+</span> Spectral
  Differential Imaging. Based on Soummer et al., 2012.

      Star location must have been previously measured using satellite spots.
      Measure annuli out from the center <span class="k">of</span> the cube <span class="k">and</span> create a
      reference set <span class="k">for</span> each annuli <span class="k">of</span> each slice. Apply KLIP to the
      reference set <span class="k">and</span> project the target slice onto the KL
      transform <span class="nb">vector</span>. Subtract the projected <span class="nb">image</span> from the
      original <span class="k">and</span> <span class="k">repeat</span> <span class="k">for</span> all slices

       <span class="o">***</span> Development code, <span class="k">not</span> intended <span class="k">for</span> public release <span class="o">***</span>
       <span class="o">***</span> Does <span class="k">not</span> work in compiled mode, but that&#39;s OK since <span class="k">not</span> intended <span class="k">for</span>
       public release <span class="o">****</span>





HISTORY<span class="o">:</span>
       2014<span class="o">-</span>05 JW modified code from Dmitry <span class="k">and</span> Tyler Barker
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="9%" />
<col width="5%" />
<col width="8%" />
<col width="7%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>numthreads</td>
<td>int</td>
<td>[1,10000]</td>
<td>5</td>
<td>Number of parallel processes to run KLIP calculation</td>
</tr>
<tr class="row-even"><td>annuli</td>
<td>int</td>
<td>[0,100]</td>
<td>5</td>
<td>Number of annuli to use</td>
</tr>
<tr class="row-odd"><td>subsections</td>
<td>int</td>
<td>[1,10]</td>
<td>4</td>
<td>Number of equal area subsections to break up each annulus into</td>
</tr>
<tr class="row-even"><td>prop</td>
<td>float</td>
<td>[0.8,1.0]</td>
<td>.99999</td>
<td>Proportion of eigenvalues used to truncate KL transform vectors</td>
</tr>
<tr class="row-odd"><td>minsep</td>
<td>float</td>
<td>[0.0,250]</td>
<td>3</td>
<td>Minimum separation between slices (pixels)</td>
</tr>
<tr class="row-even"><td>minPA</td>
<td>float</td>
<td>[0.0,360]</td>
<td>0.0</td>
<td>Minimum parallactic rotation (in degrees) for constructing reference PSF (good for disk targets)</td>
</tr>
<tr class="row-odd"><td>waveclip</td>
<td>int</td>
<td>[0,18]</td>
<td>2</td>
<td>Number of wavelength slices at the beginning and end of each cube to ignore</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_klip_adi_plus_sdi.pro</p>
</div>
<div class="section" id="simple-sdi-of-post-adi-residual">
<span id="simplesdiofpostadiresidual"></span><span id="index-88"></span><h2>Simple SDI of post ADI residual<a class="headerlink" href="#simple-sdi-of-post-adi-residual" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Apply SSDI to create a 2D subtracted image from a cube. Given the user&#8217;s specified wavelength ranges, extract the 3D datacube slices for each of those wavelength ranges. Collapse these down into 2D images by simply averaging the values at each lenslet (ignoring NANs).  Rescale Image1, then compute  diffImage = I1scaled - k* I2</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 4.3</p>
<p><strong>Inputs</strong>:   Datacube of ADI residuals</p>
<p><strong>Outputs</strong>:  Datacube with simple 2-wavelength SDI subtraction</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This recipe rescales <span class="k">and</span> subtracts <span class="m">2</span> frames in different user<span class="o">-</span>defined
  bandwidths. This recipe is used <span class="k">for</span> speckle suppression using the
  Marois et al (2000) algorithm.

  This routine does <span class="k">NOT</span> update the data structures in <span class="nb">memory</span>. You <span class="o">**</span>MUST<span class="o">**</span>
  set the keyword <span class="nb">SAVE</span><span class="o">=</span><span class="m">1</span> <span class="k">or</span> <span class="k">else</span> the output is silently discarded.

ALGORITHM<span class="o">:</span>
   Given the user&#39;s specified wavelength ranges, extract the 3D datacube slices
   <span class="k">for</span> each <span class="k">of</span> those wavelength ranges. Collapse these down into 2D images by
   simply averaging the values at each lenslet (ignoring NANs).  Rescale Image1
   using fftscale so that the PSF scale matches that <span class="k">of</span> Image2 (as computed
   from the average wavelength <span class="k">for</span> each <span class="nb">image</span>). <span class="k">Then</span> compute
      diffImage <span class="o">=</span> I1scaled <span class="o">-</span> k<span class="o">*</span> I2
   <span class="k">Then</span> hopefully output the <span class="nb">image</span> somewhere <span class="k">if</span> <span class="nb">SAVE</span><span class="o">=</span><span class="m">1</span> is set.






HISTORY<span class="o">:</span>
    2007<span class="o">-</span>11 Jerome Maire
   2009<span class="o">-</span>04<span class="o">-</span>15 MDP<span class="o">:</span> Documentation updated; slight code cleanup
   2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
   2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Renamed <span class="k">for</span> consistency
   2013<span class="o">-</span>08<span class="o">-</span>07 ds<span class="o">:</span> idl2 compiler compatible
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="6%" />
<col width="10%" />
<col width="8%" />
<col width="69%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>L1Min</td>
<td>float</td>
<td>[0.9,2.5]</td>
<td>1.55</td>
<td>Wavelength range 1, minimum wavelength [in microns]</td>
</tr>
<tr class="row-odd"><td>L1Max</td>
<td>float</td>
<td>[0.9,2.5]</td>
<td>1.57</td>
<td>Wavelength range 1, maximum wavelength [in microns]</td>
</tr>
<tr class="row-even"><td>L2Min</td>
<td>float</td>
<td>[0.9,2.5]</td>
<td>1.60</td>
<td>Wavelength range 2, minimum wavelength [in microns]</td>
</tr>
<tr class="row-odd"><td>L2Max</td>
<td>float</td>
<td>[0.9,2.5]</td>
<td>1.65</td>
<td>Wavelength range 2, maximum wavelength [in microns]</td>
</tr>
<tr class="row-even"><td>k</td>
<td>float</td>
<td>[0,10]</td>
<td>1.0</td>
<td>Scaling factor of Intensity(wav_range2) with diffImage = I1scaled - k* I2</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>5</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_simple_sdi_of_post_adi_residual.pro</p>
</div>
<div class="section" id="combine-polarization-sequence-via-double-difference">
<span id="combinepolarizationsequenceviadoubledifference"></span><span id="index-89"></span><h2>Combine Polarization Sequence via Double Difference<a class="headerlink" href="#combine-polarization-sequence-via-double-difference" title="Permalink to this headline">Â¶</a></h2>
<p><strong>Category</strong>:  PolarimetricScience,Calibration      <strong>Order</strong>: 4.4</p>
<p><strong>Inputs</strong>:  Multiple polarization pair datacubes</p>
<p><strong>Outputs</strong>:  a single Stokes datacube      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Combine a sequence <span class="k">of</span> polarized images via the SVD method, after first
       performing <span class="nb">double</span> differencing to remove systematics between the e<span class="o">-</span> <span class="k">and</span>
       o<span class="o">-</span>rays.

       See James Graham&#39;s SVD algorithm document, <span class="k">or</span> this algorithm may be hard to
       follow.  This is <span class="k">not</span> your father&#39;s imaging polarimeter any more!



       This routine assumes that it can <span class="nb">read</span> in a series <span class="k">of</span> files on disk which were written by
       the previous stage <span class="k">of</span> processing.



GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>EXPTIME,ISS_PORT,PAR_ANG,WPANGLE
DRP KEYWORDS<span class="o">:</span>CDELT3,CRPIX3,CRVAL3,CTYPE3,CUNIT3,DATAFILE,NAXISi,PC3_3





HISTORY<span class="o">:</span>
 2009<span class="o">-</span>07<span class="o">-</span>21<span class="o">:</span> MDP Started
   2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
   2013<span class="o">-</span>01<span class="o">-</span>30<span class="o">:</span> updated with some new keywords
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="6%" />
<col width="12%" />
<col width="8%" />
<col width="56%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>HWPoffset</td>
<td>float</td>
<td>[-360.,360.]</td>
<td>-29.14</td>
<td>The internal offset of the HWP. If unknown set to 0</td>
</tr>
<tr class="row-odd"><td>IncludeSystemMueller</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: Include, 0: Don&#8217;t</td>
</tr>
<tr class="row-even"><td>IncludeSkyRotation</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: Include, 0: Don&#8217;t</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>10</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_combine_polarizations_dd.pro</p>
</div>
<div class="section" id="combine-polarization-sequence">
<span id="combinepolarizationsequence"></span><span id="index-90"></span><h2>Combine Polarization Sequence<a class="headerlink" href="#combine-polarization-sequence" title="Permalink to this headline">Â¶</a></h2>
<p><strong>Category</strong>:  PolarimetricScience,Calibration      <strong>Order</strong>: 4.4</p>
<p><strong>Inputs</strong>:</p>
<p><strong>Outputs</strong>: Not specified      <strong>Output Suffix</strong>:  &#8220;-stokesdc&#8221;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Combine a sequence <span class="k">of</span> polarized images via the SVD method.

       See James Graham&#39;s SVD algorithm document, <span class="k">or</span> this algorithm may be hard to
       follow.  This is <span class="k">not</span> your father&#39;s imaging polarimeter any more!

       This routine assumes that it can <span class="nb">read</span> in a series <span class="k">of</span> files on disk which were written by
       the previous stage <span class="k">of</span> processing.



GEM<span class="o">/</span>GPI KEYWORDS<span class="o">:</span>EXPTIME,ISS_PORT,PAR_ANG,WPANGLE
DRP KEYWORDS<span class="o">:</span>CDELT3,CRPIX3,CRVAL3,CTYPE3,CUNIT3,DATAFILE,NAXISi,PC3_3
ALGORITHM<span class="o">:</span>





HISTORY<span class="o">:</span>
 2009<span class="o">-</span>07<span class="o">-</span>21<span class="o">:</span> MDP Started
 2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
 2013<span class="o">-</span>01<span class="o">-</span>30<span class="o">:</span> updated with some new keywords
 2014<span class="o">-</span>03 MP <span class="k">and</span> MM<span class="o">-</span>B<span class="o">:</span> Polarization coordinates <span class="k">and</span> angles verification <span class="k">and</span> debug
 2014<span class="o">-</span>08<span class="o">-</span>18 MPF<span class="o">:</span> some formatting cleanup, added sanity check, fixed
   which header was used <span class="k">for</span> output file
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="6%" />
<col width="12%" />
<col width="8%" />
<col width="56%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>HWPoffset</td>
<td>float</td>
<td>[-360.,360.]</td>
<td>-29.14</td>
<td>The internal offset of the HWP. If unknown set to 0</td>
</tr>
<tr class="row-odd"><td>IncludeSystemMueller</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: Include, 0: Don&#8217;t</td>
</tr>
<tr class="row-even"><td>IncludeSkyRotation</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: Include, 0: Don&#8217;t</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>10</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_combine_polarization_sequence.pro</p>
</div>
<div class="section" id="quality-check-wavelength-calibration">
<span id="qualitycheckwavelengthcalibration"></span><span id="index-91"></span><h2>Quality Check Wavelength Calibration<a class="headerlink" href="#quality-check-wavelength-calibration" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Performs a basic quality check on a wavecal based on the statistical distribution of measured inter-lenslet spacings.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.5</p>
<p><strong>Inputs</strong>:  3D wavecal</p>
<p><strong>Outputs</strong>:  if wavecal fails quality check, then recipe is failed</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       Quality check wavelength calibration<span class="o">:</span>
        Looks <span class="k">for</span> unexpected statistical anomalies <span class="k">or</span> offsets in the
        wavelength calibration, as implemented in the utility
        <span class="k">function</span> `gpi_wavecal_sanity_check`




HISTORY<span class="o">:</span>
  2013<span class="o">-</span>11<span class="o">-</span>28 MP<span class="o">:</span> Created.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="5%" />
<col width="10%" />
<col width="6%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>error_action</td>
<td>string</td>
<td>[Fail|Ask_user]</td>
<td>Ask_user</td>
<td>If the quality check fails, should the recipe immediately fail or should I alert the user and ask what they want to do?</td>
</tr>
<tr class="row-odd"><td>Display</td>
<td>int</td>
<td>[-1,100]</td>
<td>-1</td>
<td>-1 = No display; 0 = New (unused) window; else = Window number to display diagnostic plot.</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_quality_check_wavelength_calibration.pro</p>
</div>
<div class="section" id="combine-3d-datacubes">
<span id="combine3ddatacubes"></span><span id="index-92"></span><h2>Combine 3D Datacubes<a class="headerlink" href="#combine-3d-datacubes" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Combine 3D datacubes via mean or median.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 4.5</p>
<p><strong>Inputs</strong>:  3d datacubes</p>
<p><strong>Outputs</strong>:  a single combined datacube      <strong>Output Suffix</strong>:  strlowcase(method)</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre> Multiple 3D cubes can be combined into one, using either a <span class="nb">Mean</span> <span class="k">or</span> a <span class="nb">Median</span>.

 TODO<span class="o">:</span> more advanced combination methods. Improved sigma<span class="o">-</span>clipped <span class="nb">mean</span> implementation



HISTORY<span class="o">:</span>
        Jerome Maire 2008<span class="o">-</span>10
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2009<span class="o">-</span>10<span class="o">-</span>22 MDP<span class="o">:</span> Created from mediancombine_darks, converted to use
                               accumulator.
  2010<span class="o">-</span>01<span class="o">-</span>25 MDP<span class="o">:</span> Added support <span class="k">for</span> multiple methods, <span class="nb">MEAN</span> method.
  2011<span class="o">-</span>07<span class="o">-</span>30 MP<span class="o">:</span> Updated <span class="k">for</span> <span class="nb">multi</span><span class="o">-</span>extension FITS
  2012<span class="o">-</span>10<span class="o">-</span>10 MP<span class="o">:</span> Minor code cleanup
  2013<span class="o">-</span>07<span class="o">-</span>29 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="5%" />
<col width="25%" />
<col width="7%" />
<col width="56%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Method</td>
<td>enum</td>
<td>MEAN|MEDIAN|SIGMACLIP|MINIMUM</td>
<td>MEDIAN</td>
<td>How to combine images: median, mean, or mean with outlier rejection?</td>
</tr>
<tr class="row-odd"><td>sig_clip</td>
<td>int</td>
<td>0,10</td>
<td>3</td>
<td>Clipping value to be used with SIGMACLIP in sigma (stddev)</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_combine_3d_datacubes.pro</p>
</div>
<div class="section" id="median-combine-adi-datacubes">
<span id="mediancombineadidatacubes"></span><span id="index-93"></span><h2>Median Combine ADI datacubes<a class="headerlink" href="#median-combine-adi-datacubes" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Median combine all residual datacubes after an ADI (or LOCI) speckle suppression.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 4.5</p>
<p><strong>Inputs</strong>:  Many datacubes with ADI subtraction residuals</p>
<p><strong>Outputs</strong>:  Median combined residual datacube      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>    <span class="nb">Median</span> all ADI datacubes




HISTORY<span class="o">:</span>
   Jerome Maire <span class="o">:-</span> multiwavelength 2008<span class="o">-</span>08
   JM<span class="o">:</span> adapted <span class="k">for</span> GPI<span class="o">-</span>pip
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2010<span class="o">-</span>10<span class="o">-</span>19 JM<span class="o">:</span> split HISTORY keyword <span class="k">if</span> necessary
  2013<span class="o">-</span>07<span class="o">-</span>16 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>10</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_median_combine_adi_datacubes.pro</p>
</div>
<div class="section" id="pad-wavelength-calibration-edges">
<span id="padwavelengthcalibrationedges"></span><span id="index-94"></span><h2>Pad Wavelength Calibration Edges<a class="headerlink" href="#pad-wavelength-calibration-edges" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>pads the outer edges of the wavecal via extrapolation to cover lenslets whose spectra only partially fall on the detector field of view.</div></blockquote>
<p><strong>Category</strong>:  Calibration      <strong>Order</strong>: 4.6</p>
<p><strong>Inputs</strong>:  3D wavecal</p>
<p><strong>Outputs</strong>:  3D wavecal with extrapolated edges</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This primitive extrapolates from the lenslets near the edge <span class="k">of</span> the
       detector to provide estimated wavelength solutions <span class="k">for</span> the &#39;fractional&#39;
       lenslets, whose spectra only fall partially on the detector.

       Note that these fractional spectra are <span class="k">not</span> generally going to be
       useful <span class="k">for</span> scientific analyses, but <span class="k">for</span> some datacube processing
       tasks such as destriping it can be helpful to have wavelength
       solutions that cover these lenslets.





HISTORY<span class="o">:</span>
  2013<span class="o">-</span>11<span class="o">-</span>28 MP<span class="o">:</span> Created.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="4%" />
<col width="6%" />
<col width="6%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: Save output to disk, 0: Don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitvim_dispgrid</td>
<td>int</td>
<td>[0,500]</td>
<td>15</td>
<td>1-500: choose gpitv session for displaying image output and wavcal grid overplotted, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_pad_wavelength_calibration_edges.pro</p>
</div>
<div class="section" id="flexure-quicklook-for-spectra-lsqr-microlens-psf">
<span id="flexurequicklookforspectra-lsqr-microlenspsf"></span><span id="index-95"></span><h2>Flexure Quicklook for Spectra (Lsqr, microlens psf)<a class="headerlink" href="#flexure-quicklook-for-spectra-lsqr-microlens-psf" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive will extract flux from a 2D detector image into a GPI spectral cube using a least-square algorithm and microlenslet PSFs. Optionally can produce a residual detector image, solve for microphonics, and iterate the wavecal solution to find a minimum residual.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  2D detector image, wavecal, microlens PSF reference.</p>
<p><strong>Outputs</strong>:  None      <strong>Output Suffix</strong>: &#8216;quick_residual&#8217;              ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This primitive will extract flux from a 2D detector <span class="nb">image</span> into a GPI spectral cube using a least<span class="o">-</span>square, matrix inversion algorithm <span class="k">and</span> microlenslet PSFs.
       Optionally can produce a residual detector <span class="nb">image</span>, solve <span class="k">for</span> microphonics, <span class="k">and</span> iterate the wavecal solution to find a minimum residual.
       Ideally run in parrallel enviroment.





<span class="nb">where</span> in the order <span class="k">of</span> the primitives should this go by default?

pick one <span class="k">of</span> the following options <span class="k">for</span> the primitive type<span class="o">:</span>

HISTORY<span class="o">:</span>
   Began 2014<span class="o">-</span>01<span class="o">-</span>13 by Zachary Draper
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="6%" />
<col width="8%" />
<col width="8%" />
<col width="65%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>stopidl</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: stop IDL, 0: dont stop IDL</td>
</tr>
<tr class="row-odd"><td>x_lens</td>
<td>float</td>
<td>[0,281]</td>
<td>150</td>
<td>Lenslet number in x so start search</td>
</tr>
<tr class="row-even"><td>y_lens</td>
<td>float</td>
<td>[0,281]</td>
<td>150</td>
<td>Lenslet number in y so start search</td>
</tr>
<tr class="row-odd"><td>size</td>
<td>float</td>
<td>[0,30]</td>
<td>5</td>
<td>Size of region to iterate</td>
</tr>
<tr class="row-even"><td>resid</td>
<td>float</td>
<td>[0,1]</td>
<td>1</td>
<td>Save residual detector image?</td>
</tr>
<tr class="row-odd"><td>micphn</td>
<td>float</td>
<td>[0,1]</td>
<td>0</td>
<td>Solve for microphonics?</td>
</tr>
<tr class="row-even"><td>iter</td>
<td>float</td>
<td>[0,2]</td>
<td>1</td>
<td>Run iterative solver of wavecal?</td>
</tr>
<tr class="row-odd"><td>badpix</td>
<td>float</td>
<td>[0,1]</td>
<td>0</td>
<td>Weight by bad pixel map?</td>
</tr>
<tr class="row-even"><td>del_x_best</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for flexure perpendicular to dispersion shift (pixels)</td>
</tr>
<tr class="row-odd"><td>del_theta_best</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for rotation angle shift (degrees)</td>
</tr>
<tr class="row-even"><td>del_lam_best</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for flexure parrallel to dispersion shift (pixels)</td>
</tr>
<tr class="row-odd"><td>x_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Offset from wavecal in x pixels (used only if xcorrelariotn wasn&#8217;t run!)</td>
</tr>
<tr class="row-even"><td>y_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Offset from wavecal in y pixels</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_lsqr_mlens_quick_flx.pro</p>
</div>
<div class="section" id="flexure-quicklook-for-pol-lsqr-microlens-psf">
<span id="flexurequicklookforpol-lsqr-microlenspsf"></span><span id="index-96"></span><h2>Flexure Quicklook for Pol (Lsqr, microlens psf)<a class="headerlink" href="#flexure-quicklook-for-pol-lsqr-microlens-psf" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive will extract flux from a 2D detector image into a GPI spectral cube using a least-square algorithm and microlenslet PSFs. Optionally can produce a residual detector image, solve for microphonics, and iterate the wavecal solution to find a minimum residual.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  2D detector image, wavecal, microlens PSF reference.</p>
<p><strong>Outputs</strong>:  None      <strong>Output Suffix</strong>: &#8216;&#8217;            ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This primitive will extract flux from a 2D detector <span class="nb">image</span> into a GPI spectral cube using a least<span class="o">-</span>square, matrix inversion algorithm <span class="k">and</span> microlenslet PSFs.
       Optionally can produce a residual detector <span class="nb">image</span>, solve <span class="k">for</span> microphonics, <span class="k">and</span> iterate the wavecal solution to find a minimum residual.
       Ideally run in parrallel enviroment.





<span class="nb">where</span> in the order <span class="k">of</span> the primitives should this go by default?

pick one <span class="k">of</span> the following options <span class="k">for</span> the primitive type<span class="o">:</span>

HISTORY<span class="o">:</span>
   Began 2014<span class="o">-</span>01<span class="o">-</span>13 by Zachary Draper
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="6%" />
<col width="8%" />
<col width="8%" />
<col width="69%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>stopidl</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: stop IDL, 0: dont stop IDL</td>
</tr>
<tr class="row-even"><td>x_lens</td>
<td>float</td>
<td>[0,281]</td>
<td>150</td>
<td>Lenslet number in x so start search</td>
</tr>
<tr class="row-odd"><td>y_lens</td>
<td>float</td>
<td>[0,281]</td>
<td>150</td>
<td>Lenslet number in y so start search</td>
</tr>
<tr class="row-even"><td>size</td>
<td>float</td>
<td>[0,30]</td>
<td>4</td>
<td>Size of region to iterate</td>
</tr>
<tr class="row-odd"><td>resid</td>
<td>float</td>
<td>[0,1]</td>
<td>1</td>
<td>Save residual detector image?</td>
</tr>
<tr class="row-even"><td>micphn</td>
<td>float</td>
<td>[0,1]</td>
<td>0</td>
<td>Solve for microphonics?</td>
</tr>
<tr class="row-odd"><td>iter</td>
<td>float</td>
<td>[0,1]</td>
<td>1</td>
<td>Run iterative solver of wavecal?</td>
</tr>
<tr class="row-even"><td>del_x</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for flexure in detector x (pixels)</td>
</tr>
<tr class="row-odd"><td>del_y</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for flexure in detector y (pixels)</td>
</tr>
<tr class="row-even"><td>x_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Offset from wavecal in x pixels (used only if xcorrelariotn wasn&#8217;t run!)</td>
</tr>
<tr class="row-odd"><td>y_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Offset from wavecal in y pixels</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_lsqr_mlens_pol_quick_flx.pro</p>
</div>
<div class="section" id="subtract-mean-stellar-polarization">
<span id="subtractmeanstellarpolarization"></span><span id="index-97"></span><h2>Subtract Mean Stellar Polarization<a class="headerlink" href="#subtract-mean-stellar-polarization" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This description of the processing or calculation will show ; up in the Recipe Editor GUI. This is an example template for creating new ; primitives. It multiples any input cube by a constant value.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  Coronagraphic mode Stokes Datacube</p>
<p><strong>Outputs</strong>:  That datacube with an estimated stellar polarization subtracted off.      <strong>Output Suffix</strong>: &#8216;sub&#8217;</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>               Subtract an estimate <span class="k">of</span> the stellar polarization, measured from
               the <span class="nb">mean</span> polarization inside the occulting spot radius.

               This primitive is simple, but has <span class="k">not</span> been extensively tested.
               Under what circumstances, <span class="k">if</span> any, it is useful on GPI data in practice
               is still TBD.







HISTORY<span class="o">:</span>
   2014<span class="o">-</span>03<span class="o">-</span>23 MP<span class="o">:</span> Started
   2015<span class="o">-</span>02<span class="o">-</span>05 LWH<span class="o">:</span> Added more parameters
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="6%" />
<col width="9%" />
<col width="9%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_subtract_mean_stellar_polarization.pro</p>
</div>
<div class="section" id="assemble-polarization-datacube-lsqr-microlens-psf">
<span id="assemblepolarizationdatacube-lsqr-microlenspsf"></span><span id="index-98"></span><h2>Assemble Polarization Datacube (Lsqr, microlens psf)<a class="headerlink" href="#assemble-polarization-datacube-lsqr-microlens-psf" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive will extract flux from a 2D detector image into a GPI polarization cube using a least-square algorithm and microlenslet PSFs. Optionally can produce a residual detector image, solve for microphonics, and iterate the polcal solution to find a minimum residual.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  2D detector image, polcal, microlens PSF reference.</p>
<p><strong>Outputs</strong>:  GPI datacube      <strong>Output Suffix</strong>: &#8216;-podc&#8217;               ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This primitive will extract flux from a 2D detector <span class="nb">image</span> <span class="k">of</span> Wollaston spots into a GPI polarization cube using a least<span class="o">-</span>square, matrix inversion algorithm <span class="k">and</span> microlenslet PSFs.
       Optionally can produce a residual detector <span class="nb">image</span>, solve <span class="k">for</span> microphonics, <span class="k">and</span> iterate the polcal solution to find a minimum residual.
       Ideally run in parrallel enviroment.





<span class="nb">where</span> in the order <span class="k">of</span> the primitives should this go by default?

pick one <span class="k">of</span> the following options <span class="k">for</span> the primitive type<span class="o">:</span>

HISTORY<span class="o">:</span>
   Began 2014<span class="o">-</span>02<span class="o">-</span>17 by Zachary Draper
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="6%" />
<col width="8%" />
<col width="8%" />
<col width="71%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>stopidl</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>1: stop IDL, 0: dont stop IDL</td>
</tr>
<tr class="row-even"><td>np</td>
<td>float</td>
<td>[0,100]</td>
<td>2</td>
<td>Number of processors to use in reduction (double check enviroment before running)</td>
</tr>
<tr class="row-odd"><td>resid</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Save residual detector image?</td>
</tr>
<tr class="row-even"><td>micphn</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Solve for microphonics?</td>
</tr>
<tr class="row-odd"><td>iter</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Run iterative solver of polcal?</td>
</tr>
<tr class="row-even"><td>x_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Offset from wavecal in x pixels</td>
</tr>
<tr class="row-odd"><td>y_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Offset from wavecal in y pixels</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_lsqr_mlens_extract_pol.pro</p>
</div>
<div class="section" id="assemble-spectral-datacube-lsqr-microlens-psf">
<span id="assemblespectraldatacube-lsqr-microlenspsf"></span><span id="index-99"></span><h2>Assemble Spectral Datacube (Lsqr, microlens psf)<a class="headerlink" href="#assemble-spectral-datacube-lsqr-microlens-psf" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive will extract flux from a 2D detector image into a GPI spectral cube using a least-square algorithm and microlenslet PSFs. Optionally can produce a residual detector image, solve for microphonics, and iterate the wavecal solution to find a minimum residual.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  2D detector image, wavecal, microlens PSF reference.</p>
<p><strong>Outputs</strong>:  GPI datacube      <strong>Output Suffix</strong>: &#8216;spdc&#8217;                ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       This primitive will extract flux from a 2D detector <span class="nb">image</span> into a GPI spectral cube using a least<span class="o">-</span>square, matrix inversion algorithm <span class="k">and</span> microlenslet PSFs.
       Optionally can produce a residual detector <span class="nb">image</span>, solve <span class="k">for</span> microphonics, <span class="k">and</span> iterate the wavecal solution to find a minimum residual.
       Ideally run in parrallel enviroment.





<span class="nb">where</span> in the order <span class="k">of</span> the primitives should this go by default?

pick one <span class="k">of</span> the following options <span class="k">for</span> the primitive type<span class="o">:</span>

HISTORY<span class="o">:</span>
   Began 2014<span class="o">-</span>01<span class="o">-</span>13 by Zachary Draper
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="6%" />
<col width="7%" />
<col width="7%" />
<col width="67%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>stopidl</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: stop IDL, 0: dont stop IDL</td>
</tr>
<tr class="row-even"><td>np</td>
<td>float</td>
<td>[0,100]</td>
<td>4</td>
<td>Number of processors to use in reduction (double check enviroment before running)</td>
</tr>
<tr class="row-odd"><td>resid</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>Save residual detector image?</td>
</tr>
<tr class="row-even"><td>micphn</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Solve for microphonics?</td>
</tr>
<tr class="row-odd"><td>iter</td>
<td>int</td>
<td>[0,1]</td>
<td>0</td>
<td>Run iterative solver of wavecal?</td>
</tr>
<tr class="row-even"><td>badpix</td>
<td>float</td>
<td>[0,1]</td>
<td>0</td>
<td>Weight by bad pixel map?</td>
</tr>
<tr class="row-odd"><td>del_x_best</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for flexure perpendicular to dispersion shift (pixels)</td>
</tr>
<tr class="row-even"><td>del_theta_best</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for rotation angle shift (degrees)</td>
</tr>
<tr class="row-odd"><td>del_lam_best</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for flexure parrallel to dispersion shift (pixels)</td>
</tr>
<tr class="row-even"><td>x_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Offset from wavecal in x pixels</td>
</tr>
<tr class="row-odd"><td>y_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Offset from wavecal in y pixels</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_lsqr_mlens_extract.pro</p>
</div>
<div class="section" id="flexure-2d-x-correlation-with-wavecal-model-perpendicular">
<span id="flexure2dxcorrelationwithwavecalmodel-perpendicular"></span><span id="index-100"></span><h2>Flexure 2D x correlation with wavecal model (perpendicular)<a class="headerlink" href="#flexure-2d-x-correlation-with-wavecal-model-perpendicular" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive uses the relevent microlense PSF and wave cal to generate a model detector image to cross correlate with a science image.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  Science image, microlens PSF, wavecal</p>
<p><strong>Outputs</strong>:  Flexure offset in xy detector coordinates.      <strong>Output Suffix</strong>: &#8216;&#8217;              ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This primitive uses the relevent microlense PSF <span class="k">and</span> wave cal to generate a model detector <span class="nb">image</span> to cross <span class="nb">correlate</span> with a science <span class="nb">image</span>.
  The resulting output can be used as a flexure offset prior to flux extraction.



  The resulting output can be used as a flexure offset prior to flux extraction.


<span class="nb">where</span> in the order <span class="k">of</span> the primitives should this go by default?

pick one <span class="k">of</span> the following options <span class="k">for</span> the primitive type<span class="o">:</span>

HISTORY<span class="o">:</span>
   Began 2014<span class="o">-</span>01<span class="o">-</span>13 by Zachary Draper
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="11%" />
<col width="6%" />
<col width="7%" />
<col width="8%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>range</td>
<td>float</td>
<td>[0,5]</td>
<td>2</td>
<td>Range of cross corrleation search in pixels.</td>
</tr>
<tr class="row-odd"><td>resolution</td>
<td>float</td>
<td>[0,1]</td>
<td>0.01</td>
<td>Subpixel resolution of cross correlation convergence</td>
</tr>
<tr class="row-even"><td>psf_sep</td>
<td>float</td>
<td>[0,1]</td>
<td>0.01</td>
<td>PSF separation in pixels</td>
</tr>
<tr class="row-odd"><td>stopidl</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: stop IDL, 0: dont stop IDL</td>
</tr>
<tr class="row-even"><td>del_x_best</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>Best initial guess for flexure perpendicular to dispersion shift (pixels)</td>
</tr>
<tr class="row-odd"><td>badpix</td>
<td>float</td>
<td>[0,1]</td>
<td>0</td>
<td>Weight by bad pixel map?</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_img_xcorr_perp.pro</p>
</div>
<div class="section" id="flexure-2d-x-correlation-with-ulens-and-polcal-models">
<span id="flexure2dxcorrelationwithulensandpolcalmodels"></span><span id="index-101"></span><h2>Flexure 2D x correlation with ulens and polcal models<a class="headerlink" href="#flexure-2d-x-correlation-with-ulens-and-polcal-models" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive uses the relevent microlense PSF and pol cal to generate a model detector image to cross correlate with a science image and find the flexure offset.</div></blockquote>
<p><strong>Category</strong>:  PolarimetricScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  Science image, microlens PSF, wavecal</p>
<p><strong>Outputs</strong>:  Flexure offset in xy detector coordinates.      <strong>Output Suffix</strong>: &#8216;&#8217;              ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This primitive uses the relevent microlense PSF <span class="k">and</span> pol cal to generate a model detector <span class="nb">image</span> to cross <span class="nb">correlate</span> with a science <span class="nb">image</span>.
  The resulting output can be used as a flexure offset prior to flux extraction.



  The resulting output can be used as a flexure offset prior to flux extraction.


<span class="nb">where</span> in the order <span class="k">of</span> the primitives should this go by default?

pick one <span class="k">of</span> the following options <span class="k">for</span> the primitive type<span class="o">:</span>

HISTORY<span class="o">:</span>
   Began 2014<span class="o">-</span>01<span class="o">-</span>13 by Zachary Draper
         2014<span class="o">-</span>10<span class="o">-</span>01 MMB<span class="o">:</span> Changed name to distinguish from other version that JUST uses polcal file.
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="9%" />
<col width="10%" />
<col width="11%" />
<col width="56%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>range</td>
<td>float</td>
<td>[0,5]</td>
<td>2</td>
<td>Range of cross corrleation search in pixels.</td>
</tr>
<tr class="row-odd"><td>resolution</td>
<td>float</td>
<td>[0,1]</td>
<td>0.01</td>
<td>Subpixel resolution of cross correlation</td>
</tr>
<tr class="row-even"><td>stopidl</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: stop IDL, 0: dont stop IDL</td>
</tr>
<tr class="row-odd"><td>x_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>initial guess for large offsets</td>
</tr>
<tr class="row-even"><td>y_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>initial guess for large offsets</td>
</tr>
<tr class="row-odd"><td>badpix</td>
<td>float</td>
<td>[0,1]</td>
<td>0</td>
<td>Weight by bad pixel map?</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_img_xcorr_pol.pro</p>
</div>
<div class="section" id="flexure-2d-x-correlation-with-wavecal-model">
<span id="flexure2dxcorrelationwithwavecalmodel"></span><span id="index-102"></span><h2>Flexure 2D x correlation with wavecal model<a class="headerlink" href="#flexure-2d-x-correlation-with-wavecal-model" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive uses the relevent microlense PSF and wave cal to generate a model detector image to cross correlate with a science image.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  Science image, microlens PSF, wavecal</p>
<p><strong>Outputs</strong>:  Flexure offset in xy detector coordinates.      <strong>Output Suffix</strong>: &#8216;&#8217;              ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>  This primitive uses the relevent microlense PSF <span class="k">and</span> wave cal to generate a model detector <span class="nb">image</span> to cross <span class="nb">correlate</span> with a science <span class="nb">image</span>.
  The resulting output can be used as a flexure offset prior to flux extraction.



  The resulting output can be used as a flexure offset prior to flux extraction.


<span class="nb">where</span> in the order <span class="k">of</span> the primitives should this go by default?

pick one <span class="k">of</span> the following options <span class="k">for</span> the primitive type<span class="o">:</span>

HISTORY<span class="o">:</span>
   Began 2014<span class="o">-</span>01<span class="o">-</span>13 by Zachary Draper
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="7%" />
<col width="8%" />
<col width="9%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>range</td>
<td>float</td>
<td>[0,5]</td>
<td>2</td>
<td>Range of cross corrleation search in pixels.</td>
</tr>
<tr class="row-odd"><td>resolution</td>
<td>float</td>
<td>[0,1]</td>
<td>0.01</td>
<td>Subpixel resolution of cross correlation</td>
</tr>
<tr class="row-even"><td>yoff_factor</td>
<td>float</td>
<td>[0,5]</td>
<td>2</td>
<td>Fractor of resolution to which the code converges in y_offset</td>
</tr>
<tr class="row-odd"><td>psf_sep</td>
<td>float</td>
<td>[0,1]</td>
<td>0.01</td>
<td>PSF separation in pixels</td>
</tr>
<tr class="row-even"><td>stopidl</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: stop IDL, 0: dont stop IDL</td>
</tr>
<tr class="row-odd"><td>x_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>initial guess for large offsets</td>
</tr>
<tr class="row-even"><td>y_off</td>
<td>float</td>
<td>[-5,5]</td>
<td>0</td>
<td>initial guess for large offsets</td>
</tr>
<tr class="row-odd"><td>badpix</td>
<td>float</td>
<td>[0,1]</td>
<td>0</td>
<td>Weight by bad pixel map?</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_img_xcorr.pro</p>
</div>
<div class="section" id="insert-planet-into-datacube">
<span id="insertplanetintodatacube"></span><span id="index-103"></span><h2>Insert Planet into datacube<a class="headerlink" href="#insert-planet-into-datacube" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>This primitive inserts planets into reduced datacubes. It can be run multiple times to insert multiple planets.</div></blockquote>
<p><strong>Category</strong>:  SpectralScience      <strong>Order</strong>: 5.0</p>
<p><strong>Inputs</strong>:  A fully reduced datacube prior to any speckle manipulation. A planet&#8217;s distance, separation, position angle, mass, age, and formation scenario (hot/cold start).</p>
<p><strong>Outputs</strong>:  The datacube with an inserted planet.      <strong>Output Suffix</strong>: &#8216;wplnt&#8217;              ; set this to the desired output filename suffix</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>This primitive allows users to input artificial planets, based on the solar
metalicity, hybrid cloud model, hot<span class="o">/</span>cold formation scenario models <span class="k">of</span> Spiegel
<span class="k">and</span> Burrows (2012) into reduced datacubes. The planet PSF is represented by an
average <span class="k">of</span> the four satellite spots. The models span the ages <span class="k">of</span> <span class="m">1</span> to 1000
Myr, <span class="k">and</span> masses <span class="k">of</span> <span class="m">1</span><span class="o">-</span>15 Jupiter masses. <span class="k">If</span> a user specifies parameters that <span class="k">do</span>
<span class="k">not</span> represent an exact model the nearest model in age, <span class="k">then</span> mass, is used.
Currently, the intensity <span class="k">of</span> the planets is determined assuming an instrument
throughput <span class="k">of</span> 18.<span class="m">6</span>%, combined with a <span class="m">7</span>.<span class="m">9</span> meter primary mirror with a 1m
secondary. The user also has the option to scale the <span class="nb">image</span> to represent a star
<span class="k">of</span> a user<span class="o">-</span>defined magnitude. This provides the ability to simulate multiple
observing scenarios.

The stellar <span class="k">and</span> planet properties are written to the headers. Should the user
wish to <span class="k">not</span> include the planet information, it can be bypassed by
de<span class="o">-</span>activating the write_header_info keyword.

At the <span class="nb">moment</span> there is no way to determine only the star parameters <span class="k">and</span> <span class="k">not</span>
insert a planet. To <span class="k">do</span> this, the user should just put the planet distance to a
large number <span class="k">and</span> separation to a small number.


Note that the inserted separation <span class="k">and</span> position angle will be SLIGHTLY different
from the user specified values <span class="o">-</span> the proper values can be found in the header







HISTORY<span class="o">:</span>
   2013<span class="o">-</span>07<span class="o">-</span>30 PI<span class="o">:</span> Created Primitive
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="6%" />
<col width="9%" />
<col width="7%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Age</td>
<td>int</td>
<td>[1,1000]</td>
<td>10</td>
<td>Age of planet in Myr</td>
</tr>
<tr class="row-odd"><td>Mass</td>
<td>int</td>
<td>[1,15]</td>
<td>10</td>
<td>Mass of planet in Jupiter masses</td>
</tr>
<tr class="row-even"><td>model_type</td>
<td>string</td>
<td>[hot,cold]</td>
<td>hot</td>
<td>Hot or Cold Start formation scenario</td>
</tr>
<tr class="row-odd"><td>position_angle</td>
<td>float</td>
<td>[0,360]</td>
<td>45.0</td>
<td>Position angle of the planet in degrees East of North</td>
</tr>
<tr class="row-even"><td>Separation</td>
<td>float</td>
<td>[0,1800]</td>
<td>500</td>
<td>Separation in milli-arcseconds</td>
</tr>
<tr class="row-odd"><td>Star_Mag</td>
<td>float</td>
<td>[-1,8]</td>
<td>-1</td>
<td>Stellar Magnitude in H band, -1 estimates stellar magnitude from satellite spots</td>
</tr>
<tr class="row-even"><td>distance</td>
<td>float</td>
<td>[0,1000]</td>
<td>10.0</td>
<td>distance to system in parsecs</td>
</tr>
<tr class="row-odd"><td>write_header_info</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: Write planet info to headers 0: don&#8217;t write planet info to headers</td>
</tr>
<tr class="row-even"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-odd"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_insert_planet_into_cube.pro</p>
</div>
<div class="section" id="save-accumulated-stack">
<span id="saveaccumulatedstack"></span><span id="index-104"></span><h2>Save Accumulated Stack<a class="headerlink" href="#save-accumulated-stack" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Save output to disk as a FITS file. Note that you can often do this from another module by setting the &#8216;save=1&#8217; option; this is a redundant way to specify that.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 10.0</p>
<p><strong>Inputs</strong>:  Any</p>
<p><strong>Outputs</strong>:   The input is written to disk as a FITS file      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre><span class="nb">Save</span> the current accumulated stack <span class="k">of</span> images to disk.

Note that this uses whatever the currently defined suffix is, though you can
also override that here.  This is the one <span class="k">and</span> only routine that should be
used to override a suffix.

 TODO<span class="o">:</span> change output filename too, optionally?



HISTORY<span class="o">:</span>
  2014<span class="o">-</span>11<span class="o">-</span>18 Created by MMB
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="16%" />
<col width="14%" />
<col width="18%" />
<col width="37%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>suffix</td>
<td>string</td>
<td>None</td>
<td>None</td>
<td>choose the suffix</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_save_accumulated_stack.pro</p>
</div>
<div class="section" id="save-output">
<span id="saveoutput"></span><span id="index-105"></span><h2>Save Output<a class="headerlink" href="#save-output" title="Permalink to this headline">Â¶</a></h2>
<blockquote>
<div>Save output to disk as a FITS file. Note that you can often do this from another module by setting the &#8216;save=1&#8217; option; this is a redundant way to specify that.</div></blockquote>
<p><strong>Category</strong>:  ALL      <strong>Order</strong>: 10.0</p>
<p><strong>Inputs</strong>:  Any</p>
<p><strong>Outputs</strong>:   The input is written to disk as a FITS file      <strong>Output Suffix</strong>: Could not be determined automatically</p>
<p><strong>Notes</strong>:</p>
<div class="highlight-idl"><div class="highlight"><pre>       <span class="nb">Save</span> the current file to disk. Note that you can often <span class="k">do</span> this
       from another primitive by setting the &#39;<span class="nb">save</span><span class="o">=</span><span class="m">1</span>&#39; option; this is an
       optional, redundant way to specify that.

       Note that this uses whatever the currently defined suffix is, though you can
       also override that here.  This is the one <span class="k">and</span> only routine that should be
       used to override a suffix.

 TODO<span class="o">:</span> change output filename too, optionally?



HISTORY<span class="o">:</span>
       2009<span class="o">-</span>07<span class="o">-</span>21 Created by MDP.
  2009<span class="o">-</span>09<span class="o">-</span>17 JM<span class="o">:</span> added DRF parameters
  2013<span class="o">-</span>07<span class="o">-</span>17 MP<span class="o">:</span> Rename <span class="k">for</span> consistency
</pre></div>
</div>
<p><strong>Parameters</strong>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="8%" />
<col width="9%" />
<col width="9%" />
<col width="66%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Type</th>
<th class="head">Range</th>
<th class="head">Default</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>suffix</td>
<td>string</td>
<td>None</td>
<td>None</td>
<td>choose the suffix</td>
</tr>
<tr class="row-odd"><td>Save</td>
<td>int</td>
<td>[0,1]</td>
<td>1</td>
<td>1: save output on disk, 0: don&#8217;t save</td>
</tr>
<tr class="row-even"><td>gpitv</td>
<td>int</td>
<td>[0,500]</td>
<td>2</td>
<td>1-500: choose gpitv session for displaying output, 0: no display</td>
</tr>
</tbody>
</table>
<p><strong>IDL Filename</strong>: gpi_save_output.pro</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="logging.html" title="Logging and History records"
             >next</a> |</li>
        <li class="right" >
          <a href="calibdb.html" title="The Calibration Database"
             >previous</a> |</li>
        <li><a href="../index.html">GPI Data Pipeline 1.3.0 documentation</a> &raquo;</li>
          <li><a href="index.html" >GPI Data Pipeline User&#8217;s Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2008-2014, Marshall Perrin, Jerome Maire &amp; the GPI Data Analysis Team.
      Last updated on Jun 18, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>